{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, re, matplotlib, pandas, collections, importlib, sys, pickle, random, glob, dill, time\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "from importlib import reload\n",
    "from pathlib import Path\n",
    "\n",
    "\n",
    "top = Path('/Users/dp/pma/dataAndScripts/clip/meta/')\n",
    "top_dir = top\n",
    "\n",
    "os.chdir(top)\n",
    "\n",
    "sys.path.append('/Users/dp/pma/')\n",
    "import sameRiver\n",
    "import sameRiver.biotypeUtils\n",
    "import sameRiver.stacked_bargraph\n",
    "import sameRiver.readsPerGene\n",
    "import sameRiver.exp\n",
    "import sameRiver.rnaDataFileMaker\n",
    "\n",
    "importlib.reload(sameRiver.readsPerGene)\n",
    "importlib.reload(sameRiver.stacked_bargraph)\n",
    "\n",
    "from sameRiver.readsPerGene import *\n",
    "from sameRiver.stacked_bargraph import *\n",
    "\n",
    "##################################################################\n",
    "# 1. ann_counts is raw read counts. It also has biotypes added as the column 'Gene type'.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sameRiver.intronExonSplit\n",
    "import sameRiver.statsForCountsNB\n",
    "import sameRiver.biotypeDf\n",
    "import sameRiver.stacked_bargraph\n",
    "importlib.reload(sameRiver.biotypeUtils)\n",
    "importlib.reload(sameRiver.countsO)\n",
    "importlib.reload(sameRiver.stacked_bargraph)\n",
    "importlib.reload(sameRiver.statsForCountsNB)\n",
    "importlib.reload(sameRiver.readsPerGene)\n",
    "from sameRiver.readsPerGene import *\n",
    "\n",
    "\n",
    "import sameRiver.metadata.negative_metadata as negative_metadata\n",
    "import sameRiver.metadata.positive_metadata_all\n",
    "import sameRiver.negativeCounts\n",
    "import sameRiver.positiveCounts\n",
    "import sameRiver.scheme\n",
    "import sameRiver.statsForCountsNB\n",
    "importlib.reload(sameRiver.metadata.positive_metadata_all)\n",
    "importlib.reload(sameRiver.positiveCounts)\n",
    "importlib.reload(sameRiver.negativeCounts)\n",
    "importlib.reload(sameRiver.statsForCountsNB)\n",
    "importlib.reload(sameRiver.scheme)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def numeric_columns(_df):\n",
    "    return [x for x in _df.columns if (_df[x].dtype.kind in 'bifc')]\n",
    "\n",
    "\n",
    "figname = lambda _str: '/Users/dp/pma/dataAndScripts/clip/figs/{}.pdf'.format(_str)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sameRiver.biotypeArtist\n",
    "from pathlib import PurePath, Path\n",
    "\n",
    "importlib.reload(sameRiver.biotypeUtils)\n",
    "importlib.reload(sameRiver.biotypeArtist)\n",
    "importlib.reload(sameRiver.stacked_bargraph)\n",
    "importlib.reload(sameRiver.biotypeDf)\n",
    "\n",
    "# Load data and globals.\n",
    "#top = Path('/Users/dp/pma/dataAndScripts/clip/old_mapping_meta/')\n",
    "#os.chdir(top)\n",
    "print(f\"Working in {top}...\")\n",
    "\n",
    "scheme = sameRiver.scheme.scheme(Path(top, 'scheme.xlsx'))\n",
    "\n",
    "randos_rbps = [\n",
    "    'FBL', 'Rbfox1', 'Rbfox2', 'hnRNPC', \n",
    "    'hnRNPD', 'CELF1', 'SF3B1', 'PCBP1',\n",
    "    'UBA2', 'ETS2', 'CAPNS6', 'TPGS2', \n",
    "    'EPB41L5', 'CHMP3', 'CDK4', 'IDE', \n",
    "    'DCTN6', 'CCIN', 'ITPA', \n",
    "    'PCBP1 100P', 'PCBP1 100Q', 'PCBP1 dKH']\n",
    "randos = [\n",
    "    'UBA2', 'ETS2', 'CAPNS6', 'TPGS2', \n",
    "    'EPB41L5', 'CHMP3', 'CDK4', 'IDE', \n",
    "    'DCTN6', 'CCIN', 'ITPA', \n",
    "]\n",
    "rbps = [\n",
    "    'FBL', 'hnRNPC', 'SF3B1', 'PCBP1', \n",
    "    #'PCBP1:100P', 'PCBP1:100Q','PCBP1:dKH',\n",
    "    'CELF1', 'Rbfox1', 'Rbfox2', 'hnRNPD',\n",
    "    'KHDRBS2', 'A1CF', 'FUBP1', 'RPL5',#'CRNKL1',\n",
    "    ]\n",
    "# Select PCBP1s\n",
    "pcbp1_names = ['PCBP1', 'PCBP1:100P', 'PCBP1:100Q', 'PCBP1:dKH', 'Gene type']\n",
    "\n",
    "cfMuts = [\n",
    "    'KHDRBS2', 'KHDRBS2:R168C', 'A1CF', 'A1CF:E34K', 'FUBP1', 'FUBP1:R429C', 'RPL5', 'RPL5:E82K',#'PCBP1', 'PCBP1:100Q'\n",
    "]\n",
    "\n",
    "#with open(Path(top, 'data/stats.dill'), 'rb') as f:\n",
    "#    nb = dill.load(f)\n",
    "# Write ./data/positive_counts_per_read.txt:\n",
    "#nb.positives.save(write_txt=True, write_object=False)\n",
    "\n",
    "# Load and filter positive counts_per_read\n",
    "#df = pandas.read_csv('./data/positive_counts_per_read.txt', sep='\\t', index_col=0)   \n",
    "#df = nb.positives.counts_per_million_df.copy()\n",
    "#df = sameRiver.biotypeUtils.biotype_filter(df)\n",
    "\n",
    "# Load the read counts.\n",
    "rpg = readsPerGene(f'{top}/counts.txt')\n",
    "rpg.add_biotypes_column(\n",
    "    reload=True,\n",
    "    gtf_filename='/opt/genomes/repeats/repeats_and_genome.gtf',  # The newer, repeats/genome separate mapping. \n",
    "    #gtf_filename='/opt/genomes/repeats_and_ensembl_release94_GRCh38/combined_tsl1andNA.gtf' # The old mapping gtf. \n",
    ")\n",
    "df = rpg.df\n",
    "\n",
    "# Create a palette to be used for all figures.\n",
    "type_list = [#'rRNA', #'ribozyme', #'7SK', \n",
    "             #'processed_transcript', \n",
    "             'snoRNA', #'nonsense_mediated_decay', \n",
    "             'tRNA', 'scaRNA', \n",
    "             'lincRNA',\n",
    "             'snRNA', #'LTR',\n",
    "             'protein_coding', \n",
    "             'repeat', #'intergenic'\n",
    "            ]\n",
    "type_colors = sns.cubehelix_palette(n_colors=len(type_list), start=.5, rot=-2)\n",
    "\n",
    "# Lookup table of number (position in type list) -> color.\n",
    "group_to_color = {_t:type_colors[m] for m, _t in enumerate(type_list)}\n",
    "\n",
    "\n",
    "############################################################\n",
    "# Fraction of all reads by biotype bargraph.\n",
    "############################################################\n",
    "\n",
    "def prepare_to_plot(df):\n",
    "    \n",
    "    df = df.copy()\n",
    "\n",
    "    df = sameRiver.biotypeUtils.biotype_filter(df)\n",
    "    print(f\"From {set(df['Gene type'].tolist())},\")\n",
    "    \n",
    "    bio = sameRiver.biotypeDf.biotypeDf(df=df)\n",
    "    \n",
    "    ok_biotypes = bio.only_biotypes_with_a_large_enough_fraction_of_some_dataset(\n",
    "        minimum_percentage=3)\n",
    "    \n",
    "    print(f\"biotypes with a large enough fraction: {ok_biotypes}\")\n",
    "\n",
    "    fraction_all_reads = bio.fraction_of_value_by_biotype_flat()\n",
    "    fraction_all_reads.loc[:,'% reads (all RNAs)'] = fraction_all_reads['%']\n",
    "    fraction_all_reads.loc[:,'Protein'] = [scheme.gene_from_fname(x) for x in fraction_all_reads['Protein']]\n",
    "    fraction_all_reads = fraction_all_reads.groupby(['Protein', 'Gene type'])['% reads (all RNAs)'].mean().reset_index()\n",
    "\n",
    "    sample_order = fraction_all_reads[fraction_all_reads['Gene type'] == 'protein_coding']\n",
    "    sample_order.sort_values(by='% reads (all RNAs)', inplace=True)\n",
    "    return fraction_all_reads, sample_order['Protein'].tolist()\n",
    "\n",
    "def bargraphs(_df, proteins=[], name='', type_list=type_list):\n",
    "\n",
    "    # Select proteins.\n",
    "    df = _df.loc[:,[x for x in _df.columns if (scheme.gene_from_fname(x) in proteins) or x=='Gene type']]\n",
    "    \n",
    "    # Get rid of biotypes not in the type list.\n",
    "    df = df.loc[[x in type_list for x in df['Gene type']], :]\n",
    "\n",
    "    # Log what datasets were used.\n",
    "    os.makedirs(f'{top}/logs/', exist_ok=True)    \n",
    "    with open(f'{top}/logs/datasets_used_for_biotypes.py', 'a') as f:\n",
    "        f.write(f'datasets_for_{name} = ' + str(list(_df.columns)) + '\\n')\n",
    "\n",
    "    # Define color from the type list.\n",
    "    type_colors = sns.cubehelix_palette(n_colors=len(type_list), start=.5, rot=-2)\n",
    "\n",
    "    # Lookup table of number (position in type list) -> color.\n",
    "    group_to_color = {_t:type_colors[m] for m, _t in enumerate(type_list)}\n",
    "\n",
    "    # Get rid of biotypes not in the type list.\n",
    "    df_given_types = df.loc[[x in type_list for x in df['Gene type']], :]\n",
    "\n",
    "    print(f\"Before calling prepare to plot... {set(df['Gene type'].tolist())}\")\n",
    "    (fraction_all_reads, sample_order) = prepare_to_plot(df_given_types)\n",
    "\n",
    "    sameRiver.stacked_bargraph.stacked_bargraph(\n",
    "        df=fraction_all_reads, value_label='% reads (all RNAs)', sample_label='Protein',\n",
    "        hue_label='Gene type', title='% reads (all RNAs)',\n",
    "        sample_order=sample_order,#counts.proteins(),\n",
    "        group_to_color=group_to_color,\n",
    "        fname=figname(f'biotypes_{name}_all_rnas'))\n",
    "\n",
    "    # rRNA vs non-rRNA, keeping the order the same.\n",
    "    rRNA = _df.loc[:,[x for x in _df.columns if (scheme.gene_from_fname(x) in proteins) or x=='Gene type']]\n",
    "    rRNA['Gene type'] = [\n",
    "        {'rRNA': 'rRNA', 'intergenic': 'intergenic'}.get(x, 'non-rRNA') for x in rRNA['Gene type']]\n",
    "\n",
    "    type_list = ['rRNA', 'non-rRNA']#, 'intergenic']\n",
    "    type_colors = sns.cubehelix_palette(n_colors=len(type_list), start=.8, rot=-1.5)\n",
    "    group_to_color = {_t:type_colors[m] for m, _t in enumerate(type_list)}\n",
    "    \n",
    "    rRNA = rRNA.loc[[x in type_list for x in rRNA['Gene type']], :]\n",
    "    (fraction_all_reads, _) = prepare_to_plot(rRNA)\n",
    "    \n",
    "    sameRiver.stacked_bargraph.stacked_bargraph(\n",
    "        df=fraction_all_reads, value_label='% reads (all RNAs)', sample_label='Protein',\n",
    "        hue_label='Gene type', title='% reads (all RNAs)',\n",
    "        sample_order=sample_order,#counts.proteins(),\n",
    "        group_to_color=group_to_color,\n",
    "        fname=figname(f'biotypes_{name}_rRNA'), fig_height=2)\n",
    "\n",
    "    # Exon vs intron, keeping the order the same.\n",
    "    ex = _df.loc[:,[x for x in _df.columns if (scheme.gene_from_fname(x) in proteins) or x=='Gene type']]\n",
    "    # Only mRNA.\n",
    "    ex = ex.loc[[x=='protein_coding' for x in ex['Gene type']], :]\n",
    "    # Make the gene type column just the designation of intron or exon.\n",
    "    ex['Gene type'] = [x.split('::')[-1] for x in ex.index]\n",
    "    type_list = ['exon', 'intron']\n",
    "\n",
    "    type_colors = sns.cubehelix_palette(n_colors=len(type_list), start=7, rot=-1.4)\n",
    "    group_to_color = {_t:type_colors[m] for m, _t in enumerate(type_list)}\n",
    "    \n",
    "    ex = ex.loc[[x in type_list for x in ex['Gene type']], :]\n",
    "    (fraction_all_reads, _) = prepare_to_plot(ex)\n",
    "    \n",
    "    print(fraction_all_reads)\n",
    "    \n",
    "    sameRiver.stacked_bargraph.stacked_bargraph(\n",
    "        df=fraction_all_reads, value_label='% reads (all RNAs)', sample_label='Protein',\n",
    "        hue_label='Gene type', title='% reads (all RNAs)',\n",
    "        sample_order=sample_order,#counts.proteins(),\n",
    "        group_to_color=group_to_color,\n",
    "        fname=figname(f'biotypes_{name}_perc_intronic'), fig_height=2)\n",
    "    \n",
    "with open('./logs/datasets_used_for_biotypes.py', 'w') as f:\n",
    "    f.write('# Written by biotypes.ipynb.')  # Empty the file so it can be appended to.\n",
    "\n",
    "# Plot RBPs.\n",
    "bargraphs(df, proteins=cfMuts, name='Mutants')\n",
    "bargraphs(df, proteins=rbps, name='RBPs')\n",
    "bargraphs(df, proteins=randos, name='non-RBPs')\n",
    "bargraphs(df, proteins=pcbp1_names, name='PCBP1')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create a counts file of counts collapsed by biotype and perform statistics."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sameRiver.readsPerGene\n",
    "importlib.reload(sameRiver.readsPerGene)\n",
    "from sameRiver.readsPerGene import *\n",
    "# Class for creating a new metadata namespace, if wanted. (Not used currently, I think.)\n",
    "class Namespace:\n",
    "    def __init__(self, **kwargs): self.__dict__.update(kwargs)\n",
    "\n",
    "# Import object with file paths and protein names.\n",
    "import sameRiver.metadata.negative_metadata as negative_metadata\n",
    "import sameRiver.metadata.positive_metadata_all as positive_metadata\n",
    "\n",
    "# Change the counts filename to be one collapsed by biotype.\n",
    "positive_metadata.ann_counts_file = f\"{top}/data/biotype_counts.txt\"\n",
    "negative_metadata.ann_counts_file = f\"{top}/data/biotype_counts.txt\"\n",
    "\n",
    "# Load counts.\n",
    "rpg = rawReadsPerGene(f'{top_dir}/counts.txt', scheme_filename=f'{top_dir}/scheme.xlsx')\n",
    "rpg.add_biotypes_column(\n",
    "    gtf_filename='/Users/dp/pma/repeats_and_genome.gtf', reload=True)\n",
    "\n",
    "# Write a version of the counts file collapsed by biotype.\n",
    "a = rpg.df.groupby('Gene type').apply(np.sum) # Series of Gene type -> {Col: samples, Rows: sums}\n",
    "a = pandas.DataFrame(a)\n",
    "a = a.loc[:, numeric_columns(a)]\n",
    "a.to_csv(f'{top}/data/biotype_ann_counts.txt', sep='\\t')\n",
    "\n",
    "# Convert to per read.\n",
    "rpm = readsPerMillion(rpg)\n",
    "\n",
    "# Per million reads: collapse by gene type.\n",
    "rpm.df = rpm.df.groupby('Gene type').sum()#\n",
    "#print('rpm.groupby sum->', rpm.df.sum(axis=0))\n",
    "rpm.df = rpm.df[numeric_columns(rpm.df)]\n",
    "rpm.reads_per_million_df = rpm.df  # Refresh the synonym.\n",
    "rpm.df.to_csv(f'{top_dir}/data/biotype_per_read.txt', sep='\\t')\n",
    "\n",
    "# FYI: to get a Series of Gene type -> array of {Columns = protein datasets. Index = RNAs}:\n",
    "# a = rpg.df.groupby('Gene type').apply(np.array)\n",
    "\n",
    "#xpp = xlsPerProtein(rpm, xl_rate_fname='/Users/dp/pma/percentCrosslinked.xlsx')\n",
    "\n",
    "# Redefine the metadata objects to load the raw counts.\n",
    "positive_metadata = sameRiver.metadata.positive_metadata_all\n",
    "positive_metadata.ann_counts_file = f'{top_dir}/data/biotype_ann_counts.txt'\n",
    "negative_metadata.ann_counts_file = f'{top_dir}/data/biotype_ann_counts.txt'\n",
    "\n",
    "# Create counts objects of data collapsed by biotype and save.\n",
    "negatives = sameRiver.negativeCounts.negativeCounts(negative_metadata, xl_rate_fname='/Users/dp/pma/percentCrosslinked.xlsx')\n",
    "positives = sameRiver.positiveCounts.positiveCounts(positive_metadata, xl_rate_fname='/Users/dp/pma/percentCrosslinked.xlsx')\n",
    "positives.save(write_object=True, write_txt=True)  # Note this over-writes the regular data files.\n",
    "\n",
    "# Do stats and write to tables/\n",
    "nb = sameRiver.statsForCountsNB.statsForCountsNB(\n",
    "    negatives=negatives, positives=positives, data_dir=f'{positive_metadata.top_dir}/data/')\n",
    "nb.calculate_pvalues(which='per_read', test_mode=False)\n",
    "nb.write_pvals_single_file(which='per_read', outfname=f'{top}/tables/pvalues_per_read_biotypes.xlsx')\n",
    "nb.calculate_pvalues(which='per_protein')\n",
    "nb.write_pvals_single_file(which='per_protein', outfname=f'{top}/tables/pvalues_per_protein_biotypes.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rpg = rawReadsPerGene(f'{top_dir}/data/biotype_ann_counts.txt', scheme_filename=f'{top_dir}/scheme.xlsx')\n",
    "rpm = readsPerMillion(rpg)\n",
    "\n",
    "xpp = xlsPerProtein(rpm, xl_rate_fname='/Users/dp/pma/percentCrosslinked.xlsx')\n",
    "\n",
    "aves = rpm.average_by_protein()\n",
    "aves['s'] = aves.sum(axis=1)\n",
    "aves = aves.sort_values(by='s').tail(8)\n",
    "del aves['s']\n",
    "df = aves\n",
    "\n",
    "#df = pandas.read_csv(f'{top_dir}/data/biotype_per_read.txt', sep='\\t', index_col=0)\n",
    "df = df[numeric_columns(df)]\n",
    "df.clip(lower=0.1, inplace=True)\n",
    "df = np.log10(df)\n",
    "#print(df)\n",
    "\n",
    "fig = plt.figure()\n",
    "g = sns.clustermap(df.T, yticklabels=True)#, vmax=1, row_cluster=False)\n",
    "g.ax_row_dendrogram.set_visible(False)\n",
    "g.ax_col_dendrogram.set_visible(False)\n",
    "#fig.set_figheight(20)\n",
    "plt.savefig(figname('reads_per_million_biotype_clustermap'))\n",
    "#plt.show()\n",
    "plt.clf(); plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "xpp = xlsPerProtein(rpm, xl_rate_fname='/Users/dp/pma/percentCrosslinked.xlsx')\n",
    "\n",
    "aves = rpm.average_by_protein()\n",
    "pos = ['hnRNPC', 'hnRNPD', 'Rbfox1', 'Rbfox2', 'FUBP1', 'KHDRBS2',\n",
    "      'A1CF', 'CELF1', 'SF3B1', 'PCBP1', 'RPL5', 'FBL']\n",
    "negative_aves = aves.loc[:, [x for x in negative_metadata.random_proteins if x in aves]]\n",
    "positive_aves = aves.loc[:, [x for x in pos if x in aves]]\n",
    "mean_negative_aves = aves.mean(axis=1)\n",
    "aves['Negatives'] = mean_negative_aves.to_list()\n",
    "a = aves.sort_values(by='Negatives').tail(4)['Negatives']\n",
    "\n",
    "top_types = list(a.index)\n",
    "\n",
    "aves = aves.loc[top_types, :]\n",
    "\n",
    "fig = plt.figure()\n",
    "for _type in top_types:\n",
    "    neg_vals = negative_aves.loc[_type, :]/1E6\n",
    "    pos_vals = positive_aves.loc[_type, :]/1E6\n",
    "    print(neg_vals)\n",
    "    bins = np.linspace(0, max(pos_vals)*1.2, 20)\n",
    "    #plt.hist([neg_vals, pos_vals], label=['neg', 'pos'])\n",
    "    plt.hist(neg_vals, bins, alpha=0.5, label='Negatives')\n",
    "    plt.hist(pos_vals, bins, alpha=0.5, label='RBPs')# color='r')\n",
    "    plt.title(_type)\n",
    "    plt.legend()\n",
    "    plt.show()\n",
    "    plt.clf(); plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "??filter\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy import stats\n",
    "\n",
    "def make_stat_df(rpm: readsPerGene) -> pandas.DataFrame:\n",
    "    \n",
    "    aves = rpm.average_by_protein()\n",
    "    negative_aves = aves.loc[:,[x for x in negative_metadata.random_proteins if x in aves]]\n",
    "\n",
    "    mean_negative_aves = aves.mean(axis=1)\n",
    "    aves['Negatives'] = mean_negative_aves.to_list()\n",
    "\n",
    "    s = {}\n",
    "    for prot in rpm.proteins(): \n",
    "        cols = rpm.columns_for_a_protein(prot)\n",
    "        neg_vals = negative_aves\n",
    "        for RNA in rpm.df.index:\n",
    "            s.setdefault(RNA, {})\n",
    "            pos = rpm.df.loc[RNA, cols]\n",
    "            neg = negative_aves.loc[RNA, :]\n",
    "\n",
    "            std_neg = np.std(neg)\n",
    "            z = (pos.mean() - neg.mean())/np.std(neg)\n",
    "            s[RNA][f'{prot} z'] = z\n",
    "\n",
    "            t = stats.ttest_ind(a=pos, b=neg)\n",
    "            #s[prot][RNA][f'{prot} pvalue'] = t.pvalue\n",
    "            s[RNA][f'{prot} -log(pvalue)'] = -np.log10(t.pvalue)\n",
    "            #print(prot, RNA, z, t)\n",
    "    zt = pandas.DataFrame.from_dict(s, orient='index')\n",
    "    return zt\n",
    "\n",
    "#aves = rpm.average_by_protein()\n",
    "\n",
    "xpp = xlsPerProtein(rpm, xl_rate_fname='/Users/dp/pma/percentCrosslinked.xlsx')\n",
    "\n",
    "zt_pm = make_stat_df(rpm)\n",
    "zt_pm.to_excel('temp_pm.xlsx')\n",
    "\n",
    "zt_pp = make_stat_df(xpp)\n",
    "zt_pp.to_excel('temp_pp.xlsx')\n",
    "\n",
    "top_types = list(a.index)\n",
    "\n",
    "fc = {}\n",
    "for prot in set(['Negatives']) | rpm.proteins():\n",
    "    fc[prot] = {x: aves[prot][x]/mean_negative_aves[x] for x in aves.index}\n",
    "fc = pandas.DataFrame.from_dict(fc)\n",
    "\n",
    "print(zt.head(2))\n",
    "\n",
    "#fc.clip(lower=0.001, inplace=True)\n",
    "#fc = np.log2(fc)\n",
    "#print(fc.columns)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Initialize figure objects.\n",
    "\n",
    "# Will only use these biotypes.\n",
    "used_types = ['protein_coding', 'rRNA', 'snoRNA', 'repeat', 'snRNA']\n",
    "\n",
    "# Define the list of proteins to be plotted.\n",
    "protein_list = [x for x in positive_metadata.positive_proteins if x not in negative_metadata.random_proteins]\n",
    "protein_list += negative_metadata.random_proteins\n",
    "protein_list = [x for x in protein_list if x in rpm.proteins()]\n",
    "protein_list = [x for x in protein_list if x not in (#'CCIN', 'EPB41L5', \n",
    "                                                     'DDX3X', 'RARS2',)]\n",
    "protein_order = ['CDK4', 'CHMP3', 'DCTN6', 'ETS2', 'IDE', 'ITPA', 'TPGS2', 'UBA2', 'CCIN', 'EPB41L5',\n",
    "                'SMAD3', 'SMAD4', 'BARD1', 'BRCA1', 'CRNKL1',\n",
    "                'RPL5', 'KHDRBS2', 'FUBP1', 'A1CF', 'hnRNPD', 'Rbfox1', 'Rbfox2',\n",
    "                'CELF1', 'PCBP1', 'SF3B1', 'hnRNPC', 'FBL']\n",
    "\n",
    "protein_list = protein_order[::-1]\n",
    "# For mapping biotype to color.\n",
    "to_color = ['black', 'red', 'indigo', 'cyan', 'violet']\n",
    "\n",
    "def plot_bubbles(rpm, zt, fig_fname, marker_params='per_read'):\n",
    "    \n",
    "    fig = plt.figure()\n",
    "    ax = plt.subplot()\n",
    "    ax.scatter([0], [0], c='w')\n",
    "\n",
    "    for y_pos, protein in enumerate(protein_list):\n",
    "\n",
    "        ax.axhline(y=y_pos, color='k', alpha=0.5, linewidth=0.5)\n",
    "\n",
    "        for RNA in filter(lambda r: r in used_types, rpm.df.index):\n",
    "            pval = zt.loc[RNA, f'{protein} -log(pvalue)']\n",
    "            zscore = zt.loc[RNA, f'{protein} z']\n",
    "            marker = 'o' if zscore > 0 else 'D'\n",
    "\n",
    "            # Plot a color label only once.\n",
    "            label = RNA if protein=='FBL' else None\n",
    "\n",
    "            markersize = 15*np.log10(abs(zscore)) if marker_params == 'per_read' else 15*np.log10(abs(zscore))\n",
    "            \n",
    "            ax.plot(\n",
    "                pval, [y_pos], marker, color=to_color[used_types.index(RNA)], markeredgewidth=0,\n",
    "                alpha=0.5, markersize=markersize,\n",
    "                label=label)\n",
    "\n",
    "        ax.axvline(x=2, linestyle=':', color='k', alpha=0.1, linewidth=1)\n",
    "        ax.set_yticks(range(len(protein_list)))\n",
    "        ax.set_yticklabels(protein_list)\n",
    "\n",
    "    plt.legend()\n",
    "    fig.set_figheight(10)\n",
    "    fig.savefig(fig_fname)\n",
    "    plt.show(); plt.clf(); plt.close()\n",
    "\n",
    "    ##################\n",
    "    # Plot legend for bubble size interpretation.\n",
    "    fig = plt.figure()\n",
    "    ax = plt.subplot()\n",
    "    ax.scatter([0],[0], c='w')\n",
    "    \n",
    "    zscores = [2, 5, 10, 100]\n",
    "    pvals = [1, 3, 5, 7]\n",
    "    \n",
    "    for y_pos, protein in enumerate(protein_list):\n",
    "#        for RNA in filter(lambda r: r in used_types, rpm.df.index):\n",
    "        RNA = 'protein_coding'\n",
    "        marker = random.choice(['o', 'D'])\n",
    "        for pval, zscore in zip(pvals, zscores):\n",
    "            markersize = 15*np.log10(abs(zscore)) if marker_params == 'per_read' else 15*np.log10(abs(zscore))\n",
    "            ax.plot(\n",
    "                pval, [y_pos], marker, color=to_color[used_types.index(RNA)], markeredgewidth=0,\n",
    "                alpha=0.5, markersize=markersize,\n",
    "                label=label)\n",
    "                \n",
    "    ax.axvline(x=2, linestyle=':', color='k', alpha=0.1, linewidth=1)\n",
    "    ax.set_yticks(range(len(protein_list)))\n",
    "    ax.set_yticklabels(protein_list)\n",
    "        \n",
    "    fig.set_figheight(10)\n",
    "    fig.savefig(figname('legend_for_biotype_stripplot_significance_and_z_score'))\n",
    "    plt.show(); plt.clf(); plt.close()    \n",
    "    # End plot legend.\n",
    "    ##################\n",
    "    \n",
    "plot_bubbles(rpm, zt_pm, figname('biotype_stripplot_significance_and_z_score_per_read'), marker_params='per_read')\n",
    "plot_bubbles(xpp, zt_pp, figname('biotype_stripplot_significance_and_z_score_per_protein'), marker_params='per_protein')\n",
    "\n",
    "protein_order = ['CDK4', 'CHMP3', 'DCTN6', 'ETS2', 'IDE', 'ITPA', 'TPGS2', 'UBA2', 'CCIN', 'EPB41L5',\n",
    "                'SMAD3', 'SMAD4', 'BARD1', 'BRCA1', 'CRNKL1',\n",
    "                'RPL5', 'KHDRBS2', 'FUBP1', 'A1CF', 'hnRNPD', 'Rbfox1', 'Rbfox2',\n",
    "                'CELF1', 'PCBP1', 'SF3B1', 'hnRNPC', 'FBL']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pprint import pprint\n",
    "\n",
    "scheme = sameRiver.scheme.scheme(Path(top, 'scheme.xlsx'))\n",
    "df = rpm.df\n",
    "print(len(df.index))\n",
    "\n",
    "#pprint(df.sum())\n",
    "cf_lists = [\n",
    "    ('KHDRBS2', 'KHDRBS2-R168C'),\n",
    "    ('A1CF', 'A1CF-E34K'),\n",
    "    #('RPL5', 'RPL5-E82K'),\n",
    "    #('FUBP1', 'FUBP1-R429C'),\n",
    "    #('PCBP1', 'PCBP1-L100Q')\n",
    "    #('CRNKL1', 'CRNKL1-S128F'),\n",
    "    #('RARS2', 'RARS2-R6C'),\n",
    "]\n",
    "\n",
    "import scipy.stats as stats\n",
    "\n",
    "changes = {}\n",
    "\n",
    "for (wt, mut) in cf_lists:\n",
    "    print(wt, mut)\n",
    "    \n",
    "    changes[wt] = {}\n",
    "    wt_cols = [x for x in df.columns if wt in x and mut not in x]\n",
    "    mut_cols = [x for x in df.columns if mut in x]\n",
    "\n",
    "    wt_batches = [x.split('_')[0] for x in wt_cols]\n",
    "    mut_batches = [x.split('_')[0] for x in mut_cols]\n",
    "    batch_lookup = {batch_n: batch_str for batch_n, batch_str in enumerate(wt_batches)}\n",
    "    \n",
    "    def batch_cols(batch_n):\n",
    "        wts = [x for x in wt_cols if x.split('_')[0]==batch_lookup[batch_n]]\n",
    "        muts = [x for x in mut_cols if x.split('_')[0]==batch_lookup[batch_n]]\n",
    "        return (wts, muts)\n",
    "    \n",
    "    for biotype in df.index:\n",
    "        cols_paired = [batch_cols(n) for n in range(len(wt_batches))]\n",
    "        vals = [(float(df.loc[biotype, wtcol]), float(df.loc[biotype, mutcol])) for (wtcol, mutcol) in cols_paired]\n",
    "        \n",
    "        if np.mean(vals) < 100:\n",
    "            continue  # Skip categories with less than 0.01% of all signal.\n",
    "            \n",
    "        ratios = [_mut/_wt for _wt,_mut in vals if _wt>0]\n",
    "        ttest = stats.ttest_1samp(ratios, popmean=1)\n",
    "\n",
    "        if len(ratios):\n",
    "            changes[wt][biotype] = {\n",
    "                'log2vsWT': np.log2(np.mean(ratios)),\n",
    "                '-log10(P value)': -np.log10(ttest.pvalue),\n",
    "                'WT mean RPM': np.mean(vals),\n",
    "                'Gene type': biotype}\n",
    "            \n",
    "    changes[wt] = pandas.DataFrame(changes[wt]).T\n",
    "\n",
    "# Select PCBP1s\n",
    "pcbp1_names = ['PCBP1', 'PCBP1:100P', 'PCBP1:100Q', 'PCBP1:dKH', 'Gene type']\n",
    "\n",
    "# Create a palette to be used for all figures.\n",
    "type_list = ['rRNA', #'ribozyme', #'7SK', \n",
    "             'processed_transcript', 'snoRNA', #'nonsense_mediated_decay', \n",
    "             'tRNA', 'scaRNA', \n",
    "             'lincRNA',\n",
    "             'snRNA', #'LTR',\n",
    "             'protein_coding', 'repeat', #'intergenic'\n",
    "            ]\n",
    "\n",
    "\n",
    "type_colors = sns.cubehelix_palette(n_colors=len(type_list), start=.5, rot=-2)\n",
    "protein_colors = sns.cubehelix_palette(n_colors=len(cf_lists), start=.5, rot=-2)\n",
    "\n",
    "def ploty(log2vsWT, pair, axes):\n",
    "    p1 = plt.scatter(\n",
    "        log2vsWT['log2vsWT'], log2vsWT['-log10(P value)'], #'.',\n",
    "        color=protein_colors[cf_lists.index(pair)], alpha=0.5, s=100,\n",
    "        label=pair[1], edgecolors=None, linewidths=0)\n",
    "    axes.append(p1)\n",
    "    #axes[-1].set_label(pair[1])\n",
    "\n",
    "    # Add text labels to outliers on the plot.\n",
    "    for line in range(0,log2vsWT.shape[0]):\n",
    "        if log2vsWT['-log10(P value)'][line] < 2:\n",
    "            continue  # Don't label P>0.01 dots.\n",
    "        if abs(log2vsWT['log2vsWT'][line]) < np.log2(1.1):\n",
    "            continue  # Don't label FC < 1.5\n",
    "\n",
    "        plt.text(log2vsWT['log2vsWT'][line]+0.01, log2vsWT['-log10(P value)'][line], \n",
    "             log2vsWT['Gene type'][line], horizontalalignment='left', \n",
    "             size='small', color='black')#, weight='semibold')\n",
    "\n",
    "    return p1\n",
    "\n",
    "axes = []\n",
    "for (wt, mut) in cf_lists:\n",
    "    print(wt, '---->')\n",
    "    p1 = ploty(changes[wt], (wt, mut), axes)\n",
    "\n",
    "plt.legend(axes, [pair[1] for pair in cf_lists])\n",
    "plt.show()\n",
    "plt.clf(); plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas\n",
    "import sameRiver\n",
    "import numpy as np\n",
    "import sameRiver.biotypeDf\n",
    "import sameRiver.biotypeUtils\n",
    "import sameRiver.intronExonSplit\n",
    "import sameRiver.biotypeArtist\n",
    "import sameRiver.biotypeAdder\n",
    "importlib.reload(sameRiver.biotypeDf)\n",
    "importlib.reload(sameRiver.biotypeArtist)\n",
    "importlib.reload(sameRiver.biotypeUtils)\n",
    "#importlib.reload(sameRiver.countsO)\n",
    "importlib.reload(sameRiver.stacked_bargraph)\n",
    "\n",
    "# Create a palette to be used for all figures.\n",
    "type_list = [#'rRNA', #'ribozyme', #'7SK', \n",
    "             #'processed_transcript', \n",
    "             'snoRNA', #'nonsense_mediated_decay', \n",
    "             'tRNA', 'scaRNA', \n",
    "             'lincRNA',\n",
    "             'snRNA', #'LTR',\n",
    "             'protein_coding', \n",
    "    #'SINE/Alu', 'LINE/L1',\n",
    "    'Repetitive element',\n",
    "            #'repeat', #'intergenic'\n",
    "            ]\n",
    "type_colors = sns.cubehelix_palette(n_colors=len(type_list), start=.5, rot=-2)\n",
    "\n",
    "# Lookup table of number (position in type list) -> color.\n",
    "group_to_color = {_t:type_colors[m] for m, _t in enumerate(type_list)}\n",
    "\n",
    "##################################################################\n",
    "# Bargraphs of target RNA numbers.\n",
    "##################################################################\n",
    "\n",
    "# Load tools and data.\n",
    "top = '/Users/dp/pma/dataAndScripts/clip/old_mapping_meta/'\n",
    "#with open(f'{top}/data/stats.dill', 'rb') as f:\n",
    "#    nb = dill.load(f)\n",
    "\n",
    "rpg = readsPerGene(filename=f'{top}/ann_counts.txt', scheme_filename=f\"{top}/scheme.xlsx\")\n",
    "\n",
    "raw_read_df = sameRiver.combineReplicates.combine_protein_replicates(rpg.df, nb, scheme)\n",
    "#print(df.loc['>0_11_Crp#LINE/CR1::exon'])\n",
    "\n",
    "    \n",
    "def raw_cutoff(x):\n",
    "    if x < 5:\n",
    "        return 0\n",
    "    return x\n",
    "\n",
    "for col in rpg.numeric_columns():\n",
    "    rpg.df[col] = [raw_cutoff(v) for v in rpg.df[col]]\n",
    "\n",
    "def rep(_str):\n",
    "    if '/' in _str:\n",
    "        return 'Repetitive element'\n",
    "    return _str\n",
    "\n",
    "rpg.df['Gene type'] = [rep(_) for _ in rpg.df['Gene type']]\n",
    "rpm = readsPerMillion(rpg)\n",
    "xpp = xlsPerProtein(rpm, xl_rate_fname='/Users/dp/pma/percentCrosslinked.xlsx')\n",
    "\n",
    "scheme = sameRiver.scheme.scheme('./scheme.xlsx')\n",
    "\n",
    "#ba = sameRiver.biotypeAdder.biotypeAdder.load()\n",
    "\n",
    "# Per read or per protein\n",
    "which = 'per_protein'\n",
    "\n",
    "# Get the read counts table we'll be using.\n",
    "if which == 'per_read':\n",
    "    df = rpm.df#nb.positives.counts_per_million_df.copy()\n",
    "    pvals = pandas.read_excel(f\"{top}/tables/pvals_per_read.xlsx\", index_col=0)\n",
    "    #pandas.read_csv(f\"{top}/tables/pvalues_per_read.csv\", sep='\\t', index_col=0)\n",
    "elif which == 'per_protein':\n",
    "    df = xpp.df#nb.positives.counts_per_protein_df.copy()\n",
    "    pvals = pandas.read_excel(f\"{top}/tables/pvals_per_protein.xlsx\", index_col=0)#pandas.read_csv(f\"{top}/tables/pvalues_per_protein.csv\", sep='\\t', index_col=0)\n",
    "\n",
    "print(f\"Applying a raw read cutoff by setting pvalues to 1.0 if not enough raw reads.\")\n",
    "for col in pvals:\n",
    "    pvals[col] = [(p if raw_read_df.loc[gene, col]>9 else 1.) for gene, p in zip(pvals.index, pvals[col])]\n",
    "\n",
    "# Only biotypes in type list.\n",
    "df = df.loc[[(x in type_list) for x in df['Gene type']], :]\n",
    "\n",
    "# Combine protein replicates:\n",
    "import sameRiver.combineReplicates\n",
    "df = sameRiver.combineReplicates.combine_protein_replicates(df, nb, scheme)\n",
    "\n",
    "# Convert the read table of reads by RNA into three dataframes, all by biotype.\n",
    "[reads_in_targs_df, n_targets_df, all_reads_df] = sameRiver.biotypeArtist.dataframe_of_target_biotypes(\n",
    "    df, pvals, scheme, which=which, p_val_to_define_targets=1E-4)\n",
    "\n",
    "print('----=----'* 7)\n",
    "#print(f\"all_reads_df={all_reads_df.head()}\")\n",
    "all_reads_df = sameRiver.biotypeUtils.biotype_filter(all_reads_df)\n",
    "mrna = all_reads_df[all_reads_df['Gene type'] == 'protein_coding']\n",
    "to_prot = dict(zip(mrna['Protein'], mrna['% reads (all RNAs)']))\n",
    "by_mrna = sorted(to_prot.keys(), key=lambda x: to_prot[x])\n",
    "#by_mrna = filter_proteins(by_mrna)\n",
    "#print(f'pizza: after returning: n_targets_df={n_targets_df}')\n",
    "\n",
    "x = n_targets_df.groupby(by=['Protein']).sum()\n",
    "#print(f'x={x.head(5)}')\n",
    "by_target_number = x.sort_values(by=['# RNAs'], ascending=False, inplace=False).index\n",
    "\n",
    "to_remove = [ 'PCBP1 100P', 'PCBP1 100Q', 'PCBP1 dKH', 'HCT116']\n",
    "pcbp1s = ['PCBP1', 'PCBP1:100P', 'PCBP1:100Q', 'PCBP1:dKH']\n",
    "#by_target_number = pcbp1s\n",
    "#by_target_number = [x for x in by_target_number if (re.sub(':', ' ', x) not in to_remove)]\n",
    "#by_target_number = sorted(to_num.keys(), key=lambda x: to_num[x])\n",
    "randos_rbps = [\n",
    "    'FBL', 'Rbfox1', 'Rbfox2', 'hnRNPC', \n",
    "    'hnRNPD', 'CELF1', 'SF3B1', 'PCBP1',\n",
    "    'UBA2', 'ETS2', 'CAPNS6', 'TPGS2', \n",
    "    'EPB41L5', 'CHMP3', 'CDK4', 'IDE', \n",
    "    'DCTN6', 'CCIN', 'ITPA', \n",
    "    'PCBP1 100P', 'PCBP1 100Q', 'PCBP1 dKH']\n",
    "randos = [\n",
    "    'UBA2', 'ETS2', 'CAPNS6', 'TPGS2', \n",
    "    'EPB41L5', 'CHMP3', 'CDK4', 'IDE', \n",
    "    'DCTN6', #'CCIN', \n",
    "    'ITPA', \n",
    "]\n",
    "rbps = [\n",
    "    'FBL', 'hnRNPC', 'SF3B1', 'PCBP1', \n",
    "    #'PCBP1:100P', 'PCBP1:100Q','PCBP1:dKH',\n",
    "    'CELF1', 'Rbfox1', 'Rbfox2', 'hnRNPD',\n",
    "    'A1CF', 'KHDRBS2', 'FUBP1', 'RPL5',\n",
    "    ]\n",
    "chosen_columns = randos + rbps#rbps\n",
    "print(f\"pizza: by_target_number={by_target_number}\")\n",
    "print(f\"pizza: chosen_columns = {chosen_columns}\")\n",
    "print(f'pizza: by_target_number={by_target_number}')\n",
    "col_order =  [x for x in by_target_number if (x in chosen_columns)]\n",
    "print(f\"pizza: col_order={col_order}\")\n",
    "#randos_rbps_to_plot = [x for x in by_mrna if x in randos_rbps]\n",
    "#print(randos_rbps_to_plot)\n",
    "            \n",
    "def ploty(protein_order, figprefix='biotypes_bargraph_'):\n",
    "    sameRiver.stacked_bargraph.stacked_bargraph(\n",
    "        df=all_reads_df, value_label='% reads (all RNAs)', sample_label='Protein', \n",
    "        hue_label='Gene type', title='% reads (all RNAs)',\n",
    "        sample_order=protein_order,#counts.proteins(), \n",
    "        group_to_color=group_to_color,\n",
    "        fname=figname('biotypes_' + figprefix + 'all_rnas'))\n",
    "\n",
    "    print(f'pizza: right before plotting: n_targets_df={n_targets_df}')\n",
    "    sameRiver.stacked_bargraph.stacked_bargraph(\n",
    "        df=n_targets_df, value_label='# RNAs', sample_label='Protein', \n",
    "        hue_label='Gene type', title='# target RNAs',\n",
    "        sample_order=protein_order,#counts.proteins(), \n",
    "        group_to_color=group_to_color,\n",
    "        fname=figname('biotypes_' + figprefix + 'number_of_targets'))\n",
    "\n",
    "    sameRiver.stacked_bargraph.stacked_bargraph(\n",
    "        df=reads_in_targs_df, value_label='% reads (target RNAs)', sample_label='Protein', \n",
    "        hue_label='Gene type', title='% reads (in target RNAs)',\n",
    "        sample_order=protein_order,#counts.proteins(), \n",
    "        group_to_color=group_to_color,\n",
    "        fname=figname('biotypes_' + figprefix + 'target_rnas'))\n",
    "\n",
    "#ploty(randos_rbps_to_plot, 'randos_')\n",
    "\n",
    "\n",
    "#pcbp1s = [x for x in by_mrna if x in ['PCBP1', 'PCBP1 100P', 'PCBP1 100Q', 'PCBP1 dKH']]\n",
    "\n",
    "#ploty(pcbp1s, 'pcbp1_')\n",
    "\n",
    "ploty(col_order, figprefix=which)\n",
    "#print(pvals['CDK4'])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "hnRNP C bound mostly mRNA, while FBL was almost entirely crosslinked to snoRNA and rRNA (Figure 5E). \n",
    "\"\"\"\n",
    "\n",
    "\n",
    "import sameRiver\n",
    "import sameRiver.biotypeUtils\n",
    "import sameRiver.intronExonSplit\n",
    "importlib.reload(sameRiver.biotypeUtils)\n",
    "#sameRiver.biotypeUtils.make_enst_gene_name_biotype_map_file(\n",
    "#    gtf_filename='/opt/genome/Homo_sapiens.GRCh38.83/exon_CDS_ensembl_havana_only_Homo_sapiens.GRCh38.83_plus_snos_and_scar_from_ensembl_release91.gtf',\n",
    "#                         )\n",
    "\n",
    "#df = pandas.read_csv(\n",
    "#    '/Users/dfporter/pma/dataAndScripts/clip/miseq/meta/ann_counts.txt', sep='\\t',\n",
    "#    index_col=0)\n",
    "kk = '''\n",
    "_scheme = sameRiver.scheme.scheme('/Users/dfporter/pma/dataAndScripts/clip/miseq/meta/scheme.xls')\n",
    "\n",
    "counts = sameRiver.countsO.countsO(filename='/Users/dfporter/pma/dataAndScripts/clip/miseq/meta/ann_counts.txt',\n",
    "                            scheme=_scheme,\n",
    "                            index_col=0,\n",
    "                            indexing_col_name='gene_name')\n",
    "'''\n",
    "\n",
    "#bio = sameRiver.biotypeUtils.biotyper(counts)\n",
    "#bio.counts_df = counts.counts_df.copy()  # Just for safety\n",
    "\n",
    "#df = df[[x for x in df.columns if not re.search(':', x)]].copy()\n",
    "#print(df.columns)\n",
    "_types_df = bio.fraction_by_biotype()\n",
    "#print(_types_df.columns)\n",
    "_types_df = _types_df[[x not in [\n",
    "    #'PCBP1', \n",
    "    '100P', '100Q',\n",
    "    #'SF3B1',\n",
    "    'SF3B1:K700E',\n",
    "    'Unknown1', 'Unknown2', \n",
    "    #'No vector',\n",
    "#    'FBL', 'hnRNPC'\n",
    "    'HA-only',\n",
    "    'CSRP2',\n",
    "    'CAPNS6',\n",
    "    'hnRNPC:50k', 'hnRNPC:500cag', \n",
    "    'hnRNPC:4500agt', 'hnRNPC:500aac', 'hnRNPC:4500cag', \n",
    "    'hnRNPC:500agt', 'hnRNPC:0', 'hnRNPC:4500aac',\n",
    "    'hnRNPC:500tca', 'hnRNPC:4500tca',\n",
    "    ] for x in _types_df.Protein]]\n",
    "\n",
    "protein_order = counts.proteins()\n",
    "mrna = _types_df[_types_df['Gene type'] == 'protein_coding']\n",
    "to_prot = dict(zip(mrna['Protein'], mrna['Reads (% total)']))\n",
    "by_mrna = sorted(to_prot.keys(), key=lambda x: to_prot[x])\n",
    "introns_perc = sameRiver.intronExonSplit.intronExonSplit().percent_introns(counts, by_mrna)\n",
    "print(by_mrna)\n",
    "print(introns_perc)\n",
    "\n",
    "stacked_bargraph(df=_types_df, xlabel='Reads (% total)', ylabel='Protein', group_label='Gene type', \n",
    "                 protein_order=by_mrna)\n",
    "fig = plt.figure()\n",
    "sns.barplot(data=introns_perc, y='% Intronic', x='Protein', color='k')\n",
    "plt.xticks(rotation=90)\n",
    "fig.savefig('percent_intronic.pdf')\n",
    "plt.show()\n",
    "\n",
    "introns_perc['pre-mRNA score'] = [a*b for (a,b) in zip(\n",
    "    introns_perc['% Intronic'], introns_perc['Total (log10)'])]\n",
    "fig = plt.figure()\n",
    "sns.barplot(data=introns_perc, y='Total (log10)', x='Protein', color='k')\n",
    "plt.xticks(rotation=90)\n",
    "plt.ylabel('Reads in mRNA (log10)')\n",
    "fig.savefig('reads_in_mrna.pdf')\n",
    "\n",
    "fig = plt.figure()\n",
    "sns.barplot(data=introns_perc, y='pre-mRNA score', x='Protein', color='k')\n",
    "plt.xticks(rotation=90)\n",
    "plt.ylabel('pre-mRNA score')\n",
    "fig.savefig('reads_in_mrna.pdf')\n",
    "\n",
    "fig = plt.figure()\n",
    "sns.barplot(data=introns_perc, y='% Intronic (- Back.)', x='Protein', color='k')\n",
    "plt.xticks(rotation=90)\n",
    "fig.savefig('percent_intronic.pdf')\n",
    "plt.show()\n",
    "\n",
    "introns_perc['pre-mRNA score (- Back.)'] = [a*b for (a,b) in zip(\n",
    "    introns_perc['% Intronic (- Back.)'], introns_perc['Total (log10) (- Back.)'])]\n",
    "fig = plt.figure()\n",
    "sns.barplot(data=introns_perc, y='Total (log10) (- Back.)', x='Protein', color='k')\n",
    "plt.xticks(rotation=90)\n",
    "fig.savefig('reads_in_mrna.pdf')\n",
    "\n",
    "fig = plt.figure()\n",
    "sns.barplot(data=introns_perc, y='pre-mRNA score (- Back.)', x='Protein', color='k')\n",
    "plt.xticks(rotation=90)\n",
    "plt.ylabel('pre-mRNA score (- Back.)')\n",
    "fig.savefig('reads_in_mrna.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#make_gene2refseq2symbol_human()\n",
    "#make_refseq_to_biotype()\n",
    "\n",
    "def add_gene_name(df, refseq_name='/opt/genome/refseq/biotype_refseq_geneid_symbol.txt'):\n",
    "    return add_gene_name_and_type(df, refseq_name=refseq_name, add_type=False)\n",
    "\n",
    "\n",
    "#df = pandas.read_excel('/Users/dfporter/pma/miseq/Runs/170830/tables/counts.xls')\n",
    "df = pandas.read_csv(\n",
    "    '/Users/dfporter/pma/dataAndScripts/clip/miseq/meta/ann_counts.txt', sep='\\t',\n",
    "    index_col=0)\n",
    "\n",
    "#make_biotype_map_file()\n",
    "#add_biotypes_column(df)\n",
    "#add_gene_name_and_type_from_refseq(df)\n",
    "df.sort_index(by=df.columns[0], ascending=False, inplace=True)\n",
    "#print(df['Gene type'].value_counts())\n",
    "#print(df.columns)\n",
    "#df.to_excel('/Users/dfporter/pma/miseq/Proteins/counts.xls')\n",
    "print(df.head(1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def take_average_of_replicate_set(df, gene_name=None, replicate_columns=None):\n",
    "    #print(df.columns)\n",
    "    \n",
    "    if type(gene_name)==type(''):\n",
    "        replicate_columns = []\n",
    "        for col in find_hits_columns(df):\n",
    "            if re.search(gene_name, col):#.split('_')[0]):\n",
    "                replicate_columns.append(col)\n",
    "    \n",
    "    #print(replicate_columns)\n",
    "    if type(replicate_columns) == type([]):\n",
    "        sub = df[replicate_columns].copy()\n",
    "        _sum = sub.sum(axis=1)\n",
    "        _sum = [x/max(len(replicate_columns), 1) for x in _sum]\n",
    "        #print(sub.head(3))\n",
    "        return _sum\n",
    "    print(\"Some error in taking average of replicates, returning a column of zeros.\")\n",
    "    return [0] * len(df.index)\n",
    "\n",
    "def take_average_of_replicates_for_multiple_genes(df, genes='All'):\n",
    "    _results = {}\n",
    "    rnas = df.index.tolist()\n",
    "    \n",
    "    cols = set([x.split('_')[2] for x in find_hits_columns(df) if len(x.split('_')) > 2])\n",
    "    print(cols)\n",
    "    \n",
    "    for gene in cols:\n",
    "        averages = take_average_of_replicate_set(df, gene_name=gene)\n",
    "        _results[gene +'_average'] = dict(zip(rnas, averages))\n",
    "        \n",
    "    print(_results.keys())\n",
    "    averaged_df = pandas.DataFrame.from_dict(_results)\n",
    "    averaged_df['Gene'] = rnas\n",
    "\n",
    "    if 'Gene type' in df.columns:\n",
    "        averaged_df['Gene type'] = df['Gene type'].tolist()\n",
    "    \n",
    "    return averaged_df\n",
    "\n",
    "ave_df = take_average_of_replicates_for_multiple_genes(df)\n",
    "print(ave_df.head(2))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sameRiver\n",
    "reload(sameRiver.stacked_bargraph)\n",
    "#from sameRiver.stacked_bargraph import *\n",
    "def purge_no_feature(df):\n",
    "    def is_ok(x):\n",
    "        if x == '_no_feature':\n",
    "            return False\n",
    "        if x == '_ambiguous':\n",
    "            return False\n",
    "        return True\n",
    "    df['OK'] = [is_ok(n) for n in df['Gene'].tolist()]\n",
    "    print(\"Purging rows that are not clearly assigned to a gene: {0}\".format(df['OK'].value_counts()))\n",
    "    df = df[df['OK']].copy()\n",
    "    del df['OK']\n",
    "    return df\n",
    "\n",
    "def purge_columns_with_low_counts(df, min_count=100):\n",
    "    hits_columns = find_hits_columns(df)\n",
    "    to_del = set()\n",
    "    for col in hits_columns:\n",
    "        col_sum = df[col].sum()\n",
    "        if col_sum <= min_count:\n",
    "            to_del.add(col)\n",
    "    print(\"Purging columns with less than {0} counts total:\\n{1} out of {2} removed.\".format(\n",
    "        min_count, len(to_del), len(df.columns)))\n",
    "    for col in to_del:\n",
    "        del df[col]\n",
    "    return df\n",
    "\n",
    "def total_reads_per_protein(df):\n",
    "    hits_columns = find_hits_columns(df)\n",
    "    totals = collections.defaultdict(int)\n",
    "    for col in hits_columns:\n",
    "        gene = col.split('_')[0]\n",
    "        totals[gene] += df[col].sum()\n",
    "    return totals\n",
    "\n",
    "def total_reads_in_dataframe(df):\n",
    "    hits_columns = find_hits_columns(df, verbose=False)\n",
    "    total = 0\n",
    "    for col in hits_columns:\n",
    "        total += df[col].sum()\n",
    "    return total\n",
    "\n",
    "def _genes(df):\n",
    "    hits_columns = find_hits_columns(df)\n",
    "    return set([x.split('_')[0] for x in hits_columns])\n",
    "\n",
    "def _cols_of_gene(df, gene):\n",
    "    cols_to_keep = set()\n",
    "    hits_columns = find_hits_columns(df, verbose=False)\n",
    "    for col in hits_columns:\n",
    "        if col.split('_')[0] == gene:\n",
    "            cols_to_keep.add(col)\n",
    "    return list(cols_to_keep)\n",
    "\n",
    "def biotypes_with_enough_reads(df, cutoff=1000):\n",
    "    biotypes_to_keep = set()\n",
    "    hits_columns = find_hits_columns(df)\n",
    "    for biotype in set(df['Gene type'].tolist()):\n",
    "        all_for_type = 0\n",
    "        sub = df[df['Gene type']==biotype]\n",
    "        for col in hits_columns:\n",
    "            all_for_type += sub[col].sum()\n",
    "        if all_for_type >= cutoff:\n",
    "            biotypes_to_keep.add(biotype)\n",
    "    print(\"Biotypes with at least {0} reads across all datasets (Ignoring 'Unknown'): {1}.\".format(\n",
    "        cutoff, len(biotypes_to_keep - set(['Unknown']))))\n",
    "    return biotypes_to_keep - set(['Unknown'])\n",
    "\n",
    "def fraction_by_biotype(df):\n",
    "    for_df = []\n",
    "    usable_types = biotypes_with_enough_reads(df)\n",
    "    sub = df.copy()\n",
    "    sub['keep'] = [(x in usable_types) for x in df['Gene type']]\n",
    "    sub = sub[sub['keep']]\n",
    "    del sub['keep']\n",
    "    totals = total_reads_per_protein(sub)\n",
    "    for biotype in biotypes_with_enough_reads(df):\n",
    "        print(biotype)\n",
    "        sub = df[df['Gene type']==biotype]\n",
    "        for gene in _genes(df):\n",
    "            this_gene = sub[_cols_of_gene(sub, gene)]\n",
    "            reads_in_biotype = total_reads_in_dataframe(this_gene)\n",
    "            for_df.append({'Gene type': biotype, 'Protein': gene,\n",
    "                           'Reads': reads_in_biotype,\n",
    "                        'Reads (% total)': 100 * reads_in_biotype/totals[gene]})\n",
    "    return pandas.DataFrame(for_df)\n",
    "\n",
    "def classify_a_gene_name(x, _type):\n",
    "    if re.search('SNORD', x):\n",
    "        return 'C/D box snoRNA'\n",
    "    if re.search('SNORA', x):\n",
    "        return 'H/ACA box snoRNA'\n",
    "    if re.search('SCARNA', x):\n",
    "        return 'scaRNA'\n",
    "    if re.search('RNU', x):\n",
    "        return 'snRNA'\n",
    "\n",
    "    if _type == 'snRNA':\n",
    "        return 'snRNA'\n",
    "    if _type == 'protein_coding':\n",
    "        return 'protein_coding'\n",
    "    if _type == 'lincRNA':\n",
    "        return 'lincRNA'\n",
    "    return 'Unknown'\n",
    "\n",
    "def classify_gene_names(df):\n",
    "    #name, _type = zip(df['Gene name'].tolist(), df['Gene type'].tolist())\n",
    "    df['Class'] = [classify_a_gene_name(gene_name, gene_type) for (gene_name, gene_type) in \\\n",
    "                  zip(df.index, df['Gene type'].tolist())]\n",
    "    return df\n",
    "#df = pandas.read_excel('/Users/dfporter/pma/miseq/Proteins/counts.xls')\n",
    "#                    sep='\\t')#, index_col=False)\n",
    "#print(df)\n",
    "\n",
    "df = ave_df.copy()\n",
    "\n",
    "df = classify_gene_names(df)\n",
    "#print(df)\n",
    "\n",
    "\n",
    "for_pie = df[df['Class']!='Unknown']\n",
    "for_pie = for_pie[for_pie['Class']!='protein_coding']\n",
    "print\n",
    "#sums = for_pie.hnRNPC_171104_CGATTA_AAC.groupby(for_pie.Class).sum()\n",
    "\n",
    "#sums = for_pie.FBL.groupby(for_pie.Class).sum()\n",
    "add_gene_name_and_type_from_refseq(df)\n",
    "add_gene_name_and_type_from_refseq(for_pie)\n",
    "#print(df.columns)\n",
    "#print('====')\n",
    "nc_df = [collections.defaultdict(dict)]\n",
    "#hm = '''\n",
    "for gene in ['hnRNPC', 'FBL', 'PCBP1', '100Q', '100P', 'UBA2', 'TPGS2', 'ITPA', 'CDK4', 'NoVector',\n",
    "            'PCBP1:WT', 'PCBP1:100Q', 'PCBP1:100P']:\n",
    "    gene += '_average'\n",
    "    if gene not in for_pie:\n",
    "        continue\n",
    "    #print(gene, '->')\n",
    "    sums = for_pie[gene].groupby(for_pie.Class).sum()\n",
    "    sums = sums.to_dict()\n",
    "#    sums['Protein'] = gene\n",
    "#    sums['Gene'] = gene\n",
    "    for _type, _reads in sums.items():\n",
    "        row = {}\n",
    "        row['Gene type'] = _type\n",
    "        row['Reads'] = _reads\n",
    "        row['Protein'] = gene\n",
    "        row['Reads (% total)'] = _reads/max(1, for_pie[gene].sum())\n",
    "        nc_df.append(row)\n",
    "    #print(sums)\n",
    "    #nc_df['']\n",
    "    #plt.axis('equal')\n",
    "    #plt.pie(sums, labels=sums.index)\n",
    "    #plt.title(gene.split('_')[0])\n",
    "    #plt.gcf().savefig('figs/ncRNA_type_pie_{0}.svg'.format(gene.split('_')[0]))\n",
    "    #plt.show()\n",
    "    #plt.clf()\n",
    "#'''\n",
    "nc_df = pandas.DataFrame(nc_df)\n",
    "#find_hits_columns(df)\n",
    "nc_df = purge_columns_with_low_counts(nc_df)\n",
    "nc_df = nc_df.dropna(axis=0, how='any')\n",
    "#df = purge_no_feature(nc_df)\n",
    "print(nc_df)\n",
    "#types_df = fraction_by_biotype(df)\n",
    "protein_order = ['hnRNPC', 'FBL', 'SF3B1', 'SF3B1:K700E', \n",
    "                 'PCBP1', 'PCBP1:100Q','PCBP1:100P',\n",
    "                'PCBP1:WT', '100Q', '100P',\n",
    "                'UBA2', 'TPGS2', 'ITPA', 'CDK4', 'NoVector']\n",
    "                \n",
    "nc_df['Protein'] = [x.split('_')[0] for x in nc_df.Protein]\n",
    "sameRiver.stacked_bargraph.stacked_bargraph(\n",
    "    df=nc_df, xlabel='Reads (% total)', ylabel='Protein', group_label='Gene type', \n",
    "                 #protein_order=protein_order\n",
    "                                           )\n",
    "#print(types_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#protein_order = ['SF3B1', 'SF3B1:K700E']\n",
    "protein_order = ['hnRNPC', 'FBL', 'SF3B1', 'SF3B1:K700E', 'PCBP1:WT', 'PCBP1:100Q','PCBP1:100P',\n",
    "                'PCBP1', '100P', '100Q', 'UBA2', 'TPGS2', 'ITPA', 'CDK4', 'NoVector']\n",
    "#protein_order = ['NoVector', 'UBA2', 'TPGS2', 'ITPA', 'CDK4']\n",
    "#protein_order = ['PCBP1', '100Q', '100P']\n",
    "#protein_order = ['hnRNPC', 'FBL']\n",
    "#df = pandas.read_excel('/Users/dfporter/pma/miseq/Proteins/counts.xls')\n",
    "\n",
    "#df = purge_columns_with_low_counts(df)\n",
    "#df = purge_no_feature(df)\n",
    "print(df)\n",
    "add_gene_name_and_type_from_refseq(df)\n",
    "_types_df = fraction_by_biotype(df)\n",
    "\n",
    "#types_df['keep'] = [x in ['FBL', 'hnRNPC'] for x in types_df.Protein]\n",
    "#types_df = types_df.loc[types_df['keep']]\n",
    "#del types_df['keep']\n",
    "print(_types_df)\n",
    "#protein_order = [#'hnRNPC', 'FBL', 'SF3B1', 'SF3B1:K700E', \n",
    "#                 'PCBP1', 'PCBP1:100Q','PCBP1:100P',\n",
    "#                'PCBP1', '100Q', '100P'\n",
    "#                ]\n",
    "\n",
    "stacked_bargraph(df=_types_df, xlabel='Reads (% total)', ylabel='Protein', group_label='Gene type', \n",
    "                 protein_order=protein_order)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#?sns.color_palette\n",
    "sns.palplot(sns.light_palette('black'))\n",
    "plt.show()\n",
    "plt.clf()\n",
    "#sns.choose_colorbrewer_palette(, as_cmap=True, data_type='s')\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
