{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys, dill, pickle, pandas, re, HTSeq, os, importlib, operator, functools, random, matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "\n",
    "from matplotlib import patches\n",
    "\n",
    "import sklearn\n",
    "from sklearn.neighbors.kde import KernelDensity\n",
    "\n",
    "top_dir = top = '/Users/dp/pma/dataAndScripts/clip/meta/'\n",
    "os.chdir(top_dir)\n",
    "\n",
    "sys.path.append('/Users/dp/pma/')\n",
    "import sameRiver\n",
    "from sameRiver import *\n",
    "\n",
    "import sameRiver.peakFinder\n",
    "import sameRiver.peakSeqFinder\n",
    "\n",
    "importlib.reload(sameRiver.peakFinder)\n",
    "importlib.reload(sameRiver.peakSeqFinder)\n",
    "\n",
    "figname = lambda _str: f'/Users/dp/pma/dataAndScripts/clip/figs/{_str}.pdf'\n",
    "\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams['ps.fonttype'] = 42\n",
    "matplotlib.rc('font',**{'family':'sans-serif','sans-serif':['Arial']})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sameRiver.peakArtist\n",
    "import sameRiver.rnaDataFileMaker\n",
    "\n",
    "maker = sameRiver.rnaDataFileMaker.rnaDataFileMaker()\n",
    "RNAs = maker.make_from_gtf_file(\n",
    "    gtf_filename='/Users/dp/pma/repeats_and_genome.gtf',  # New mapping method gtf.\n",
    "    #gtf_filename='/opt/genomes/test.gtf',\n",
    "    #gtf_filename='/opt/genomes/repeats_and_ensembl_release94_GRCh38/combined_tsl1andNA.gtf',  # Old mapping gtf.\n",
    "    output_data_filename=f\"{top}/data/rna.data\",\n",
    ")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "importlib.reload(sameRiver.peakArtist)\n",
    "pk = sameRiver.peakArtist.peakArtist(\n",
    "    scheme_filename=f'{top}/scheme.xlsx',\n",
    "    bedgraph_folder=f'{top}/beds/read_start/',\n",
    "    file_with_total_read_counts=f'{top}/data/total_read_numbers.txt',\n",
    "    rnas_obj=RNAs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sameRiver.artist\n",
    "from typing import List, Mapping, Union\n",
    "importlib.reload(sameRiver.artist)\n",
    "import glob\n",
    "from sameRiver.artist import draw_gene_body, draw_ivs, array_from_htseq, signal_for_protein_in_region, smooth_y_values\n",
    "\n",
    "def drawNlines(_x: np.array, y_arrays: List[np.array]=[], labels: list=[], title: str='') -> None:\n",
    "    linewidth=1#0.5\n",
    "    a = 1\n",
    "\n",
    "    pal = sns.cubehelix_palette(len(labels), start=0.14, rot=0.9)\n",
    "    for n, (label, y_array) in enumerate(zip(labels, y_arrays)):\n",
    "        plt.plot(\n",
    "            _x, y_array, label=label, c=pal[n], linewidth=linewidth, alpha=a)\n",
    "\n",
    "def plotN(\n",
    "    _x: np.array, y_arrays: List[np.array]=[], labels: list=[], \n",
    "    title: str='', rna='rna', smooth=False, yaxislabel=''):\n",
    "    \n",
    "    plt.clf()\n",
    "    fig = plt.figure()\n",
    "    \n",
    "    if smooth:\n",
    "        xys = [smooth_y_values(y) for y in y_arrays]\n",
    "        _x = xys[0][0]\n",
    "        y_arrays = [_[1] for _ in xys]\n",
    "\n",
    "    drawNlines(_x, y_arrays, labels=labels)\n",
    "\n",
    "    plt.legend()#.remove()\n",
    "    plt.title(title)\n",
    "    plt.ylabel(yaxislabel)\n",
    "    fig.set_figheight(4)  # 2 for publication.\n",
    "    fig.set_figwidth(6)  # 3 for publication.\n",
    "    fig.savefig(figname(f'signal_smooth_{title}'))\n",
    "    plt.show()\n",
    "    plt.clf()\n",
    "    plt.close()\n",
    "\n",
    "print(pk.gene_name_to_bedgraph_filenames('RPL5:E82K', glob.glob(pk.bedgraph_folder + '/*_+.wig')))\n",
    "rna = '>0_5S#rRNA'\n",
    "rna = '>0_rRNA_Hsa#rRNA'\n",
    "rna = '>0_U2#snRNA'\n",
    "rna = 'PURPL'\n",
    "q = pk.genomic_iv_of_gene(rna)\n",
    "print(q)\n",
    "labels = ['RPL5', 'FUBP1', 'PCBP1', 'SF3B1']\n",
    "labels = ['SF3B1', 'FUBP1', 'KHDRBS2', 'KHDRBS2:R168C', 'A1CF',  'A1CF:E34K']\n",
    "xy_arrs = [pk.signal_for_protein_in_region(label, rna) for label in labels]\n",
    "print(xy_arrs[0])\n",
    "#xy_arrs = [(_x[10:], _y[10:]) for _x,_y in xy_arrs]\n",
    "#print(xy_arrs)\n",
    "x_arrs = [_[0] for _ in xy_arrs]\n",
    "y_arrs = [_[1] for _ in xy_arrs]\n",
    "print(x_arrs)\n",
    "plotN(x_arrs[0], y_arrs, labels=labels, title=rna, smooth=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import joypy\n",
    "\n",
    "xys = [smooth_y_values(y) for y in y_arrs]\n",
    "smoothed_ys = [_[1] for _ in xys]\n",
    "a = np.array(smoothed_ys)\n",
    "df = pandas.DataFrame(a.T, columns=labels)\n",
    "\n",
    "fig = plt.figure()\n",
    "j = joypy.joyplot(df, kind='values', #ylim='own', \n",
    "              linecolor=\"k\", fill=False, x_range=range(len(df.index)))\n",
    "plt.savefig(figname('ridgeline_plot_PURPL_per_read'))\n",
    "plt.show(); plt.clf(); plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def to_per_protein(_df):\n",
    "    for protein in _df.columns:\n",
    "        xl_rate = pk.scheme.percent_xl_of_protein_from_file(protein)\n",
    "        if xl_rate == 0:\n",
    "            xl_rate = pk.scheme.percent_xl_of_protein_from_file(protein.replace(':', ' '))\n",
    "        print(f\"{protein}: {xl_rate}\")\n",
    "        _df[protein] = 1E4 * xl_rate * _df[protein]\n",
    "    return _df\n",
    "\n",
    "xldf = df.copy()\n",
    "\n",
    "q = to_per_protein(xldf)\n",
    "\n",
    "fig = plt.figure()\n",
    "j = joypy.joyplot(q, kind='values', #ylim='own', \n",
    "              linecolor=\"k\", fill=False, x_range=range(len(df.index)))\n",
    "plt.savefig(figname('ridgeline_plot_PURPL_per_protein'))\n",
    "plt.show(); plt.clf(); plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sameRiver.artist\n",
    "importlib.reload(sameRiver.artist)\n",
    "from sameRiver.artist import draw_gene_body, draw_ivs, array_from_htseq, signal_for_protein_in_region, smooth_y_values\n",
    "\n",
    "##################\n",
    "# Definitions\n",
    "\n",
    "def draw3lines(_x, y_wt, y_100p, y_100q, title=''):\n",
    "    linewidth=1#0.5\n",
    "    a = 1\n",
    "\n",
    "    plt.plot(_x, y_wt, label='WT', c='#142C37', linewidth=linewidth, alpha=a)\n",
    "    plt.plot(_x, y_100p, label='L100P', c='#A5689A', linewidth=linewidth, alpha=a)\n",
    "    plt.plot(_x, y_100q, label='L100Q', c='#899FC9', linewidth=linewidth, alpha=a)\n",
    "    \n",
    "def plot3(_x, y_wt, y_100p, y_100q, title='', rna='rna', ylabel='', smooth=False):\n",
    "    plt.clf()\n",
    "    fig = plt.figure()\n",
    "\n",
    "    \n",
    "    if smooth:\n",
    "        (_x, sy_wt) = smooth_y_values(y_wt)\n",
    "        (_x, sy_100p) = smooth_y_values(y_100p)\n",
    "        (_x, sy_100q) = smooth_y_values(y_100q)\n",
    "        draw3lines(_x, sy_wt, sy_100p, sy_100q)\n",
    "    else:\n",
    "        draw3lines(_x, y_wt, y_100p, y_100q)\n",
    "\n",
    "    plt.legend().remove()\n",
    "    plt.title(title)\n",
    "    plt.ylabel(ylabel)\n",
    "    fig.set_figheight(2)  # 2 for publication.\n",
    "    fig.set_figwidth(3)  # 3 for publication.\n",
    "    fig.savefig(\n",
    "        '/Users/dp/pma/dataAndScripts/clip/figs/signal_smooth_{}_{}.pdf'.format(title, ylabel))\n",
    "    plt.show()\n",
    "    plt.clf()\n",
    "    plt.close()\n",
    "\n",
    "    \n",
    "def iv_str(_str, strand):\n",
    "    s = _str.split(':')\n",
    "    a, b = [int(x.rstrip('\\n')) for x in s[1].split('-')]\n",
    "    return HTSeq.GenomicInterval(s[0], a, b, strand)\n",
    "\n",
    "##################\n",
    "# Input selections\n",
    "\n",
    "\"\"\"chr11:65892136-65900526\n",
    "\n",
    "FOSL1\n",
    "chr11:65892136-65900526\n",
    "id = NM_001300844\n",
    "--------------\n",
    "Exon number: 3\n",
    "Amino acid coding number: 219\n",
    "chr11:65892136-65893296\n",
    "~\n",
    "U2:\n",
    "repeats:115971-116159\n",
    "\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "#rna = 'RNU4-2'\n",
    "#rna = 'GAPDH'\n",
    "#rna = 'FOSL1'\n",
    "rna = '>0_U1#snRNA'\n",
    "#rna = '>0_U2#snRNA'\n",
    "#rna = '>0_U3#snRNA'\n",
    "#rna = '>0_U4#snRNA'\n",
    "#rna = '>0_rRNA_Hsa#rRNA'\n",
    "#rna = '>0_5S#rRNA'\n",
    "#rna = 'CDKN1A'\n",
    "#rna = 'SOX9'\n",
    "#rna = 'PCBP2'\n",
    "#rna = 'SMAD3'\n",
    "#rna = 'SPRYD4'\n",
    "#rna = 'AGPAT1'\n",
    "#rna = 'GSK3A'\n",
    "#rna = 'CDKN1A'\n",
    "\n",
    "#rna = iv_str('chr11:65892136-65893296', '-')  # FOSL1 leftmost exon\n",
    "#rna = iv_str('repeats:115971-116159', '-') # U2\n",
    "#rna = iv_str('repeats:1217629-1217846', '-') # U3\n",
    "#rna = iv_str('repeats:10075-10220', '-')\n",
    "#rna = iv_str('chr19:42230186-42230867', '-')  # Last GSK3A exon\n",
    "\n",
    "##################\n",
    "# Obtaining arrays of values at selection\n",
    "\n",
    "# Per million reads\n",
    "#pk.signal_for_protein_in_region('RPL5', '>0_5S#rRNA')\n",
    "_x, y_wt  = pk.signal_for_protein_in_region('PCBP1', rna)\n",
    "_x, y_100p = pk.signal_for_protein_in_region('PCBP1:100P', rna)\n",
    "_x, y_100q = pk.signal_for_protein_in_region('PCBP1:100Q', rna)\n",
    "#y_wt, _x = signal_for_protein_in_region('PCBP1', ssr, rna, bedfname_to_total_reads)\n",
    "#y_100p, _x = signal_for_protein_in_region('PCBP1:100P', ssr, rna, bedfname_to_total_reads)\n",
    "#y_100q, _x = signal_for_protein_in_region('PCBP1:100Q', ssr, rna, bedfname_to_total_reads)\n",
    "#y_wt, _ = signal_for_protein_in_region('PCBP1', ssr05, rna, bedfname_to_total_reads)\n",
    "\n",
    "##################\n",
    "# Derivatives of selections\n",
    "scheme = sameRiver.scheme.scheme('./scheme.xlsx')\n",
    "# Per billion proteins\n",
    "y_wt_per_prot = 10 * scheme.percent_xl_of_protein_from_file('PCBP1') * y_wt\n",
    "y_100p_per_prot = 10 * scheme.percent_xl_of_protein_from_file('PCBP1 L100P') * y_100p\n",
    "y_100q_per_prot = 10 * scheme.percent_xl_of_protein_from_file('PCBP1 L100Q') * y_100q\n",
    "\n",
    "\n",
    "# Per cell\n",
    "df = pandas.read_excel('/Users/dp/pma/dataAndScripts/tables/lysateExpression.xlsx')\n",
    "\n",
    "to_expression = dict(df.groupby('Protein')['Value'].mean())\n",
    "\n",
    "y_100p_per_cell = 1E-2 * to_expression['PCBP1 100P'] * y_100p_per_prot\n",
    "y_100q_per_cell = 1E-2 * to_expression['PCBP1 100Q'] * y_100q_per_prot\n",
    "y_wt_per_cell = 1E-2 * to_expression['PCBP1'] * y_wt_per_prot\n",
    "\n",
    "##################\n",
    "# Plot.\n",
    "\n",
    "if type(rna) != type(''):\n",
    "    title = '{}_{}_{}'.format(rna.chrom, rna.start, rna.end)\n",
    "else:\n",
    "    title = rna\n",
    "\n",
    "smooth = True\n",
    "\n",
    "plot3(\n",
    "    _x, y_wt, y_100p, y_100q, \n",
    "    rna=rna, title=title, ylabel='Cross-links per million cross-links', smooth=smooth)\n",
    "\n",
    "plot3(\n",
    "    _x, y_wt_per_prot, y_100p_per_prot, y_100q_per_prot,\n",
    "    rna=rna, title=title, ylabel='Cross-links per billion proteins', smooth=smooth)\n",
    "\n",
    "plot3(\n",
    "    _x, y_wt_per_cell, y_100p_per_cell, y_100q_per_cell, \n",
    "    rna=rna, title=title, ylabel='Cross-links per cell (relative)', smooth=smooth)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "iv = HTSeq.GenomicInterval('a',1,20,'-')\n",
    "print(type(iv.strand))\n",
    "dict([('+', '-'), ('-', '+')])[str(iv.strand)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "importlib.reload(sameRiver.statsForCountsNB)\n",
    "\n",
    "#nb = sameRiver.statsForCountsNB.statsForCountsNB(negatives=negatives, positives=positives)\n",
    "#nb.calculate_pvalues()\n",
    "nb.write_pvals_single_file(\n",
    "    which='per_read', outfname='data/pvals_for_signal_at_{}_peaks_per_read.xlsx'.format(gene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.stats as stats\n",
    "import statsmodels\n",
    "import statsmodels.api as sm\n",
    "\n",
    "detected = np.array([1, 1, 2, 1, 2, 3, 5, 8]) * 100\n",
    "average = 10 * 100\n",
    "#detected = np.log(detected)\n",
    "#average = np.log(average)\n",
    "\n",
    "mean_negative = np.average(detected)\n",
    "max_negative = np.max(detected)\n",
    "vmr = (np.std(detected)**2)/mean_negative\n",
    "\n",
    "pois = stats.poisson(mean_negative)\n",
    "p = 1 - pois.cdf(average)\n",
    "\n",
    "print(detected, 'vs', average)\n",
    "print('mu=', mean_negative)\n",
    "print('std=', np.std(detected))\n",
    "print('VMR=', vmr)\n",
    "\n",
    "q = sm.NegativeBinomial(detected, np.array([1] * len(detected)), loglike_method='nb2')\n",
    "res = q.fit(disp=0)\n",
    "mu = res.predict()[0]\n",
    "size = 1. / res.params[1]\n",
    "prob = size / (size + mu)\n",
    "print('NB fit mu=', mu)\n",
    "\n",
    "dist_est = stats.nbinom(size, prob)\n",
    "\n",
    "nb_p = 1 - dist_est.cdf(average)\n",
    "print('Poisson p value=', p)\n",
    "print('NB pvalue=', nb_p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#genomic_fasta_fname = '/opt/genome/repeats_and_ensembl_release94_GRCh38/combined.fa'\n",
    "#genomic_fasta = dict( (s[1], s[0]) for s in HTSeq.FastaReader(genomic_fasta_fname, raw_iterator=True) )\n",
    "\n",
    "motif = 'GCATG'\n",
    "\n",
    "import sameRiver.motifTrackWriter\n",
    "importlib.reload(sameRiver.motifTrackWriter)\n",
    "from sameRiver.motifTrackWriter import motifTrackWriter\n",
    "        \n",
    "mtw = motifTrackWriter()\n",
    "mtw.write_motif_track(genomic_fasta, motif, motif2='TGCATG')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
