{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, re, matplotlib, pandas, collections, importlib, sys, pickle\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "\n",
    "import __init__\n",
    "sys.path = __init__.add_paths(sys.path)\n",
    "\n",
    "from importlib import reload\n",
    "import sameRiver\n",
    "from sameRiver import *\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def split_types(df, col):\n",
    "    def later_part(_str):\n",
    "        try:\n",
    "            return _str.split('::')[1]\n",
    "        except:\n",
    "            return ''\n",
    "    exons = [later_part(x) == 'exon' for x in df[col]]\n",
    "    introns = [later_part(x) == 'intron' for x in df[col]]\n",
    "    return df[exons].copy(), df[introns].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df = pandas.read_csv('/Users/dfporter/pma/dataAndScripts/clip/miseq/meta/ann_counts.txt', sep='\\t', index_col=0)\n",
    "top_dir = '/Users/dp/pma/dataAndScripts/clip/meta/'\n",
    "df = pandas.read_csv(f'{top_dir}/ann_counts.txt', sep='\\t', index_col=0)\n",
    "df['gene'] = df.index\n",
    "\n",
    "exons, introns = split_types(df, 'gene')\n",
    "results = []\n",
    "for col in df.columns:\n",
    "    if col in ['gene', 'gene_name', 'Gene type']:\n",
    "        continue\n",
    "    if re.search('nknown', col):\n",
    "        continue\n",
    "    if re.search('No vec', col):\n",
    "        continue\n",
    "        \n",
    "    skip = False\n",
    "    for pat in ['PCBP', '100', 'SF3B', 'nknown', 'No vec', 'hnRNPC:']:\n",
    "        if re.search(pat, col):\n",
    "            skip = True\n",
    "    if skip:\n",
    "        continue\n",
    "        \n",
    "    exonic = exons[col].sum()\n",
    "    intronic = introns[col].sum()\n",
    "    total = exonic + intronic\n",
    "    if total < 1E3:\n",
    "        continue\n",
    "    results.append({\n",
    "        '% Exonic': 100 * exonic/total,\n",
    "        '% Intronic': 100 *intronic/total,\n",
    "        'Total': exonic + intronic,\n",
    "        'Total (log10)': np.log10(exonic + intronic),\n",
    "        'Intronic / total': (intronic/total) / np.log10(exonic + intronic),\n",
    "        'Protein_rep': col,\n",
    "        'Protein': col.split('_')[1],\n",
    "    })\n",
    "\n",
    "df = pandas.DataFrame(results)\n",
    "df.sort_values(by='% Intronic', inplace=True)\n",
    "\n",
    "_kwargs = {'palette': sns.cubehelix_palette(2*len(set(df['Protein'])), start=0, rot=-0.2)}\n",
    "\n",
    "fig = plt.figure()\n",
    "sns.lmplot(data=df, x='Total (log10)', y='% Intronic')\n",
    "fig.set_figheight(10)\n",
    "plt.show()\n",
    "plt.clf()\n",
    "\n",
    "fig = plt.figure()\n",
    "\n",
    "sns.barplot(data=df, y='Protein', x='% Intronic', **_kwargs)\n",
    "fig.set_figheight(12)\n",
    "plt.show()\n",
    "plt.clf()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "excel_of_target_RNAs = '/Users/dp/pma/dataAndScripts/clip/miseq/meta/tables/pvals.xlsx'\n",
    "\n",
    "\n",
    "#excel_of_target_RNAs = '/Users/dfporter/pma/dataAndScripts/clip/miseq/meta/ann_counts.txt'\n",
    "\n",
    "xl = pandas.ExcelFile(excel_of_target_RNAs)\n",
    "\n",
    "results = []\n",
    "for sheet_name in xl.sheet_names:\n",
    "    print(sheet_name)\n",
    "\n",
    "    skip = False\n",
    "    for pat in ['PCBP', '100', 'SF3B', 'nknown', 'No vec', 'hnRNPC:', 'hnRNPC ']:\n",
    "        if re.search(pat, sheet_name):\n",
    "            skip = True\n",
    "    if skip:\n",
    "        continue\n",
    "\n",
    "    df = xl.parse(sheet_name)\n",
    "    print(df.iloc[0])\n",
    "    if len(df.index) == 0:\n",
    "        continue\n",
    "        \n",
    "    df['gene'] = df.index\n",
    "    df = df[df['P value']<0.01]\n",
    "    exons, introns = split_types(df, 'gene')\n",
    "    total = len(exons.index) + len(introns.index)\n",
    "    \n",
    "    if total < 1:\n",
    "        print(df)\n",
    "        continue\n",
    "    \n",
    "    n_rnas = len(set([x.split('::')[0] for x in df.gene]))\n",
    "    \n",
    "    results.append({\n",
    "        'Gene': sheet_name,\n",
    "        '% Exons': 100* len(exons.index)/total,\n",
    "        '% Introns': 100* len(introns.index)/total,\n",
    "        'Total': len(exons.index) + len(introns.index),\n",
    "        'Target RNAs': n_rnas,\n",
    "    })\n",
    "\n",
    "df = pandas.DataFrame(results)\n",
    "print(df)\n",
    "df.sort_values(by='% Introns', inplace=True)\n",
    "\n",
    "_kwargs = {'palette': sns.cubehelix_palette(2*len(set(df['Gene'])), start=0, rot=-0.2)}\n",
    "\n",
    "fig = plt.figure()\n",
    "sns.barplot(data=df, x='% Introns', y='Gene', **_kwargs)\n",
    "fig.savefig('/Users/dfporter/pma/dataAndScripts/clip/figs/Percent_introns_in_target_RNAs_bargraph.svg')\n",
    "plt.show()\n",
    "plt.clf()\n",
    "\n",
    "df.sort_values(by='Target RNAs', inplace=True)\n",
    "fig = plt.figure()\n",
    "sns.barplot(data=df, x='Target RNAs', y='Gene', **_kwargs)\n",
    "fig.savefig('/Users/dfporter/pma/dataAndScripts/clip/figs/Number_of_target_RNAs_bargraph.svg')\n",
    "plt.show()\n",
    "plt.clf()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
