{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# Options:\n",
    "# Get Refseq ID <-> is_C/D_box_snoRNA link\n",
    "# Go through sams/refseq/\n",
    "# Or, using only the consensus:\n",
    "# Load the deletion bedgraphs.\n",
    "# Get (Refseq ID or genomic interval) <-> is_C/D_box_snoRNA link\n",
    "# Count in regions using bedgraph.\n",
    "\n",
    "# Sequences, with a single ID from Refseq, are in:\n",
    "# /opt/indexes/refseq_index/refseq_rna_reformatted.fasta\n",
    "# The counts file should have the same Refseq IDs as the fasta.\n",
    "# These can be extracted first, since it would be a slow step, and saved.\n",
    "# The GTF can then be used to look up genomic coordinates, which can also be saved.\n",
    "\n",
    "# Finally, the bedgraphs are read and mapped to snoRNAs from their genomic coordinates,\n",
    "# and the sequences are used to partition. That data is sent here for plotting.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re, pandas, HTSeq, glob, collections, importlib\n",
    "import numpy as np\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as patches\n",
    "import seaborn as sns\n",
    "\n",
    "import sameRiver\n",
    "from sameRiver import snoUtils\n",
    "from sameRiver.signal_normalizer import signal_normalizer\n",
    "from sameRiver.bedgraphs import bedgraphs\n",
    "from sameRiver.area import *\n",
    "importlib.reload(sameRiver.area)\n",
    "\n",
    "\"\"\"Text below is copied from: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4053766/\n",
    "Insights into snoRNA biogenesis and processing from PAR-CLIP of snoRNA core proteins and small RNA sequencing\n",
    "\n",
    "Curated C/D and H/ACA box snoRNA annotation\n",
    "SnoRNA sequences were initially obtained from snoRNA­LBME­db. BLAT searches at the UCSC Genome Browser \n",
    "(http://genome.ucsc.edu/cgi­bin/hgBlat) were then conducted to map sequences to the human genome assembly hg19.\n",
    "Based on the data available at snoRNA­LBME­db and inspection of evolutionary conservation profiles at the \n",
    "UCSC Genome Browser, we next annotated C, C’, D’ and D box motifs, and the terminal closing stem. For some \n",
    "snoRNAs, terminal end complementarity could be extended beyond the initial sequence obtained from snoRNA LBME db. \n",
    "In these cases sequence coordinates were adjusted accordingly.\n",
    "\n",
    "Each individual snoRNA is represented by 4 lines, which correspond to the description of the snoRNA, \n",
    "the sequence, annotation of C (C), D (D), C'(c) and D'(d) boxes, and the location of the terminal stem region in\n",
    "parenthesis notation. The description line contains the snoRNA name, its coordinates (chromosome, start, stop and \n",
    "strand) in the hg19 version of the human genome assembly from the University of California Santa Cruz, and the \n",
    "modifications sites and target molecules of the guide regions located upstream of the D’ box (d) or D box (D).\n",
    "\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "            \n",
    "class snoRNA(complex_area):\n",
    "    \n",
    "    def __init__(self, name, seq, box, structure):\n",
    "        self.iv = None\n",
    "        self.sub_areas = {}\n",
    "        self.regions = {}\n",
    "        \n",
    "        self.name, self.seq, self.box = name, seq, box\n",
    "        self.structure = structure\n",
    "        self.c_box, self.c_prime_box, self.d_box, self.d_prime_box = [None] * 4\n",
    "        self.boxes = []\n",
    "        self.parse_seq()\n",
    "    \n",
    "    def check(self, re_result):\n",
    "        if re_result is None:\n",
    "            return False\n",
    "        if len(re_result.groups()) > 1:\n",
    "            print(\"??\", re_result)\n",
    "            return False\n",
    "        return True\n",
    "    \n",
    "    def parse_seq(self):\n",
    "        cp = re.search('(c+)', self.box)\n",
    "        if self.check(cp):\n",
    "            self.c_prime_box = (cp.start(), cp.end())\n",
    "        C = re.search('(C+)', self.box)\n",
    "        if self.check(C):\n",
    "            self.c_box = (C.start(), C.end())\n",
    "        dp = re.search('(d+)', self.box)\n",
    "        if self.check(dp):\n",
    "            self.d_prime_box = (dp.start(), dp.end())\n",
    "        D = re.search('(D+)', self.box)\n",
    "        if self.check(D):\n",
    "            self.d_box = (D.start(), D.end())\n",
    "        self.boxes = {'C': self.c_box, \n",
    "                      'Cp': self.c_prime_box, \n",
    "                      'D': self.d_box, \n",
    "                      'Dp': self.d_prime_box}\n",
    "        self.define_CD_box_snoRNA_regions()\n",
    "    \n",
    "    def name_from_split(self, sep='|'):\n",
    "        self.name = self.name.split(sep)[0]\n",
    "        self.name = self.name.lstrip('>')\n",
    "\n",
    "    def define_CD_box_snoRNA_regions(self):\n",
    "        all_regions = ['preC', 'C', 'loop1', 'Dp', 'loop2', 'Cp', 'loop3', 'D', 'postD']\n",
    "        lens = {}\n",
    "        regions = dict(zip(self.boxes.keys(), self.boxes.values())) # Copy.\n",
    "        if self.boxes['C'] is not None:\n",
    "            lens['preC'] = self.boxes['C'][0]\n",
    "            regions['preC'] = (0, self.boxes['C'][0])\n",
    "            lens['C'] = self.boxes['C'][1] - self.boxes['C'][0]\n",
    "            if self.boxes['Dp'] is not None:\n",
    "                lens['loop1'] = self.boxes['Dp'][0] - self.boxes['C'][1]\n",
    "                regions['loop1'] = (self.boxes['C'][1], self.boxes['Dp'][0])\n",
    "            elif self.boxes['Cp'] is not None:\n",
    "                lens['C to Cp'] = self.boxes['Cp'][0] - self.boxes['C'][1]\n",
    "                regions['C to Cp'] = (self.boxes['C'][1], self.boxes['Cp'][0])\n",
    "            elif self.boxes['D'] is not None:\n",
    "                lens['C to D'] = self.boxes['D'][0] - self.boxes['C'][1]\n",
    "                regions['C to D'] = (self.boxes['C'][1], self.boxes['D'][0])\n",
    "            else:\n",
    "                lens['C to end'] = len(self.seq) - self.boxes['C'][1]\n",
    "                regions['C to end'] = (self.boxes['C'][1], len(self.seq))\n",
    "        if self.boxes['Dp'] is not None:\n",
    "            lens['Dp'] = self.boxes['Dp'][1] - self.boxes['Dp'][0]\n",
    "            if self.boxes['Dp'] is not None:\n",
    "                lens['loop2'] = self.boxes['Cp'][0] - self.boxes['Dp'][1]\n",
    "                regions['loop2'] = (self.boxes['Dp'][1], self.boxes['Cp'][0])\n",
    "            elif self.boxes['D'] is not None:\n",
    "                lens['Dp to D'] = self.boxes['D'][0] - self.boxes['Dp'][1]\n",
    "                regions['Dp to D'] = (self.boxes['Dp'][1], self.boxes['D'][0])\n",
    "            else:\n",
    "                lens['Dp to end'] = len(self.seq) - self.boxes['Dp'][1]\n",
    "                regions['Dp to end'] = (self.boxes['Dp'][1], len(self.seq))\n",
    "        if self.boxes['Cp'] is not None:\n",
    "            lens['Cp'] = self.boxes['Cp'][1] - self.boxes['Cp'][0]\n",
    "            if self.boxes['D'] is not None:\n",
    "                lens['loop3'] = self.boxes['D'][0] - self.boxes['Cp'][1]\n",
    "                regions['loop3'] = (self.boxes['Cp'][1], self.boxes['D'][0] )\n",
    "            else:\n",
    "                lens['Cp to end'] = len(self.seq) - self.boxes['Cp'][1]\n",
    "                regions['Cp to end'] = (self.boxes['Cp'][1] - len(self.seq))\n",
    "        if self.boxes['D'] is not None:\n",
    "            lens['D'] = self.boxes['D'][1] - self.boxes['D'][0]\n",
    "            lens['postD'] = len(self.seq) - self.boxes['D'][1]\n",
    "            regions['postD'] = (self.boxes['D'][1], len(self.seq))\n",
    "        region_sum = sum(lens.values())\n",
    "        #print(\"Seq length {0}. region lens {1}\".format(len(self.seq), region_sum))\n",
    "        self.region_lens = lens\n",
    "        self.set_regions(regions)\n",
    "        return lens\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def read_snoRNA_structures(\n",
    "        ann='/Users/dfporter/pma/miseq/Proteins/FBL/reformat_boxes.txt'):\n",
    "    a_sno = []\n",
    "    snos = complex_area_set()\n",
    "    for li in open(ann).readlines():\n",
    "        if li[0] == '>':\n",
    "            if len(a_sno) > 0:\n",
    "                complex_area = snoRNA(*a_sno)\n",
    "                complex_area.name_from_split()\n",
    "                snos.add(complex_area)\n",
    "            a_sno = [li.rstrip('\\n')]\n",
    "        else:     \n",
    "            a_sno.append(li.rstrip('\\n'))\n",
    "    return snos\n",
    "snos = read_snoRNA_structures()\n",
    "coords = snoUtils.genomic_coordinates_of_snoRNAs()\n",
    "snos.read_coordinates(coords)\n",
    "print(coords)\n",
    "\n",
    "print('len', len(snos.complex_areas))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def get_refseq_ids_of_targets(in_file='/Users/dfporter/pma/miseq/Proteins/counts.xls'):\n",
    "    df = pandas.read_excel(in_file)\n",
    "    df.sort_values(by=['FBL_170723_AGCTAG_AGT'], ascending=False, inplace=True)\n",
    "    return set(df['Gene'].tolist()[:200])\n",
    "\n",
    "\n",
    "ids = get_refseq_ids_of_targets()\n",
    "snos_refseq = snoUtils.refseq_ids_of_snoRNAs()\n",
    "ids = snos_refseq & ids\n",
    "#fa = fasta()\n",
    "#seqs = dict([(name, x) for name, x in fa.items() if name in ids])\n",
    "#print(\"Seqs: {0}\".format(len(seqs)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def add_snoRNA_binding_from_a_bedgraph(\n",
    "        snos, fname, norm_ga, cutoff=1, norm_len=50):\n",
    "    print(fname)\n",
    "    ga = bedgraphs(fname)\n",
    "    snos.add_coverage(\n",
    "            ga, norm_ga, norm_len=norm_len, cutoff=cutoff)\n",
    "        #sno.add_coverage_to_normalized_length(\n",
    "        #    ga, norm_ga, norm_len=norm_len, cutoff=cutoff)\n",
    "        #print(norm_ga)\n",
    "    return norm_ga\n",
    "\n",
    "\n",
    "def add_snoRNA_binding_from_bedgraphs(snos, bedgraph_list, cutoff=1, norm_len=20):\n",
    "    all_regions = ['preC', 'C', 'loop1', 'Dp', 'loop2', 'Cp', 'loop3', 'D', 'postD']\n",
    "    \n",
    "    # Determine the size of the average regions to normalize to.\n",
    "    region_average_lens = snos.average_region_lengths()\n",
    "    normalizer = signal_normalizer(region_average_lens)\n",
    "    normalizer.add_region('Whole area', norm_len)\n",
    "    for fname in bedgraph_list:\n",
    "        print(fname)\n",
    "        add_snoRNA_binding_from_a_bedgraph(\n",
    "            snos, fname, normalizer, cutoff=cutoff, norm_len=norm_len)\n",
    "    normalizer.normalize_regions_by_length()\n",
    "    return normalizer\n",
    "\n",
    "in_dir = '/Users/dfporter/pma/miseq/Runs/170830/sams/consensus/'\n",
    "in_dir = '/Volumes/Seagate/hiseq/170924/sams/consensus/'\n",
    "bedgraph_list = glob.glob(in_dir + '/{prefix}*_deletions_+.wig'.format(prefix='AGCTAG'))\n",
    "print(bedgraph_list)\n",
    "print(snos)\n",
    "normalizer = add_snoRNA_binding_from_bedgraphs(snos, bedgraph_list)\n",
    "\n",
    "#print(normalizer.ga)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "#sns.hls_palette(8, l=.3, s=.8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_arr(norm_ga):\n",
    "    print(norm_ga)\n",
    "    xvals = range(len(norm_ga))\n",
    "    yvals = np.array(norm_ga)\n",
    "    plt.clf()\n",
    "    plt.plot(xvals, yvals)\n",
    "    plt.show()\n",
    "    plt.clf()\n",
    "\n",
    "norm_ga = normalizer.ga['Whole area']\n",
    "print(norm_ga)\n",
    "print('-' * 10)\n",
    "all_regions = ['preC', 'C', 'loop1', 'Dp', 'loop2', 'Cp', 'loop3', 'D', 'postD']\n",
    "combined = normalizer.put_regions_together(all_regions)\n",
    "\n",
    "#plot_arr(combined)\n",
    "plt.clf()\n",
    "start = 0\n",
    "pal = sns.cubehelix_palette(len(all_regions), start=18, rot=-.5)\n",
    "for n, name in enumerate(all_regions):\n",
    "    region_len = normalizer.lengths[name]\n",
    "    print(region_len, len(normalizer.ga[name]))\n",
    "    print(normalizer.ga[name])\n",
    "    plt.plot(range(start, start+region_len), normalizer.ga[name], c=pal[n])\n",
    "    start += region_len\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_regions = ['preC', 'C', 'loop1', 'Dp', 'loop2', 'Cp', 'loop3', 'D', 'postD']\n",
    "fraction = {}\n",
    "running_fraction = []\n",
    "for region in all_regions:\n",
    "    fraction[region] = normalizer.lengths[region]/len(combined)\n",
    "    if region == 'preC':\n",
    "        running_fraction.append(fraction[region])\n",
    "    else:\n",
    "        running_fraction.append(fraction[region] + running_fraction[-1])\n",
    "print(fraction)\n",
    "print(sum(fraction.values()))\n",
    "print(running_fraction)\n",
    "pre_c_a, pre_c_b = np.array([0, running_fraction[0]])  # preC, not plotted\n",
    "seg1_a, seg1_b = np.array([running_fraction[0], running_fraction[1]])  # line up, C box\n",
    "seg2_a, seg2_b = np.array([running_fraction[1], running_fraction[2]])  # left circle, loop\n",
    "seg3_a, seg3_b = np.array([running_fraction[2], running_fraction[3]])  # line up on top, D prime box\n",
    "D_to_C_a, D_to_C_b = np.array([running_fraction[3], running_fraction[4]])  # Dprime to Cprime box loop\n",
    "seg4_a, seg4_b = np.array([running_fraction[4], running_fraction[5]])  # line down on top, C prime box\n",
    "seg5_a, seg5_b = np.array([running_fraction[5], running_fraction[6]])  # right circle, loop\n",
    "seg6_a, seg6_b = np.array([running_fraction[6], running_fraction[7]])  # line down bottom right, D box\n",
    "post_d_a, post_d_b = np.array([running_fraction[7], running_fraction[8]])  # postD, not plotted\n",
    "\n",
    "print(normalizer.lengths)\n",
    "print(combined)\n",
    "print(len(combined))\n",
    "print(sum(normalizer.lengths.values()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.clf()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "c_seg1_a, c_seg1_b = np.array([10, 10.01])  # line up\n",
    "c_seg2_a, c_seg2_b = np.array([0.0, 0.99])  # circle\n",
    "c_seg3_a, c_seg3_b = np.array([10.99, 11.])  # line down \n",
    "\n",
    "def c_segmented_curve(t):\n",
    "\n",
    "    x = 0\n",
    "    y = 0\n",
    "    deg = 0\n",
    "    hairpin_width = 0.5\n",
    "    x0 = 0.1\n",
    "    y0 = 0.1\n",
    "    if(c_seg1_a <= t < c_seg1_b):\n",
    "        x = x0\n",
    "        y = ((t-seg1_a)/(seg1_b - seg1_a)) + y0\n",
    "        deg = 90\n",
    "        \n",
    "    if(c_seg2_a <= t < c_seg2_b):\n",
    "        radius = 1 #(seg2_b - seg2_a)/3\n",
    "        center = [0+x0, radius+1+y0]\n",
    "        fraction_through = (t-seg2_a)/(seg2_b - seg2_a)\n",
    "        compression = 0.15 + 0.7 * fraction_through\n",
    "        v = 1-(compression) \n",
    "        x = center[0] + radius * np.cos(v*np.pi + np.pi/2)\n",
    "        y = center[1] + radius * np.sin(v*np.pi + np.pi/2)\n",
    "        slope_perpendicular = y/x\n",
    "        deg = 180 + 90 + np.rad2deg(v*np.pi + np.pi/2)\n",
    "\n",
    "    if (c_seg3_a <= t <= c_seg3_b):\n",
    "        x = hairpin_width + x0\n",
    "        y = 1 - ((t-seg6_a)/(seg6_b-seg6_a)) + y0\n",
    "        deg = 270\n",
    "    \n",
    "    return x, y, deg\n",
    "\n",
    "\n",
    "\n",
    "def segmented_curve(t):\n",
    "\n",
    "    x = 0\n",
    "    y = 0\n",
    "    deg = 0\n",
    "    hairpin_width = 0.1\n",
    "    x0 = 0.1\n",
    "    y0 = 0.1\n",
    "    \n",
    "    if(pre_c_a <= t < pre_c_b):\n",
    "        deg = 60\n",
    "        seg_len = pre_c_b - pre_c_a\n",
    "        frac_through = (t - pre_c_a)/seg_len\n",
    "        x = x0 - 1.5 + (frac_through) \n",
    "        y = y0 - 1.5 + frac_through\n",
    "        \n",
    "    if(seg1_a <= t < seg1_b):\n",
    "        x = x0\n",
    "        y = ((t-seg1_a)/(seg1_b - seg1_a)) + y0\n",
    "        deg = 90\n",
    "        \n",
    "    if(seg2_a <= t < seg2_b):\n",
    "        # at t = seg2_a, x,y = [0, 0.1]\n",
    "        # at t = seg2_b, x,y = [0, 0.3]\n",
    "        # x gets divided by two so that it goes from 0 to 0.5\n",
    "        # diameter must therefore be seg2_b - seg2_a\n",
    "        radius = 1 #(seg2_b - seg2_a)/3\n",
    "        center = [0+x0, radius+1+y0]\n",
    "        fraction_through = (t-seg2_a)/(seg2_b - seg2_a)\n",
    "        \n",
    "        compression = 0.10 + 0.85 * fraction_through\n",
    "        v = 1-(compression) \n",
    "        \n",
    "        x = center[0] + radius * np.cos(v*np.pi + np.pi/2)\n",
    "        y = center[1] + radius * np.sin(v*np.pi + np.pi/2)\n",
    "        slope_perpendicular = y/x\n",
    "        deg = 180 + 90 + np.rad2deg(v*np.pi + np.pi/2)\n",
    "        \n",
    "    if(seg3_a <= t < seg3_b):\n",
    "        x = x0\n",
    "        y = ((t-seg3_a)/(seg3_b - seg3_a)) + 3 + y0\n",
    "        deg = 90\n",
    "        \n",
    "    if(D_to_C_a <= t < D_to_C_b):\n",
    "        radius = 0.5 #(seg2_b - seg2_a)/3\n",
    "        center = [1+x0, radius+4+y0]\n",
    "        fraction_through = (t - D_to_C_a)/(D_to_C_b - D_to_C_a)\n",
    "        print(t, fraction_through)\n",
    "        compression = 0.10 + 0.85 * fraction_through\n",
    "        v = 1-(compression) \n",
    "        \n",
    "        x = center[0] + radius * np.cos(2*v*np.pi + np.pi/2)\n",
    "        y = center[1] + radius * np.sin(2*v*np.pi + np.pi/2)\n",
    "        slope_perpendicular = y/x\n",
    "        deg = 180 + 90 + np.rad2deg(v*np.pi + np.pi/2)   \n",
    "        \n",
    "    if(seg4_a <= t < seg4_b):\n",
    "        x = hairpin_width + x0\n",
    "        y = 4 - ((t-seg4_a)/(seg4_b-seg4_a)) + y0\n",
    "        deg = 270\n",
    "        \n",
    "    if(seg5_a <= t < seg5_b):\n",
    "        radius = 1\n",
    "        center = [hairpin_width+x0, radius+1+y0]\n",
    "        fraction_through = (t-seg5_a)/(seg5_b - seg5_a)\n",
    "        compression = 0.1 + 0.85 * fraction_through\n",
    "        v = 1 - compression\n",
    "        x = center[0] + radius * np.cos(v*np.pi - np.pi/2)\n",
    "        y = center[1] + radius * np.sin(v*np.pi - np.pi/2)\n",
    "        deg = 270 + np.rad2deg(v*np.pi - np.pi/2)\n",
    "    \n",
    "    if (seg6_a <= t <= seg6_b):\n",
    "        x = hairpin_width + x0\n",
    "        y = 1 - ((t-seg6_a)/(seg6_b-seg6_a)) + y0\n",
    "        deg = 270\n",
    "    if(post_d_a <= t < post_d_b):\n",
    "        deg = 270\n",
    "        seg_len = post_d_b - post_d_a\n",
    "        frac_through = (t - post_d_a)/seg_len\n",
    "        x = hairpin_width + x0 - frac_through\n",
    "        y =  -0.5 + y0  - frac_through\n",
    "    return x, y, deg\n",
    "\n",
    "xvals = range(len(combined))\n",
    "yvals = np.array(combined)/1000\n",
    "fig1 = plt.figure()\n",
    "ax1 = fig1.add_subplot(111, aspect='equal')\n",
    "position = np.arange(len(xvals))\n",
    "\n",
    "pal = sns.color_palette(\"RdBu_r\", 100)\n",
    "sns.palplot(pal)\n",
    "plt.show()\n",
    "plt.clf()\n",
    "#pal = sns.cubehelix_palette(100)\n",
    "def to_color(val, pal):\n",
    "    _min = 0\n",
    "    _max = max(yvals)\n",
    "    return pal[int(100*val/_max)-1]\n",
    "\n",
    "\n",
    "print(normalizer.lengths)\n",
    "for pos, x, bar_length in zip(position, xvals, yvals):\n",
    "    #mybars = plt.rectangle(pos, y, align='center', linewidth=0)\n",
    "    #plt.xticks(position, x)\n",
    "    t = pos/len(position)\n",
    "\n",
    "    x, y, deg = segmented_curve(pos/len(position))\n",
    "    #deg = tangent(t)\n",
    "    ax1.add_patch(patches.Circle(\n",
    "        (x, y),   # (x,y)\n",
    "        linestyle='None',\n",
    "        color=to_color(bar_length, pal),\n",
    "        radius=.05,          # width\n",
    "        #bar_length,          # height\n",
    "        #angle=deg,\n",
    "    ))\n",
    "ax1.set_xlim(-4, 3)\n",
    "ax1.set_ylim(-3, 8)\n",
    "\n",
    "#for spine in plt.gca().spines.values():\n",
    "\n",
    "fig1.set_figwidth(10)\n",
    "fig1.set_figheight(10)\n",
    "#plt.tick_params(top='off', bottom='off', left='off', right='off', labelleft='off', labelbottom='on')\n",
    "#plt.show()\n",
    "plt.savefig('/Users/dfporter/pma/easyCLIP paper/snoRNA_bargraph.pdf')\n",
    "plt.clf(); plt.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
