{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re, pandas, HTSeq, glob, collections, importlib, random, pickle, sys\n",
    "import numpy as np\n",
    "import pprint as pp\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as patches\n",
    "\n",
    "sys.path.append('/Users/dp/pma/')\n",
    "import sameRiver\n",
    "from sameRiver import snoUtils\n",
    "from sameRiver.signal_normalizer import signal_normalizer\n",
    "#from sameRiver.bedgraphs import bedgraphs\n",
    "from sameRiver.area import *\n",
    "importlib.reload(sameRiver.area)\n",
    "import sameRiver.gapped_area\n",
    "importlib.reload(sameRiver.gapped_area)\n",
    "\n",
    "import sameRiver.RNA\n",
    "import sameRiver.set_of_named_mRNAs\n",
    "import sameRiver.gapped_area_mRNA\n",
    "\n",
    "importlib.reload(sameRiver.RNA)\n",
    "importlib.reload(sameRiver.set_of_named_mRNAs)\n",
    "importlib.reload(sameRiver.gapped_area_mRNA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "importlib.reload(sameRiver.RNA)\n",
    "importlib.reload(sameRiver.set_of_named_mRNAs)\n",
    "importlib.reload(sameRiver.gapped_area_mRNA)\n",
    "\n",
    "_mRNAs = sameRiver.set_of_named_mRNAs.set_of_named_mRNAs()\n",
    "\n",
    "_mRNAs.create_set_of_named_mRNAs(\n",
    "    gtf_filename='/opt/genomes/repeats/repeats_and_genome.gtf'\n",
    "    #gtf_filename='/opt/genome/Homo_sapiens.GRCh38.83/exon_CDS_ensembl_havana_only_Homo_sapiens.GRCh38.83_plus_snos_and_scar_from_ensembl_release91_plus_snRNA_this_release.gtf'\n",
    "                                )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_mRNAs.keep_only_mRNA()\n",
    "_mRNAs.example()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "genomic_fasta_fname = '/opt/genome/refseq/GRCh38_latest_genomic.fna'\n",
    "genomic_fasta_fname = '/opt/genome/Homo_sapiens.GRCh38.83/Homo_sapiens.GRCh38.dna.primary_assembly.fa'\n",
    "genomic_fasta = dict( (s[1], s[0]) for s in HTSeq.FastaReader(genomic_fasta_fname, raw_iterator=True) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "rRNA18 = \"\"\">NR_003286.4 Homo sapiens RNA, 18S ribosomal N5 (RNA18SN5), ribosomal RNA\n",
    "TACCTGGTTGATCCTGCCAGTAGCATATGCTTGTCTCAAAGATTAAGCCATGCATGTCTGAGTACGCACG\n",
    "GCCGGTACAGTGAAACTGCGAATGGCTCATTAAATCAGTTATGGTTCCTTTGGTCGCTCGCTCCTCTCCT\n",
    "ACTTGGATAACTGTGGTAATTCTAGAGCTAATACATGCCGACGGGCGCTGACCCCCTTCGCGGGGGGGAT\n",
    "GCGTGCATTTATCAGATCAAAACCAACCCGGTCAGCCCCTCTCCGGCCCCGGCCGGGGGGCGGGCGCCGG\n",
    "CGGCTTTGGTGACTCTAGATAACCTCGGGCCGATCGCACGCCCCCCGTGGCGGCGACGACCCATTCGAAC\n",
    "GTCTGCCCTATCAACTTTCGATGGTAGTCGCCGTGCCTACCATGGTGACCACGGGTGACGGGGAATCAGG\n",
    "GTTCGATTCCGGAGAGGGAGCCTGAGAAACGGCTACCACATCCAAGGAAGGCAGCAGGCGCGCAAATTAC\n",
    "CCACTCCCGACCCGGGGAGGTAGTGACGAAAAATAACAATACAGGACTCTTTCGAGGCCCTGTAATTGGA\n",
    "ATGAGTCCACTTTAAATCCTTTAACGAGGATCCATTGGAGGGCAAGTCTGGTGCCAGCAGCCGCGGTAAT\n",
    "TCCAGCTCCAATAGCGTATATTAAAGTTGCTGCAGTTAAAAAGCTCGTAGTTGGATCTTGGGAGCGGGCG\n",
    "GGCGGTCCGCCGCGAGGCGAGCCACCGCCCGTCCCCGCCCCTTGCCTCTCGGCGCCCCCTCGATGCTCTT\n",
    "AGCTGAGTGTCCCGCGGGGCCCGAAGCGTTTACTTTGAAAAAATTAGAGTGTTCAAAGCAGGCCCGAGCC\n",
    "GCCTGGATACCGCAGCTAGGAATAATGGAATAGGACCGCGGTTCTATTTTGTTGGTTTTCGGAACTGAGG\n",
    "CCATGATTAAGAGGGACGGCCGGGGGCATTCGTATTGCGCCGCTAGAGGTGAAATTCTTGGACCGGCGCA\n",
    "AGACGGACCAGAGCGAAAGCATTTGCCAAGAATGTTTTCATTAATCAAGAACGAAAGTCGGAGGTTCGAA\n",
    "GACGATCAGATACCGTCGTAGTTCCGACCATAAACGATGCCGACCGGCGATGCGGCGGCGTTATTCCCAT\n",
    "GACCCGCCGGGCAGCTTCCGGGAAACCAAAGTCTTTGGGTTCCGGGGGGAGTATGGTTGCAAAGCTGAAA\n",
    "CTTAAAGGAATTGACGGAAGGGCACCACCAGGAGTGGAGCCTGCGGCTTAATTTGACTCAACACGGGAAA\n",
    "CCTCACCCGGCCCGGACACGGACAGGATTGACAGATTGATAGCTCTTTCTCGATTCCGTGGGTGGTGGTG\n",
    "CATGGCCGTTCTTAGTTGGTGGAGCGATTTGTCTGGTTAATTCCGATAACGAACGAGACTCTGGCATGCT\n",
    "AACTAGTTACGCGACCCCCGAGCGGTCGGCGTCCCCCAACTTCTTAGAGGGACAAGTGGCGTTCAGCCAC\n",
    "CCGAGATTGAGCAATAACAGGTCTGTGATGCCCTTAGATGTCCGGGGCTGCACGCGCGCTACACTGACTG\n",
    "GCTCAGCGTGTGCCTACCCTACGCCGGCAGGCGCGGGTAACCCGTTGAACCCCATTCGTGATGGGGATCG\n",
    "GGGATTGCAATTATTCCCCATGAACGAGGAATTCCCAGTAAGTGCGGGTCATAAGCTTGCGTTGATTAAG\n",
    "TCCCTGCCCTTTGTACACACCGCCCGTCGCTACTACCGATTGGATGGTTTAGTGAGGCCCTCGGATCGGC\n",
    "CCCGCCGGGGTCGGCCCACGGCCCTGGCGGAGCGCTGAGAAGACGGTCGAACTTGACTATCTAGAGGAAG\n",
    "TAAAAGTCGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTA\"\"\"\n",
    "def read_fasta(seq):\n",
    "    _dict = {}\n",
    "    current_seq_name = ''\n",
    "    for li in seq.split('\\n'):\n",
    "        if li[0] == '>':\n",
    "            current_seq_name = li[1:]\n",
    "            _dict[current_seq_name] = ''\n",
    "        else:\n",
    "            _dict[current_seq_name] += li.rstrip('\\n')\n",
    "    return _dict\n",
    "a = read_fasta(rRNA18)\n",
    "a = list(a.values())[0]\n",
    "for chrom, seq in genomic_fasta.items():\n",
    "    print('Searching ', chrom)\n",
    "    m = re.search(a, seq)\n",
    "    print(m)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def rc(_seq):\n",
    "    _dict = {'A': 'T', 'T':'A', 'C':'G', 'G':'C', 'N':'N'}\n",
    "    return ''.join([_dict[x.upper()] for x in _seq[::-1]])\n",
    "\n",
    "\n",
    "def is_orf(_seq, non_atg_ok=False, not_end_in_stop_ok=False, \n",
    "           not_divisible_by_three_ok=False):\n",
    "    if (len(_seq) % 3):\n",
    "        print(\"Not div by 3\")\n",
    "        if not not_divisible_by_three_ok:\n",
    "            return False\n",
    "           \n",
    "    if (_seq[:3] != 'ATG'):\n",
    "        print('Does not start with an ATG')\n",
    "        if not non_atg_ok:\n",
    "            return False\n",
    "        \n",
    "    if not((_seq[-3:] == 'TAG') or (_seq[-3:] == 'TGA') or (_seq[-3:] == 'TAA')):\n",
    "        print('Does not end in a stop')\n",
    "        if not not_end_in_stop_ok:\n",
    "            return False\n",
    "        \n",
    "    return True\n",
    "\n",
    "\n",
    "def grab_sequence_from_iv_with_offset(iv, fasta_dict, offset=0):\n",
    "    seq = ''\n",
    "    if iv.chrom in fasta_dict:\n",
    "        seq = fasta_dict[iv.chrom][iv.start - 1 + offset:iv.end + offset].upper()\n",
    "        if iv.strand == '-':\n",
    "            seq = rc(seq)\n",
    "    else:\n",
    "        print(\"Chrom name {0} not found\".format(iv.chrom))\n",
    "        return ''\n",
    "    return seq\n",
    "\n",
    "\n",
    "def read_transcript_sequence(self, fasta_dict):\n",
    "    # Move to be a method.\n",
    "    self.area_sequences = []\n",
    "    self.cds_sequence = ''\n",
    "    \n",
    "    _sequences = [] \n",
    "    for n, _area in enumerate(self.areas):  # For each genomic iv.\n",
    "        _sequences.append(grab_sequence_from_iv_with_offset(_area.iv, fasta_dict, offset=0))\n",
    "    \n",
    "    return _sequences\n",
    "\n",
    "\n",
    "def read_exon_sequence(\n",
    "        self, fasta_dict, offset=0):\n",
    "    #print('===',)\n",
    "    done = False\n",
    "    successful = False\n",
    "    \n",
    "\n",
    "    self.exon_seq_list = []\n",
    "    for n, _exon in enumerate(self.elements['exon']):\n",
    "        self.exon_seq_list.append(\n",
    "            grab_sequence_from_iv_with_offset(_exon, fasta_dict, offset=offset))\n",
    "    self.exon_sequence = ''.join(self.exon_seq_list)\n",
    "    return True\n",
    "\n",
    "\n",
    "def read_premRNA_sequence(\n",
    "        self, fasta_dict, offset=0):\n",
    "    \n",
    "    #print(self.name, '---')\n",
    "    self.premRNA_sequence = ''\n",
    "    for n, _region in enumerate(self.elements['mRNA']):\n",
    "    #    print(_region)\n",
    "        try:\n",
    "            self.premRNA_sequence += grab_sequence_from_iv_with_offset(\n",
    "                _region, fasta_dict, offset=offset)\n",
    "        except:\n",
    "            print(_region)\n",
    "    #print('mRNA element:', self.elements['mRNA'])\n",
    "    #print(self.premRNA_sequence, len(self.premRNA_sequence))\n",
    "    return True\n",
    "\n",
    "def read_cds_sequence(\n",
    "        self, fasta_dict, allow_offset_search=False, offset=0):\n",
    "    #print('===',)\n",
    "    done = False\n",
    "    successful = False\n",
    "    \n",
    "    cds_sequence = ''\n",
    "    \n",
    "    if not allow_offset_search:\n",
    "        self.cds_seq_list = []\n",
    "        for n, _cds in enumerate(self.elements['CDS']):\n",
    "            self.cds_seq_list.append(\n",
    "                grab_sequence_from_iv_with_offset(_cds, fasta_dict, offset=offset))\n",
    "        self.cds_sequence = ''.join(self.cds_seq_list)\n",
    "        return self.cds_sequence, self.cds_seq_list\n",
    "    \n",
    "    offset = 0\n",
    "    tried_offset = set()\n",
    "    offset_to_try = set([-2, -1, 0, 1, 2])\n",
    "    \n",
    "    while not done:\n",
    "        cds_sequence, seq_list = read_cds_seqence(self, fasta_dict, allow_offset_search=False,\n",
    "                                       offset=offset)\n",
    "        orf = is_orf(cds_sequence)\n",
    "        if not orf:\n",
    "            offset = list(offset_to_try - tried_offset)\n",
    "            if len(offset) == 0:\n",
    "                print(\"Could not find acceptible sequence for {0} at any offset.\".format(self.name))\n",
    "                self.cds_sequence, self.cds_seq_list = (cds_sequence, seq_list)\n",
    "                done = True\n",
    "            else:\n",
    "                offset = min(offset)\n",
    "                tried_offset.add(offset)\n",
    "                print(\"Trying offset {0}..,\".format(offset))\n",
    "        else:\n",
    "            successful = True\n",
    "            self.cds_sequence, self.cds_seq_list = (cds_sequence, seq_list)\n",
    "            if (offset != 0):\n",
    "                print(\"Found offset sequence at offset {0}\".format(offset))\n",
    "            else:\n",
    "                pass\n",
    "                #print(\"OK sequnce for {0}\".format(self.name))  # Found without offset.\n",
    "            done = True\n",
    "    #print('+++')\n",
    "    \n",
    "    return successful\n",
    "\n",
    "import random, copy\n",
    "import sameRiver.gapped_area_mRNA\n",
    "gaps = {}\n",
    "success, failure = (set(), set())\n",
    "\n",
    "for n , (name, _mRNA) in enumerate(_mRNAs.mRNAs.items()):\n",
    "    gaps[name] = sameRiver.gapped_area_mRNA.gapped_area_mRNA()\n",
    "    gaps[name].name = name\n",
    "    gaps[name].elements = copy.copy(_mRNA.elements)\n",
    "    _mRNA = gaps[name]\n",
    "    #read_premRNA_sequence(gaps[name], genomic_fasta)\n",
    "    \n",
    "    #read_exon_sequence(gaps[name], genomic_fasta)\n",
    "    \n",
    "    #if not read_cds_sequence(_mRNA, genomic_fasta):\n",
    "    #    print(\"Read {0}/{1} sequences before stopping.\".format(n, len(_mRNAs.mRNAs)))\n",
    "    #    failure.add(_mRNA.name)\n",
    "    #else:\n",
    "    #    success.add(_mRNA.name)\n",
    "    gaps[name].process()\n",
    "#print(_mRNAs.mRNAs)\n",
    "print(\"Failures: {0}\".format(failure))\n",
    "print(\"# success {0}. # Failure {1}\".format(len(success), len(failure)))\n",
    "    #if n > 1000:\n",
    "    #    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gap_rnas = sameRiver.set_of_named_mRNAs.set_of_named_mRNAs()\n",
    "gap_rnas.mRNAs = gaps\n",
    "print(gap_rnas.mRNAs['ENST00000548101'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sameRiver.bedgraphs\n",
    "\n",
    "def add_mRNA_binding_from_bedgraphs(rnas, bedgraph_list, gene, cutoff=20, norm_len=20):\n",
    "    \"\"\"Add normalized signal to a single ga per bedgraph +/- .wig pair.\"\"\"\n",
    "    \n",
    "    print(\"Reading {}...\".format('/Users/dfporter/pma/tables/pvals.xlsx'))\n",
    "    try:\n",
    "        targs = pandas.read_excel('/Users/dfporter/pma/tables/pvals.xlsx',\n",
    "                                 sheet_name=gene)\n",
    "    except:\n",
    "        print(\"Could not open sheet {}. Skipping...\".format(gene))\n",
    "        return signal_normalizer({'left_UTR': 50, 'CDS': 50, 'right_UTR': 50})\n",
    "    \n",
    "    print(\"Translating ENST to gene names...\")\n",
    "    to_do = set(rnas.mRNAs.keys())#[enst_to_name.get(x) for x in rnas.mRNAs])\n",
    "\n",
    "    print(targs.head())\n",
    "    targs = [x.split('::')[0] for x in targs.index]\n",
    "    targs = set([name_to_enst.get(x, x) for x in targs])\n",
    "    print(\"Finding overlap...\")\n",
    "    to_do = to_do & targs\n",
    "    total_rnas = len(to_do)\n",
    "    \n",
    "    all_regions = ['left_UTR', 'CDS', 'right_UTR']\n",
    "    \n",
    "    # Determine the size of the average regions to normalize to.\n",
    "    region_average_lens = rnas.average_region_lengths()\n",
    "    for k, v in region_average_lens.items():\n",
    "        region_average_lens[k] = int(v/10)\n",
    "    normalizer = signal_normalizer(region_average_lens)\n",
    "    \n",
    "    #normalizer = signal_normalizer({'left_UTR': 50, 'CDS': 50, 'right_UTR': 50})\n",
    "    normalizer.add_region('Whole area', norm_len)\n",
    "    for rna in rnas.mRNAs.values():\n",
    "        rna.zero()\n",
    "    for n_bedgraph, fname in enumerate(bedgraph_list, start=1):\n",
    "        \n",
    "        print(\"Loading bedgraph {}/{}: {}...\".format(\n",
    "            n_bedgraph, len(bedgraph_list), fname))\n",
    "        \n",
    "        ga = sameRiver.bedgraphs.bedgraph(fname).ga\n",
    "        print(fname)\n",
    "        #for iv, val in ga[HTSeq.GenomicInterval('1', 0, 1E9, '+')].steps():\n",
    "        #    print(iv, val)\n",
    "        #    if random.randint(0, 100) == 1:\n",
    "        #        break\n",
    "\n",
    "        for n_rna, name in enumerate(rnas.mRNAs):#to_do):#rnas.mRNAs.values()):\n",
    "            \n",
    "            mRNA = rnas.mRNAs[name]\n",
    "            #print(mRNA.name)\n",
    "            \n",
    "            if len(mRNA.areas) == 0:\n",
    "                continue\n",
    "            #if mRNA.areas[0].iv.strand == '-':\n",
    "            #    continue\n",
    "\n",
    "            if random.randint(0, 50000) == 1:\n",
    "                print(mRNA.name, \"{}/{}\".format(n_rna, total_rnas))\n",
    "  \n",
    "            #total = sum([sum(_arr) for _arr in mRNA.raw_signal_by_type.values()])\n",
    "\n",
    "            \n",
    "            mRNA.add_coverage_from_mRNA(\n",
    "                ga, normalizer, cutoff=cutoff)  # Normalized.\n",
    "            \n",
    "            #if total > 10 and random.randint(0, 100) == 0:\n",
    "            #    print(normalizer.ga)\n",
    "                \n",
    "    normalizer.normalize_areas_by_number_of_arrays_added()\n",
    "    print([(name, sum(x)) for name, x in normalizer.ga.items()])\n",
    "    \n",
    "    normalizer.normalize_regions_by_length()\n",
    "    print(\"Number of arrays added: {0}\".format(normalizer.added_to_area))\n",
    "    print([(name, sum(x)) for name, x in normalizer.ga.items()])\n",
    "    return normalizer\n",
    "\n",
    "\n",
    "gtf_translation_fname = '/opt/genome/Homo_sapiens.GRCh38.83/enst_transcript_id_name_biotype_map.txt'\n",
    "transl_df = pandas.read_csv(\n",
    "    gtf_translation_fname, sep='\\t')\n",
    "enst_to_name = dict(zip(transl_df.transcript_id, transl_df.gene_name))\n",
    "name_to_enst = dict(zip(transl_df.gene_name, transl_df.transcript_id))\n",
    "print(\"Loaded translations...\")\n",
    "\n",
    "df = pandas.read_excel('/Users/dfporter/pma/miseq/meta/scheme.xls')\n",
    "print(df.head())\n",
    "gene_bc = dict(zip(df['Gene'].tolist(), df.P6_BC))\n",
    "plots = []\n",
    "for gene, p6bc in gene_bc.items():\n",
    "    skip = False\n",
    "    for _str in ['AURKA', 'YBX', 'STD', 'hnRNPC', 'FBL', '100', 'PCBP1',\n",
    "                'SF3B1', '700', 'nknown']:\n",
    "        if _str in gene:\n",
    "            skip = True\n",
    "    \n",
    "    if skip:\n",
    "        continue\n",
    "    print('Gene {}'.format(gene))\n",
    "    \n",
    "    in_dir = '/Users/dfporter/pma/miseq/Runs/171210/'\n",
    "    in_dir = '/Users/dfporter/pma/miseq/meta/beds/read_start/'\n",
    "    #in_dir = '/Volumes/Seagate/hiseq/170924/sams/consensus/'\n",
    "    #in_dir = '/Users/dfporter/pma/miseq/Runs/170915/sams/consensus/'\n",
    "\n",
    "    bedgraph_list = glob.glob(in_dir + '/{prefix}*_+.wig'.format(prefix=p6bc))\n",
    "    #in_dir = '/Users/dfporter/pma/miseq/Runs/170524/sams/consensus/' # TGCCGA\n",
    "    #bedgraph_list = glob.glob(in_dir + '/{prefix}*_deletions_+.wig'.format(prefix='TGCCGA'))  # DDX3X \n",
    "\n",
    "    #bedgraph_list = glob.glob(in_dir + '/{prefix}*_TCA_+.wig'.format(prefix='24TGAGTG'))  # hnRNPC (?)\n",
    "\n",
    "    #bedgraph_list = glob.glob(in_dir + '/{prefix}*_deletions_+.wig'.format(prefix='AGCTAG'))  # FBL \n",
    "    #bedgraph_list = glob.glob(in_dir + '/{prefix}*_deletions_+.wig'.format(prefix='TAGTTG'))  # CSRP2 \n",
    "\n",
    "    #bedgraph_list = glob.glob(in_dir + '/{prefix}*_deletions_+.wig'.format(prefix='TAGTTG'))  # DDX3X \n",
    "\n",
    "    #bedgraph_list = glob.glob(in_dir + '/{prefix}*_deletions_+.wig'.format(prefix='TTAGGC'))  # RQCD1 \n",
    "    #bedgraph_list += glob.glob(in_dir + '/{prefix}*_deletions_+.wig'.format(prefix='CCACTC'))  # RQCD1 \n",
    "\n",
    "    #bedgraph_list = glob.glob(in_dir + '/{prefix}*_deletions_+.wig'.format(prefix='TGAGTG'))  # RQCD1 131L\n",
    "    #bedgraph_list += glob.glob(in_dir + '/{prefix}*_deletions_+.wig'.format(prefix='GTCGTC'))  # RQCD1 131L\n",
    "\n",
    "    #bedgraph_list = glob.glob(in_dir + '/{prefix}*_deletions_+.wig'.format(prefix='ATCGTG'))  # PCBP1 in 171210\n",
    "    #bedgraph_list += glob.glob(in_dir + '/{prefix}*_deletions_+.wig'.format(prefix='ATCGTG'))  # PCBP1 in 171210\n",
    "\n",
    "    print(bedgraph_list)\n",
    "    normalizer = add_mRNA_binding_from_bedgraphs(gap_rnas, bedgraph_list, gene)\n",
    "    plots.append(\n",
    "        plot_normalized_signal(normalizer, gene))\n",
    "\n",
    "\n",
    "print('done.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def seq_around_point_of_highest_coverage(self, fasta_dict, bedgraph_name='Signal',):\n",
    "    point = point_of_highest_coverage(self, bedgraph_name=bedgraph_name)\n",
    "    if point:\n",
    "        iv_for_seq = self.genomic_point_of_highest_coverage\n",
    "        iv_for_seq.start -= 5\n",
    "        iv_for_seq.end += 5\n",
    "        seq = grab_sequence_from_iv_with_offset(iv_for_seq, fasta_dict=fasta_dict)\n",
    "        #print(seq)\n",
    "        self.seq_around_highest_point = seq\n",
    "        return seq\n",
    "    else:\n",
    "        self.seq_around_highest_point = ''\n",
    "        return ''\n",
    "\n",
    "    \n",
    "def grab_sequence_from_iv_with_offset(iv, fasta_dict, offset=0):\n",
    "    seq = ''\n",
    "    if iv.chrom in fasta_dict:\n",
    "        seq = fasta_dict[iv.chrom][iv.start - 1 + offset:iv.end + offset].upper()\n",
    "        if iv.strand == '-':\n",
    "            seq = rc(seq)\n",
    "    else:\n",
    "        print(\"Chrom name {0} not found\".format(iv.chrom))\n",
    "        return ''\n",
    "    return seq\n",
    "    \n",
    "    \n",
    "def point_of_highest_coverage(self, bedgraph_name='Signal',):\n",
    "    #  Move to be a method.\n",
    "    if bedgraph_name not in self.bedgraphs:\n",
    "        print(\"No {0} in bedgraphs.\".format(bedgraph_name))\n",
    "        self.coverage_array = np.array([0])\n",
    "        return False\n",
    "    arr = []\n",
    "    \n",
    "    n_steps = 0\n",
    "    for _, _v in self.bedgraphs['Signal'].steps():\n",
    "        n_steps += 1\n",
    "    \n",
    "    mRNA = self.elements['mRNA'][0]\n",
    "\n",
    "    for n, (step_iv, value) in enumerate(self.bedgraphs['Signal'][mRNA].steps()):\n",
    "        if (step_iv.end - step_iv.start) < 1E6:\n",
    "            arr += [value] * (step_iv.end - step_iv.start)\n",
    "            \n",
    "    arr = np.array(arr)  # Replace this with sensible code when we can look up the methods.\n",
    "    if len(arr) > 0:\n",
    "        _max = arr.max()\n",
    "        self.coverage_array = arr\n",
    "    else:\n",
    "        self.coverage_array = np.array([])\n",
    "        _max = 0\n",
    "        return False\n",
    "    \n",
    "    indexes_of_maxes = []\n",
    "    for n, v in enumerate(arr):\n",
    "        if v == _max:\n",
    "            indexes_of_maxes.append(n)\n",
    "    max_index = random.choice(indexes_of_maxes)\n",
    "    \n",
    "    if _max > 0:\n",
    "        cv = self.bedgraphs['Signal'][mRNA]\n",
    "        self.genomic_point_of_highest_coverage = cv.iv\n",
    "        self.genomic_point_of_highest_coverage.start = cv.iv.start + max_index \n",
    "        self.genomic_point_of_highest_coverage.end = self.genomic_point_of_highest_coverage.start + 1\n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "for n , (name, _mRNA) in enumerate(_mRNAs.mRNAs.items()):\n",
    "    #print(_mRNA.__dict__)\n",
    "    p = seq_around_point_of_highest_coverage(_mRNA, genomic_fasta)\n",
    "\n",
    "    \n",
    "seqs = []\n",
    "for n, (name, _mRNA) in enumerate(_mRNAs.mRNAs.items()):\n",
    "    if (len(_mRNA.coverage_array) > 0) and max(_mRNA.coverage_array) > 1:\n",
    "        print(_mRNA.seq_around_highest_point)\n",
    "        print(_mRNA.genomic_point_of_highest_coverage)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def item_summary(v):\n",
    "    q = ''\n",
    "    def _len(_object):\n",
    "        try:\n",
    "            return len(_object)\n",
    "        except:\n",
    "            return 'none'\n",
    "    def _try_float(w):\n",
    "        try:\n",
    "            return float(w)\n",
    "        except:\n",
    "            return 0.\n",
    "    if type(v) == type([]):\n",
    "        #q = \"type is list. There are {0} element(s).\\n\".format(len(v))\n",
    "        #try:\n",
    "        q += \"Lengths are: \" + \", \".join([str(_len(x)) for x in v])\n",
    "        q += \" -> sum= {0}\\n\".format(sum([_try_float(_len(x)) for x in v]))\n",
    "        try:\n",
    "            q += [item_summary(x) for x in v]\n",
    "            q += '~' * 12 + '\\n'\n",
    "        except:\n",
    "            pass\n",
    "\n",
    "        #except:\n",
    "        #    pass\n",
    "        q += '\\n'\n",
    "        return q\n",
    "    elif type(v) == type(''):\n",
    "        q += \"String is {0} chars long.\\n\".format(len(v))\n",
    "        return q\n",
    "    else:\n",
    "        return ''\n",
    "def summary(self, return_summary=False, print_summary=True):\n",
    "    li = '=' * 50 + '\\n'\n",
    "    for k, v in self.__dict__.items():\n",
    "        li += '-' * 20 + '\\n'\n",
    "        li += ':'.join([self.name, k]) + '\\n'\n",
    "        li += 'First chars: ' + str(v)[:100] + '\\n'\n",
    "        li += item_summary(v)\n",
    "    if 'bedgraphs' in self.__dict__:\n",
    "        for k, v in self.bedgraphs.items():\n",
    "            li += 'Bedgraph \"{0}\": '.format(k)\n",
    "            li += str(v.chrom_vectors) + '\\n'\n",
    "            #  Assume only one chromosome.\n",
    "            _chrom = [x for x in v.chrom_vectors.keys()][0]\n",
    "            li += \"Chrom {0} \".format(_chrom)\n",
    "            li += \"Strand {0}\\n\".format(self.areas[0].iv.strand)\n",
    "            li += 'Coverage array length: '\n",
    "            arr_len = 0\n",
    "            for exon_n, ex in self.exons.items():\n",
    "                arr = []\n",
    "                for iv, coverage in v[ex.iv].steps():\n",
    "                    arr += [coverage] * (iv.end - iv.start)\n",
    "                arr_len += len(arr)\n",
    "            li += \"{0}\".format(arr_len)\n",
    "    if print_summary:\n",
    "        print(li)\n",
    "for mRNA in _mRNAs.mRNAs.values():\n",
    "    print(mRNA.__dict__.keys())\n",
    "    summary(mRNA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "\n",
    "def plot_normalized_signal(normalizer, gene_name='Gene name', plots=[]):\n",
    "    print(\"Plotting normalized signal for {}...\".format(gene_name))\n",
    "    \n",
    "    all_regions = ['left_UTR', 'CDS', 'right_UTR']\n",
    "    combined = normalizer.put_regions_together(['left_UTR', 'CDS', 'right_UTR'])\n",
    "    #plot_arr(combined)\n",
    "    plt.clf()\n",
    "    fig = plt.figure()\n",
    "    start = 0\n",
    "    pal = sns.cubehelix_palette(len(all_regions), start=18, rot=-.5)\n",
    "    for n, name in enumerate(all_regions):\n",
    "        print(sum(normalizer.ga[name]))\n",
    "        region_len = normalizer.lengths[name]\n",
    "        print(region_len, len(normalizer.ga[name]))\n",
    "\n",
    "        plt.plot(range(start, start+region_len), normalizer.ga[name], c=pal[n])\n",
    "        start += region_len\n",
    "    plt.xlabel('Position in average mRNA (scaled)')\n",
    "    plt.ylabel('Fraction of hnRNP C reads')\n",
    "    print ('--')\n",
    "    print(sum(combined))\n",
    "    #plt.ylim((0,0.03))\n",
    "    #plt.xlim(0, 800)\n",
    "    fig.savefig('figs/normalized_coverage_{}.pdf'.format(gene_name))\n",
    "    plt.show()\n",
    "    \n",
    "    plt.clf()\n",
    "    plt.close()\n",
    "    \n",
    "    plots.append({'x': range(start, start+region_len),\n",
    "                 'Gene name': gene_name,\n",
    "                 'y': copy.copy(normalizer.ga)})\n",
    "    return plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
