{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, re, matplotlib, pandas, collections, importlib, sys\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "\n",
    "sys.path.append('/Users/dp/pma/')\n",
    "\n",
    "import sameRiver\n",
    "import sameRiver.heatmapMaker\n",
    "import sameRiver.biotypeAdder\n",
    "import sameRiver.scheme\n",
    "import sameRiver.positiveCounts\n",
    "from sameRiver.readsPerGene import *\n",
    "import sameRiver.metadata.negative_metadata as negative_metadata\n",
    "import sameRiver.metadata.positive_metadata_all as positive_metadata\n",
    "\n",
    "# Global file paths.\n",
    "top = '/Users/dp/pma/dataAndScripts/clip/old_mapping_meta/'\n",
    "figs_dir = f'/Users/dp/pma/dataAndScripts/clip/figs/'\n",
    "\n",
    "# Move to top directory and make some output folders.\n",
    "os.chdir(top)\n",
    "os.makedirs(f'{top}/logs/', exist_ok=True)\n",
    "os.makedirs(f'{top}/tables/', exist_ok=True)\n",
    "\n",
    "# Useful function.\n",
    "def numeric_columns(df):\n",
    "    return [x for x in df.columns if (df[x].dtype.kind in 'bifc')]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "importlib.reload(sameRiver.heatmapMaker)\n",
    "\n",
    "#####################################################################\n",
    "# Heatmap figure early in paper showing reads per gene for RBPs.\n",
    "#####################################################################\n",
    "\n",
    "positive_metadata.positive_proteins = [\n",
    "    'FBL', 'hnRNPC', 'SF3B1', 'PCBP1',\n",
    "    'CELF1', 'Rbfox1', 'Rbfox2', 'hnRNPD',\n",
    "    'A1CF', 'FUBP1', 'KHDRBS2', #'RPL5', \n",
    "]\n",
    "\n",
    "#positives = sameRiver.positiveCounts.positiveCounts(positive_metadata, xl_rate_fname='/Users/dp/pma/percentCrosslinked.xlsx')\n",
    "#positives.save(write_object=True, write_txt=True)\n",
    "\n",
    "# Using a readsPerGene object to ensure blacklist is applied.\n",
    "rpg = readsPerGene(f'{top}/data/positive_counts_per_read.txt', scheme_filename=f'{top}/scheme.xlsx')\n",
    "df = rpg.df\n",
    "\n",
    "numeric = df[numeric_columns(df)].copy()\n",
    "\n",
    "# RBPs only:\n",
    "numeric = numeric.loc[:, [x for x in numeric.columns if any([pos in x for pos in positive_metadata.positive_proteins ])]]\n",
    "\n",
    "with open(f'{top}/logs/datasets_used_for_heatmap.py', 'w') as f:\n",
    "    f.write('datasets_used_for_heatmap = ' + str(list(numeric.columns)) + '\\n')\n",
    "    \n",
    "\n",
    "def figure_index(numeric):\n",
    "    scheme = sameRiver.scheme.scheme('scheme.xlsx')\n",
    "    labels = collections.defaultdict(int)\n",
    "    \n",
    "    def relabel(cols_to_relabel):\n",
    "        # This function gives replicate names based on Exp number in order to properly pair WT and MUT by batch.\n",
    "        \n",
    "        # Get (Exp number as string, protein, fname).\n",
    "        as_tuples = [(fname.split('_')[0], scheme.gene_from_fname(fname), fname) for fname in cols_to_relabel]\n",
    "        # Get (fname, protein, Exp number as float)\n",
    "        as_tuples = [(fname, p, float(re.search('(\\d+)', exp).group(1))) for exp, p, fname in as_tuples]\n",
    "        # Organize by protein name.\n",
    "        by_gene = collections.defaultdict(list)\n",
    "        for (fname, prot, exp_number) in as_tuples:\n",
    "            by_gene[prot].append((fname, exp_number))\n",
    "        # Get a mapping of protein name -> [fname1, fname2, ...] ordered by replicate based on exp number.\n",
    "        for prot in by_gene:\n",
    "            by_gene[prot] = sorted(by_gene[prot], key=lambda x: x[1])  # Sort by exp number\n",
    "            by_gene[prot] = [x[0] for x in by_gene[prot]]  # Just keep the fname.\n",
    "        \n",
    "        # Get and return the new column names in \"A1CF:E34K R2\" format.\n",
    "        rep_numbers = [by_gene[scheme.gene_from_fname(fname)].index(fname)+1 for fname in cols_to_relabel]\n",
    "        return [f'{scheme.gene_from_fname(fname)} R{n}' for fname, n in zip(cols_to_relabel, rep_numbers)]\n",
    "    \n",
    "    new_index = relabel(numeric.columns)\n",
    "    numeric.columns = new_index\n",
    "\n",
    "# Change column names from Exp#_protein_BC_BC format to Protein R# format.\n",
    "figure_index(numeric)\n",
    "\n",
    "# Subset to the most bound RNAs.\n",
    "numeric['sum'] = numeric.sum(axis=1)\n",
    "numeric = numeric.head(1000).copy()  # Used 5000 for the figure for RBPs.\n",
    "del numeric['sum']\n",
    "\n",
    "# Make a correlation heatmap. Default is by spearman.\n",
    "hm = sameRiver.heatmapMaker.heatmapMaker()\n",
    "hm.heatmap(\n",
    "    numeric, table_filename=f'{top}/tables/spearman_correlations_randos.xlsx',\n",
    "    fig_filename=figs_dir + '/spearman_correlations_randos_heatmap.pdf',\n",
    "    edit_column_names=False, cutoff=10,\n",
    "    cmap=sns.cubehelix_palette(100, start=.5, rot=-0.5, as_cmap=True))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "importlib.reload(sameRiver.heatmapMaker)\n",
    "#_scheme = sameRiver.scheme.scheme('/Users/dfporter/pma/dataAndScripts/clip/miseq/Runs/190326/meta/scheme.xls')\n",
    "\n",
    "top_dir = '/Users/dp/pma/dataAndScripts/clip/meta/'\n",
    "df = pandas.read_csv(f'{top_dir}/data/negative_counts_per_read.txt', sep='\\t', index_col=0)\n",
    "\n",
    "#ba = sameRiver.biotypeAdder.biotypeAdder.load()\n",
    "#df = ba.add_biotypes_column_from_gene_name(df)\n",
    "\n",
    "#df['Gene type'] = [re.sub('.*/.*', 'Repetitive element', x) for x in df['Gene type']]\n",
    "#df['Gene type'] = [re.sub('LTR/.*', 'LTR', x) for x in df['Gene type']]\n",
    "#df['Gene type'] = [re.sub('DNA/hAT.*', 'DNA/hAT', x) for x in df['Gene type']]\n",
    "df = df[df['Gene type']!='Repetitive element']\n",
    "\n",
    "to_del = []\n",
    "#to_del = [x for x in df.columns if not re.search('PCBP1', x)]\n",
    "to_del.extend([x for x in df.columns if (re.search('Exp61_PCBP1-100P', x))])\n",
    "to_del.extend([x for x in df.columns if (re.search('Exp61_PCBP1-dKH', x))])\n",
    "for col in to_del:\n",
    "    del df[col]\n",
    "\n",
    "\n",
    "numeric = df[numeric_columns(df)].copy()\n",
    "\n",
    "hm = sameRiver.heatmapMaker.heatmapMaker()\n",
    "\n",
    "figs_dir = '/Users/dp/pma/dataAndScripts/clip/figs/'\n",
    "tables_dir = './tables/'\n",
    "\n",
    "if not os.path.exists(tables_dir):\n",
    "    os.system('mkdir ' + tables_dir)\n",
    "\n",
    "hm.heatmap(numeric, table_filename=tables_dir + '/spearman_correlations_randos.xlsx',\n",
    "          fig_filename=figs_dir + '/spearman_correlations_randos_heatmap.pdf',\n",
    "          edit_column_names=False, cutoff=10)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sameRiver\n",
    "from sameRiver import *\n",
    "\n",
    "from sameRiver.stacked_bargraph import *\n",
    "\n",
    "\n",
    "importlib.reload(sameRiver.stacked_bargraph)\n",
    "from sameRiver.stacked_bargraph import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Figure 1.\n",
    "FBL and hnRNPC were anti-correlated\n",
    "\"\"\"\n",
    "import sameRiver.heatmapMaker\n",
    "\n",
    "importlib.reload(sameRiver.countsO)\n",
    "importlib.reload(sameRiver.statsForCounts)\n",
    "importlib.reload(sameRiver.heatmapMaker)\n",
    "\n",
    "\n",
    "def figure_5b_heatmap(_df):\n",
    "\n",
    "    df = _df.copy()\n",
    "    sameRiver.countsO.countsO.drop_odds(df)\n",
    "\n",
    "    hits_columns = sameRiver.countsO.countsO.find_hits_columns(df, verbose=True)\n",
    "    \n",
    "    df = df[hits_columns]\n",
    "    \n",
    "    n_cols = len(hits_columns)\n",
    "    \n",
    "\n",
    "    df['sum'] = df.sum(axis=1, numeric_only=True)\n",
    "    df['average'] = [x/max(n_cols, 1) for x in df['sum'].tolist()]\n",
    "    \n",
    "    df = df[df['average']>100].copy()\n",
    "    print(\"Clustering using {0} RNAs.\".format(len(df.index)))\n",
    "    \n",
    "    del df['average']\n",
    "    del df['sum']\n",
    "    \n",
    "    plt.clf()\n",
    "    \n",
    "    def _log(_):\n",
    "        try:\n",
    "            if _ <= 0:\n",
    "                return 0\n",
    "            else:\n",
    "                return np.log10(_)\n",
    "        except:\n",
    "                return 0\n",
    "\n",
    "    #fig, ax = plt.subplots()\n",
    "    m = df.copy()#df[hits_columns].copy()\n",
    "    m = m.dropna(axis=0, how='any')\n",
    "\n",
    "    \n",
    "    corr_m = m.corr(method='spearman')\n",
    "    \n",
    "    corr_m.to_excel('../tables/spearman_correlations.xls')\n",
    "    \n",
    "    fig = plt.figure()\n",
    "    \n",
    "    sns.set(font_scale=1)\n",
    "    \n",
    "    hm = sns.clustermap(corr_m, square=True, yticklabels=1, xticklabels=1,\n",
    "                     cmap=sns.light_palette('black', as_cmap=True)\n",
    "                    #cmap=sns.cubehelix_palette(100, start=.5, rot=-2, as_cmap=True))\n",
    "                    )\n",
    "    #ax.tick_params(labelsize=1)\n",
    "    \n",
    "    plt.savefig('../figs/heatmap_FBL_hnRNPC.pdf')\n",
    "    plt.show()\n",
    "    plt.clf()\n",
    "    plt.close()\n",
    "\n",
    "\n",
    "def is_usable(col):\n",
    "    if ('nknown' in col) or ('xPCBP1' in col) or ('ox' in col) or ('xSF3B1' in col) \\\n",
    "    or('hnRNPC:' in col) or ('CSRP2' in col):\n",
    "        return False\n",
    "    return True\n",
    "\n",
    "def in_fig5b(col):\n",
    "    if re.search('hnRNPC', col) or re.search('FBL', col):\n",
    "        if not re.search(':', col):\n",
    "            return True\n",
    "    return False\n",
    "\n",
    "hm = sameRiver.heatmapMaker.heatmapMaker()\n",
    "\n",
    "figs_dir = '/Users/dfporter/pma/dataAndScripts/clip/figs/'\n",
    "tables_dir = './tables/'\n",
    "\n",
    "if not os.path.exists(tables_dir):\n",
    "    os.system('mkdir ' + tables_dir)\n",
    "\n",
    "hm.heatmap(df, table_filename=tables_dir + '/spearman_correlations_randos.xlsx',\n",
    "          fig_filename=figs_dir + '/spearman_correlations_randos_heatmap.pdf',\n",
    "          edit_column_names=False)\n",
    "\n",
    "df = counts.counts_per_million_df[\n",
    "    [x for x in counts.counts_per_million_df.columns if in_fig5b(x)]]\n",
    "if len(df.columns):\n",
    "    hm.heatmap(df, table_filename=tables_dir + '/FBL_hnRNPC_spearman_correlations.xlsx',\n",
    "              fig_filename=figs_dir + '/spearman_correlations_FBL_hnRNPC_heatmap.pdf',\n",
    "              cutoff=50)\n",
    "\n",
    "df = counts.counts_per_million_df[\n",
    "    [x for x in counts.counts_per_million_df.columns if (('PCBP1' in x) and ('ox' not in x))]\n",
    "]\n",
    "#print(df)\n",
    "df = counts.counts_per_million_df\n",
    "hm.heatmap(df, table_filename=tables_dir + '/PCBP1_spearman_correlations.xlsx',\n",
    "          fig_filename=figs_dir + '/spearman_correlations_PCBP1_heatmap.pdf',\n",
    "        edit_column_names=False,\n",
    "          cutoff=100, annot=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(counts.raw_counts_df.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Figure 5A.\n",
    "First, the data was consistent between replicates (Figure 5A). \n",
    "\"\"\"\n",
    "\n",
    "def replicate_heatmap_figure_5a(\n",
    "    _counts, proteins=set(['FBL', 'hnRNPC'])):\n",
    "    \n",
    "    df = _counts.counts_df.copy()\n",
    "    drop_odds(df)\n",
    "    \n",
    "    df = df.dropna(axis=0, how='any')\n",
    "    \n",
    "    for protein in proteins:\n",
    "        \n",
    "        cols = _counts.cols_of_protein(protein)\n",
    "        sub = df[cols].copy()\n",
    "        print(cols)\n",
    "        sub['sums'] = sub.sum(axis=1)\n",
    "        #return\n",
    "        print(\"Jointplot for \", protein)\n",
    "        print('---')\n",
    "        \n",
    "        if len(cols) < 2:\n",
    "            continue\n",
    "            \n",
    "        #sub = sub.loc[sub['sums']>=0.5]\n",
    "        x = cols[0]\n",
    "        y = cols[1]\n",
    "        _x = np.log10(np.array(sub[cols[0]]))\n",
    "        _y = np.log10(np.array(sub[cols[1]]))\n",
    "        #print(_x, _y)\n",
    "        xmin = _x.min()\n",
    "        xmax = _x.max()\n",
    "        ymin = _y.min()\n",
    "        ymax = _y.max()\n",
    "        xy = np.array([_x, _y]).T\n",
    "        print(xy)\n",
    "        print(xy.ndim)\n",
    "        print(\"[:, 0]->\", xy[:, 0])\n",
    "        print(\"[:, 1]->\", xy[:, 1])\n",
    "        print(\"[0, :]->\", xy[0, :])\n",
    "        \n",
    "        def above(arr):\n",
    "            output = []\n",
    "            for num in arr:\n",
    "                num = np.array(num)\n",
    "                #print(num)\n",
    "                if not np.isfinite(num).all():\n",
    "                    #print(\"infinite\")\n",
    "                    output.append(False)\n",
    "                elif np.all(num > 1):\n",
    "                    #print('True')\n",
    "                    output.append(True)\n",
    "                else:\n",
    "                    output.append(False)\n",
    "            return output #[True] * len(arr)\n",
    "        print('xy[~np.isnan(xy)]=', xy[~np.isnan(xy)])\n",
    "        print('np.isfinite(xy)=', np.isfinite(xy))\n",
    "        print('np.isfinite(xy).all(axis=1)=', np.isfinite(xy).all(axis=1))\n",
    "        tr = np.isfinite(xy).all(axis=1)\n",
    "        print('transpose np.isfinite(xy).all(axis=1)=', np.transpose(np.isfinite(xy).all(axis=1)))\n",
    "        print('tr = ', tr)\n",
    "        print('tr.ndim', tr.ndim)\n",
    "        tr = np.array(tr).flatten()\n",
    "        print(np.array(above(xy)).flatten())\n",
    "        print('sum:', sum([1 for val in np.array(above(xy)).flatten() if val]))\n",
    "        print('above: ', xy[np.array(above(xy))])\n",
    "        print(tr.ndim)\n",
    "        print('after flatten:', tr)\n",
    "        print('xy[np.isfinite(xy)]=', xy[np.isfinite(xy)])\n",
    "        print('xy[np.isfinite(xy).all(axis=1)=', xy[tr])\n",
    "        print('------')\n",
    "        xy = xy[np.array(above(xy))]\n",
    "        if len(xy[:, 0]) == 0:\n",
    "            continue\n",
    "        print(xy)\n",
    "        #print('xy[:, 0]---->', xy[:, 0])\n",
    "        #print('xy[:, 0]---->', np.array(xy[:, 0]).flatten())\n",
    "        plt.clf(); plt.close()\n",
    "        sns.set_style('ticks')\n",
    "        plt.figure(figsize=(4,4))\n",
    "        #fig, ax = plt.subplots()\n",
    "#        ax = axs[0]\n",
    "        hb = sns.jointplot(np.array(xy[:, 0]), np.array(xy[:, 1]), \n",
    "                           #kind='hexbin',#stat_func=kendalltau,\n",
    "                          #xlim=(0,3), ylim=(0,3),\n",
    "                          #marginal_kws=dict(rug=True)\n",
    "                           alpha=0.2, \n",
    "                           color='k',\n",
    "                         #shade=True,\n",
    "                           \n",
    "                          )#, bins=1000)#, alpha=0.2)#,data=sub)\n",
    "        hb.set_axis_labels('Log10 reads/gene Rep 1', 'Log10 reads/gene Rep 2')\n",
    "#        ax.axis([1* x for x in (0, 1, 0, 1)])\n",
    "        #cb = fig.colorbar(hb, ax=ax)\n",
    "        #plt.xlabel('Log10 reads/gene Rep 1')\n",
    "        #plt.ylabel('Log10 reads/gene Rep 2')\n",
    "        #plt.tight_layout()\n",
    "        fig = plt.gcf()\n",
    "        \n",
    "        fig.savefig('../figs/{0}_replicates.pdf'.format(protein))\n",
    "        plt.show()\n",
    "        plt.clf()\n",
    "        plt.close()\n",
    "        print('==')\n",
    "\n",
    "        \n",
    "def keep(_df, to_keep=['hnRNPC', 'FBL']):\n",
    "    \n",
    "    def has_gene(_x):\n",
    "        found = False\n",
    "\n",
    "        for gene in to_keep:\n",
    "            #if re.search(':', _x):\n",
    "            #    continue\n",
    "            m = re.search(gene, _x)\n",
    "\n",
    "            if m is not None:\n",
    "                found = True\n",
    "        if found:\n",
    "            return _x\n",
    "        else:\n",
    "            return 'None'\n",
    "        \n",
    "    _df.columns = [has_gene(x) for x in _df.columns]\n",
    "    \n",
    "    if 'None' in _df.columns:\n",
    "        del _df['None']\n",
    "        \n",
    "    return _df\n",
    "\n",
    "\n",
    "#df = pandas.read_csv('/Users/dfporter/pma/miseq/meta/counts.txt', sep='\\t')\n",
    "#_df = keep(df,\n",
    "#          to_keep=['TTA']\n",
    "          #to_keep=[ 'CDK4', 'TPGS2', 'UBA2', 'ITPA']\n",
    "          #to_keep=protein_order)\n",
    "          #to_keep=['SF3B1'])\n",
    "#         )\n",
    "\n",
    "replicate_heatmap_figure_5a(counts)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "def line_nums():\n",
    "    counts_by_fname = {}\n",
    "    scheme = sameRiver.scheme.scheme('./scheme.xlsx')\n",
    "    total_counts = collections.defaultdict(int)\n",
    "    for bedname in glob.glob('./beds/*bed'):\n",
    "        _ = sum(1 for i in open(bedname, 'rb'))\n",
    "        bname = os.path.basename(bedname).split('.bed')[0]\n",
    "        if bname in scheme.long_basename_to_info:\n",
    "            total_counts[scheme.gene_from_fname(bname)] += _\n",
    "            counts_by_fname[bname] = _\n",
    "    return total_counts, counts_by_fname\n",
    "total_counts, counts_by_fname = line_nums()\n",
    "print(total_counts, '\\n', counts_by_fname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rows = []\n",
    "replicate = collections.defaultdict(int)\n",
    "for k, v in counts_by_fname.items():\n",
    "    prot = k#.split('_')[1]\n",
    "    replicate[prot] += 1\n",
    "    rows.append({\n",
    "        'Protein': prot,\n",
    "        #'Replicate': replicate[prot],\n",
    "        'Unique mapped reads': v\n",
    "    })\n",
    "\n",
    "df = pandas.DataFrame(rows)\n",
    "randos = negative_metadata.random_proteins + ['HCT116', 'CCIN', 'EPB41L5',\n",
    "      'PCBP1-100P', 'PCBP1-100Q', 'PCBP1-dKH', 'SF3B1-K700E']\n",
    "#df = df.loc[[(re.sub(':', '-', x) not in randos) for x in df['Protein']]]\n",
    "df.sort_values(by='Unique mapped reads', inplace=True, ascending=False)\n",
    "print(df)\n",
    "df['Unique mapped reads (millions)'] = df['Unique mapped reads']/1E6\n",
    "sns.set(font_scale=2)\n",
    "sns.set_style('ticks')\n",
    "plt.figure(figsize=(10,10))\n",
    "g = sns.barplot(x='Protein', y='Unique mapped reads (millions)', data=df, color='k')\n",
    "plt.xticks(rotation=90)\n",
    "plt.savefig('../../figs/num_mapped_reads_barchart.pdf')\n",
    "plt.show()\n",
    "plt.clf()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import dill\n",
    "\n",
    "print(\"Loading {}...\".format('data/stats.dill'))\n",
    "with open('data/stats.dill', 'rb') as f:\n",
    "    statsO = dill.load(f)\n",
    "print(\"...Loaded.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clipped_names_rna = {\n",
    "    'ITPA', 'CCIN', 'CDK4', 'TPGS2', 'IDE',\n",
    "    'UBA2', 'EPB41L5', 'DCTN6',\n",
    "    #'CAPNS2', \n",
    "    'CHMP3',  #<- VPS24 RNA name\n",
    "    'ETS2',\n",
    "    #'CMAS', \n",
    "    #'CSRP2',\n",
    "    #'FBL', #'HNRNPC',\n",
    "    #'PCBP1'\n",
    "}\n",
    "clipped_names = clipped_names_rna\n",
    "\n",
    "exons = ['{}::exon'.format(x) for x in clipped_names_rna]\n",
    "\n",
    "# Average counts per protein\n",
    "numeric_cols = statsO.positives.numeric_columns(statsO.positives.counts_per_million_df)\n",
    "\n",
    "print(numeric_cols)\n",
    "#counts = [\n",
    "#    (x, dict(\n",
    "#        [#(name, np.mean(val)) \n",
    "#         (x, x) for x in statsO.negatives.counts_per_million_df.loc[x]]\n",
    "#    )) for x in exons]\n",
    "print(os.getcwd())\n",
    "df = pandas.read_csv('./data/positive_counts_per_read.txt', sep='\\t', index_col=0)\n",
    "numeric_cols = statsO.positives.numeric_columns(df)\n",
    "#print(df)\n",
    "\n",
    "# Subset to the RNAs of interest.\n",
    "counts = []\n",
    "for x in exons:\n",
    "    y = dict(df.loc[x, numeric_cols])\n",
    "    y['gene_name'] = x\n",
    "    counts.append(y)\n",
    "\n",
    "def to_prot(x):\n",
    "    return x.split('_')[1]\n",
    "\n",
    "df = pandas.DataFrame(counts)\n",
    "df.index = df['gene_name']\n",
    "\n",
    "proteins = set([to_prot(x) for x in df.columns])\n",
    "proteins = proteins & set(clipped_names)\n",
    "averaged = []\n",
    "for protein in proteins:\n",
    "    sub = df.loc[:, [x for x in df.columns if to_prot(x)==protein]]\n",
    "    \n",
    "#    print(sub)\n",
    "    x = dict(sub.mean(axis=1))\n",
    "    x['Protein'] = protein\n",
    "    averaged.append(x)\n",
    "df = pandas.DataFrame(averaged).T\n",
    "df.columns = df.loc['Protein']\n",
    "df.fillna(0.)\n",
    "#del df['name']\n",
    "df.drop(['Protein'], inplace=True)\n",
    "print('Post drop')\n",
    "print(df.columns)\n",
    "\n",
    "skippy = \"\"\"\n",
    "pvals = [\n",
    "    (x, dict(\n",
    "        [(name, -np.log10(np.max([1E-6, np.mean(val)])) )  for name, val in statsO.pvals_kde[x].items()]\n",
    "    )) for x in exons]\n",
    "\n",
    "def to_df(counts):\n",
    "    print(counts[0][1].keys())\n",
    "    rows = []\n",
    "    for rna, _dict in counts:\n",
    "\n",
    "        to_del = set()\n",
    "        for protein, pval in _dict.items():\n",
    "            if protein not in clipped_names:\n",
    "                to_del.add(protein)\n",
    "            \n",
    "        rows.append({\n",
    "            'RNA': rna.split('::')[0]\n",
    "        })\n",
    "        \n",
    "        rows[-1].update(_dict)\n",
    "    df = pandas.DataFrame(rows)\n",
    "    df = df[df.columns[[(x not in to_del) for x in df.columns]]]\n",
    "    \n",
    "    \n",
    "    df.index = df.RNA\n",
    "    print('.jsfad;lf,', df)\n",
    "    return df\n",
    "\n",
    "df = to_df(pvals)\n",
    "df = df[df.index]\n",
    "\n",
    "fig = plt.figure()\n",
    "sns.heatmap(df, cmap='Greys')\n",
    "fig.set_figwidth(5)\n",
    "plt.title('-log10(P value)')\n",
    "plt.xlabel('Protein')\n",
    "fig.savefig('../../figs/randos_bind_their_own_rnas_pvalue.pdf')\n",
    "plt.show()\n",
    "plt.clf()\n",
    "for _ in statsO.pvals_kde.items():\n",
    "    print(_)\n",
    "    break\"\"\"\n",
    "\n",
    "#df = to_df(counts)\n",
    "#df = df[df.index]\n",
    "\n",
    "fig = plt.figure()\n",
    "df = df.astype(np.float64)\n",
    "order = df.columns.to_list()\n",
    "print(df)\n",
    "print('--')\n",
    "r_order = sorted(df.index, key=lambda x: order.index(x.split('::')[0]))\n",
    "print(r_order)\n",
    "df = df.reindex(r_order)\n",
    "print(df)\n",
    "df = np.clip(df, a_min=0, a_max=5E3)\n",
    "\n",
    "sns.heatmap(df, cmap='Greys')\n",
    "fig.set_figwidth(5)\n",
    "plt.xlabel('Protein')\n",
    "plt.title('Counts per million')\n",
    "fig.savefig('../../figs/randos_bind_their_own_rnas_counts.pdf')\n",
    "plt.show()\n",
    "plt.clf()\n",
    "#df.replace(np.nan, 0.1, inplace=True)\n",
    "#df.replace(0, 0.1, inplace=True)\n",
    "\n",
    "#m = np.clip(df, a_min=0.1, a_max=1E3)\n",
    "#print(m)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
