{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, re, matplotlib, pandas, collections, importlib, sys\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "\n",
    "import __init__\n",
    "sys.path = __init__.add_paths(sys.path)\n",
    "\n",
    "from importlib import reload\n",
    "import sameRiver\n",
    "from sameRiver import *\n",
    "import pandas as pd\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "beds = '/Users/dfporter/pma/dataAndScripts/clip/miseq/Runs/180815/beds/'\n",
    "scheme = '/Users/dfporter/pma/dataAndScripts/clip/miseq/Runs/180815/scheme.xlsx'\n",
    "\n",
    "top_dir = '/Users/dfporter/pma/dataAndScripts/clip/miseq/all/'\n",
    "importlib.reload(sameRiver.exp)\n",
    "importlib.reload(sameRiver.scheme)\n",
    "ex = sameRiver.exp.exp(name='all', file_paths={\n",
    "    'scheme': top_dir + '/scheme.xls',\n",
    "    'beds': top_dir + '/beds/',\n",
    "    'wigs': top_dir + '/beds/read_start/',\n",
    "    'fastq': top_dir + '/fastq/',\n",
    "    'sams': top_dir + '/sams/',\n",
    "    'STAR index': '/labs/khavari/dfporter/genome/GRCh38.gencode.29/star_index',\n",
    "    'STAR': 'STAR',\n",
    "    'STAR repeats index': '/labs/khavari/dfporter/genome/repeats/star_repeats/',\n",
    "    })\n",
    "\n",
    "ex.read_scheme()\n",
    "print(os.getcwd())\n",
    "\n",
    "if os.path.exists('/Users/dfporter/pma/dataAndScripts/clip/miseq/all/line_numbers.xlsx'):\n",
    "    df = pandas.read_excel('/Users/dfporter/pma/dataAndScripts/clip/miseq/all/line_numbers.xlsx')\n",
    "else:\n",
    "    print(\"Run ex.line_numbers() to generate file.\")\n",
    "\n",
    "for col in ['Gene', 'Replicate']:#, 'Percent XL']:\n",
    "    df[col] = [ex.scheme.long_basename_to_info.get(x, '')[col] for x in df.basename]\n",
    "#df['Percent XL'] = ['{:.3%}'.format(x) for x in df['Percent XL']]\n",
    "df['Total read pairs'] = [x/4 for x in df['r1r2_split line number']]\n",
    "df['Read pairs with insert'] = [x/4 for x in df['r1r2_clipped line number']]\n",
    "df['Mapped, unique reads'] = df['bed line number']\n",
    "df['% adapter dimer'] = ['{:.2%}'.format((tot-inst)/tot) for tot,inst in zip(\n",
    "    df['Total read pairs'], df['Read pairs with insert'])]\n",
    "df['% mapped, unique (of reads with insert)'] = [\n",
    "    '{:.2%}'.format(mapped/inst) for mapped, inst in zip(\n",
    "    df['Mapped, unique reads'], df['Read pairs with insert'])\n",
    "]\n",
    "df = df[['Gene', 'Replicate', 'Total read pairs', 'Read pairs with insert',\n",
    "        'Mapped, unique reads', '% adapter dimer', '% mapped, unique (of reads with insert)']]\n",
    "df['Gene/rep'] = ['{} {}'.format(x, y) for x, y in zip(df.Gene, df.index)]\n",
    "df.sort_values(by='Mapped, unique reads', inplace=True)\n",
    "\n",
    "fig = plt.figure()\n",
    "sns.set_style('ticks')\n",
    "#sns.set_style('whitegrid')\n",
    "sns.barplot(y='Gene/rep', x='Mapped, unique reads', log=True, data=df, color='k')\n",
    "#plt.plot([1E5,1E5], [0, len(df.index)], 'k')\n",
    "fig.set_figheight(7)\n",
    "plt.show()\n",
    "plt.clf()\n",
    "print(df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import six\n",
    "\n",
    "#https://stackoverflow.com/questions/26678467/export-a-pandas-dataframe-as-a-table-image\n",
    "\n",
    "def render_mpl_table(data, col_width=3.0, row_height=0.625, font_size=14,\n",
    "                     header_color='#40466e', row_colors=['#f1f1f2', 'w'], edge_color='w',\n",
    "                     bbox=[0, 0, 1, 1], header_columns=0,\n",
    "                     ax=None, **kwargs):\n",
    "    if ax is None:\n",
    "        size = (np.array(data.shape[::-1]) + np.array([0, 1])) * np.array([col_width, row_height])\n",
    "        fig, ax = plt.subplots(figsize=size)\n",
    "        ax.axis('off')\n",
    "\n",
    "    mpl_table = ax.table(cellText=data.values, bbox=bbox, colLabels=data.columns, **kwargs)\n",
    "\n",
    "    mpl_table.auto_set_font_size(False)\n",
    "    mpl_table.set_fontsize(font_size)\n",
    "\n",
    "    for k, cell in  six.iteritems(mpl_table._cells):\n",
    "        cell.set_edgecolor(edge_color)\n",
    "        if k[0] == 0 or k[1] < header_columns:\n",
    "            cell.set_text_props(weight='bold', color='w')\n",
    "            cell.set_facecolor(header_color)\n",
    "        else:\n",
    "            cell.set_facecolor(row_colors[k[0]%len(row_colors) ])\n",
    "    return ax\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_scheme = sameRiver.scheme.scheme('/Users/dfporter/pma/dataAndScripts/clip/miseq/meta/scheme.xls')\n",
    "\n",
    "df = _scheme.scheme_df\n",
    "df['ID'] = ['_'.join(str(x) for x in [gene, l5, l3]) for gene, l5, l3 in zip(\n",
    "    df.Gene, df.P6_BC, df.P3_BC)]\n",
    "\n",
    "df = df[[('STD' not in _id) for _id in df.ID]]\n",
    "df = df[[('nknown' not in _id) for _id in df.ID]]\n",
    "df = df[[('100' not in _id) for _id in df.ID]]\n",
    "df = df[[('PCBP1' not in _id) for _id in df.ID]]\n",
    "df = df[[('SF3B1' not in _id) for _id in df.ID]]\n",
    "df = df[[('YBX' not in _id) for _id in df.ID]]\n",
    "\n",
    "for col in ['Reads', 'Mapped reads']:\n",
    "    df[col] = np.nan_to_num(df[col].tolist())\n",
    "df.sort_values(by='Gene', inplace=True)\n",
    "df = df[['Gene', 'ID', 'Reads', 'Mapped reads']]\n",
    "df['Reads'] = df.apply(lambda x: \"{:,}\".format(int(x['Reads'])), axis=1)\n",
    "df['Mapped reads'] = df.apply(lambda x: \"{:,}\".format(int(x['Mapped reads'])), axis=1)\n",
    "\n",
    "#??df.style.format({'Reads': \"{:.2%}\", 'Mapped reads': \"{:,}\"})\n",
    "print(df)\n",
    "ax = render_mpl_table(\n",
    "    df[['Gene', 'Reads', 'Mapped reads']].copy(), \n",
    "    header_columns=0, col_width=1., font_size=5, row_height=0.2)\n",
    "\n",
    "df['Reads'] = [int(x.replace(',', '')) for x in df['Reads']]\n",
    "df['Mapped reads'] = [int(x.replace(',', '')) for x in df['Mapped reads']]\n",
    "\n",
    "plt.savefig('/Users/dfporter/pma/dataAndScripts/clip/figs/table_of_read_numbers.pdf')\n",
    "plt.clf()\n",
    "plt.close()\n",
    "print(df.head())\n",
    "\n",
    "sns.set_style('whitegrid')\n",
    "fig = plt.figure()\n",
    "\n",
    "_kwargs = {'palette': sns.cubehelix_palette(2*len(set(df['ID'])), start=0, rot=-0.2)}\n",
    "\n",
    "df['Reads (log 10)'] = np.log10(df['Reads'])\n",
    "sns.barplot(data=df, x='Reads (log 10)', y='ID', **_kwargs)\n",
    "fig.set_figheight(20)\n",
    "plt.show()\n",
    "plt.clf()\n",
    "\n",
    "fig = plt.figure()\n",
    "\n",
    "df['Mapped reads (log 10)'] = np.log10(df['Mapped reads'])\n",
    "df.sort_values(by='Mapped reads', inplace=True, ascending=False)\n",
    "sns.set_style('whitegrid')\n",
    "#sns.set(font_scale=0.7)\n",
    "sns.barplot(data=df, x='Mapped reads (log 10)', y='ID', **_kwargs)\n",
    "\n",
    "fig.set_figheight(12)\n",
    "plt.savefig('/Users/dfporter/pma/dataAndScripts/clip/figs/mapped_reads_log10_barplot.pdf')\n",
    "plt.show()\n",
    "plt.clf()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
