{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import HTSeq, os, importlib, pandas, re, collections, sys\n",
    "from pathlib import Path\n",
    "\n",
    "#os.chdir('/Users/dp/pma/dataAndScripts/clip/miseq/v2all/')\n",
    "sys.path.append('/Users/dp/pma/')\n",
    "import sameRiver\n",
    "import sameRiver.peakFinder\n",
    "import sameRiver.peakSeqFinder\n",
    "import sameRiver.peakSignalFinder\n",
    "\n",
    "importlib.reload(sameRiver.peakFinder)\n",
    "importlib.reload(sameRiver.peakSeqFinder)\n",
    "importlib.reload(sameRiver.peakSignalFinder)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genomic_fasta_fname = '/opt/genomes/repeats_and_ensembl_release94_GRCh38/combined.fa'  # Old, repeats-first mapping\n",
    "#genomic_fasta_fname = '/opt/genomes/gencode.v29/GRCh38.primary_assembly.genome.fa.gz'  # Newer, separate mapping.\n",
    "genomic_fasta = dict( (s[1], s[0]) for s in HTSeq.FastaReader(genomic_fasta_fname, raw_iterator=True) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(genomic_fasta_fname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "importlib.reload(sameRiver.peakSeqFinder)\n",
    "\n",
    "top_dir = '/Users/dp/pma/dataAndScripts/clip/old_mapping_meta/'\n",
    "protein_coding = pandas.read_csv(f'{top_dir}/temp/enst_transcript_id_name_biotype_map.txt', sep='\\t')\n",
    "protein_coding = protein_coding.loc[[x=='protein_coding' for x in protein_coding.transcript_biotype],:]\n",
    "protein_coding = set(protein_coding['gene_name'])\n",
    "\n",
    "def write_fasta(prot, locs, pvals_at_peak):\n",
    "\n",
    "    # This heights file is raw read counts, which we only use to sort by height.\n",
    "    heights = pandas.read_excel(\n",
    "        Path(top_dir, 'data', f'signal_at_{prot}_peak_locations.xlsx'),\n",
    "        index_col=0)\n",
    "    \n",
    "    heights = heights[[x for x in heights.columns if (prot in x)]]\n",
    "\n",
    "    print(\"Getting sequences for each peak...\")\n",
    "\n",
    "    cutoff = 2#0.2  # P value cutoff for signals at peak locations.\n",
    "    \n",
    "    if cutoff < 2:\n",
    "        # Subset peak locations to those with P<cutoff.\n",
    "        if prot not in pvals_at_peak:\n",
    "            print(f\"{prot} not in {Path(top_dir, 'tables', 'pvals_at_peaks_per_read.xlsx')}\")\n",
    "            return\n",
    "        rnas_pass_filter = pvals_at_peak[pvals_at_peak[prot]<cutoff].copy()\n",
    "        print(f'pass pvalue cutoff {cutoff} = {rnas_pass_filter.shape}')\n",
    "        locs_past_filter = locs[[(x in rnas_pass_filter.index) for x in locs.index]]\n",
    "    else:\n",
    "        locs_past_filter = locs\n",
    "        \n",
    "    print('locations that passed p value filter:', locs_past_filter.shape)\n",
    "    \n",
    "    # Remove all ncRNA.\n",
    "    locs_past_filter = locs_past_filter.loc[[x in protein_coding for x in locs_past_filter.index], :]\n",
    "    \n",
    "    pvals_per_read = pandas.read_excel(\n",
    "        Path(top_dir, 'tables', 'pvals_per_read.xlsx'), index_col=0)    \n",
    "    \n",
    "    # P value cutoff for total signal in the given RNA.\n",
    "    cutoff = 2#0.2\n",
    "    \n",
    "    if cutoff < 2:\n",
    "        rnas_pass_filter = pvals_per_read[pvals_per_read[prot]<cutoff].copy()\n",
    "        targets = [x.split('::')[0] for x in rnas_pass_filter.index]\n",
    "\n",
    "        print('pass pvalue cutoff = ', rnas_pass_filter.shape)\n",
    "        locs_past_filter = locs_past_filter[[(x in targets) for x in locs_past_filter.index]]\n",
    "        print('locations that passed p value filter:', locs_past_filter.shape)\n",
    "\n",
    "        heights = heights[[x in locs_past_filter.index for x in heights.index]]\n",
    "\n",
    "    heights['sum'] = heights.sum(axis=1)\n",
    "    to_height = dict(zip(heights.index, heights['sum']))\n",
    "    locs_past_filter['height'] = [to_height.get(x, -100) for x in locs_past_filter.index]\n",
    "\n",
    "    top_n = 10000\n",
    "    min_height = 5#10\n",
    "    \n",
    "    if len(locs_past_filter.index) > top_n:\n",
    "        locs_past_filter.sort_values(by='height', inplace=True, ascending=False)\n",
    "        locs_past_filter = locs_past_filter.head(top_n).copy()\n",
    "        \n",
    "    if min_height:\n",
    "        locs_past_filter = locs_past_filter.loc[[x>=min_height for x in locs_past_filter.height], :]\n",
    "        \n",
    "    print(\"Past height cutoff:\", locs_past_filter.shape )\n",
    "\n",
    "    if locs_past_filter.shape[0] == 0:\n",
    "        return\n",
    "\n",
    "    # Find and write fasta sequences.\n",
    "    a = pkseq.get_seqs(locs_past_filter[[prot]])\n",
    "    pkseq.write_seqs_for_motif_searches(outfolder=f'{top_dir}/data/seqs/', write_controls=True)\n",
    "    \n",
    "locs = pandas.read_excel(f'{top_dir}/data/peak_locations.xlsx', index_col=0)\n",
    "\n",
    "#nb = sameRiver.statsForCountsNB.statsForCountsNB.load()\n",
    "\n",
    "pkseq = sameRiver.peakSeqFinder.peakSeqFinder(fasta_dict=genomic_fasta)\n",
    "pvals_at_peak = pandas.read_excel(\n",
    "    Path(top_dir, 'tables', 'pvals_at_peaks_per_read.xlsx'),\n",
    "    #'/Users/dfporter/pma/dataAndScripts/clip/miseq/v2all/data/pvals_for_signal_at_{}_peaks_per_read.xlsx'.format(prot),\n",
    "    index_col=0)\n",
    "#print(\"...... pvals_at_peak=\", pvals_at_peak)\n",
    "selected_proteins = [\n",
    "    'CELF1',\n",
    "    'PCBP1',\n",
    "    'hnRNPD', \n",
    "    'Rbfox2',\n",
    "    'Rbfox1', \n",
    "    'FBL',\n",
    "    'hnRNPC', \n",
    "    'SF3B1',\n",
    "    'A1CF',\n",
    "    'KHDRBS2',\n",
    "    'FUBP1',\n",
    "#     'RARS2', 'RPL5',\n",
    "    #'UBA2', 'ETS2',\n",
    "#     'CRNKL1',\n",
    "     #'CCIN',\n",
    "]\n",
    "#selected_proteins = ['SF3B1', 'Rbfox2', 'CELF1']\n",
    "for prot in locs.columns:\n",
    "    if prot not in selected_proteins:\n",
    "        continue\n",
    "    print('\\n', prot)\n",
    "    write_fasta(prot, locs, pvals_at_peak)\n",
    "#rna_obj = pickle.load(open('./data/rna.data', 'rb'))\n",
    "homer_command = \"\"\"\n",
    "# For some reason, homer1 works while homer2 doesn't.\n",
    "\n",
    "export PATH=$PATH:/Users/dp/homer/bin/\n",
    "\"\"\"\n",
    "seq_dir = f'{top_dir}/data/seqs/'\n",
    "as_one_cmd = []\n",
    "outdir = './output/'\n",
    "print(f'mkdir {outdir}')\n",
    "\n",
    "for prot in locs.columns:\n",
    "    if prot not in selected_proteins:\n",
    "        continue\n",
    "    cmd = \"{fp} {d}/{g}.fa  fasta {o}/{g} -fasta {d}/randoms_for_{g}.fa\".format(fp='./bin/findMotifs.pl', d=seq_dir, g=prot, o=outdir)\n",
    "    cmd += ' -rna -len 3,4,5,6,7,8,9,12'  # -len 3,4,5,6,8,9,12\n",
    "    print(cmd)\n",
    "    as_one_cmd.append(cmd)\n",
    "    \n",
    "print(f'mkdir {outdir}/; rm {outdir}/*/*; rm {outdir}/*/*/*;' + '; '.join(as_one_cmd))\n",
    "\n",
    "# For Rbfox2, what seems to work is a cutoff of 0.1,\n",
    "# and this homer command ./bin/findMotifs.pl /Users/dfporter/pma/dataAndScripts/clip/miseq/v2all/data/seqs//Rbfox2.fa \\\n",
    "# fasta ./Rbfox2 -fasta /Users/dfporter/pma/dataAndScripts/clip/miseq/v2all/data/seqs//randoms_for_Rbfox2.fa  \\\n",
    "# -rna\n",
    "# Adding -homer1 doesn't generally make a huge difference\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, pandas, importlib\n",
    "import sameRiver\n",
    "import sameRiver.negativeCounts\n",
    "import sameRiver.scheme\n",
    "import sameRiver.annCounts\n",
    "import sameRiver.positiveCounts\n",
    "importlib.reload(sameRiver.annCounts)\n",
    "importlib.reload(sameRiver.negativeCounts)\n",
    "os.chdir('/Users/dfporter/pma/dataAndScripts/clip/miseq/v2all/')\n",
    "top_dir = './'\n",
    "\n",
    "class Namespace:\n",
    "    def __init__(self, **kwargs):\n",
    "        self.__dict__.update(kwargs)\n",
    "        \n",
    "negative_metadata = Namespace(\n",
    "    top_dir = top_dir,\n",
    "    scheme_file_with_random_proteins = top_dir + '/scheme.xlsx',\n",
    "    ann_counts_file = top_dir + '/ann_counts.txt',\n",
    "    bed_file_dir = top_dir + '/beds/',\n",
    "    random_proteins = [\n",
    "        'CAPNS2', #'CCIN',\n",
    "        'CDK4', 'CHMP3',\n",
    "        'DCTN6', #'EPB41L5',\n",
    "        'ETS2', 'IDE',\n",
    "        'ITPA', 'TPGS2', 'UBA2',\n",
    "    ])\n",
    "\n",
    "positive_metadata = Namespace(\n",
    "    top_dir = top_dir,\n",
    "    scheme_file = top_dir + '/scheme.xlsx',\n",
    "    ann_counts_file = top_dir + '/ann_counts.txt',\n",
    "    bed_file_dir = top_dir + '/beds/',\n",
    "    positive_proteins = [\n",
    "        'CAPNS2', 'CCIN', 'CDK4', 'CHMP3',\n",
    "        'DCTN6', 'EPB41L5', 'ETS2', 'IDE',\n",
    "        'ITPA', 'TPGS2', 'UBA2',\n",
    "        'FBL', 'HCT116', 'hnRNPC',\n",
    "        'SF3B1',\n",
    "        'PCBP1', 'PCBP1:100P', 'PCBP1:100Q',\n",
    "        'PCBP1:dKH', 'CELF1',\n",
    "        'Rbfox1', 'Rbfox2', 'hnRNPD',\n",
    "    ]\n",
    ")\n",
    "positives = sameRiver.positiveCounts.positiveCounts(positive_metadata)\n",
    "positives.normalize_ann_counts(use_bed_dir='./beds/', load_total_read_numbers=True)\n",
    "negatives = sameRiver.negativeCounts.negativeCounts(negative_metadata)\n",
    "negatives.normalize_ann_counts(use_bed_dir='./beds/', load_total_read_numbers=True)\n",
    "negatives.get_lowest_values_for_all_proteins()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sameRiver.statsForCountsNB\n",
    "importlib.reload(sameRiver.statsForCountsNB)\n",
    "importlib.reload(sameRiver.scheme)\n",
    "\n",
    "scheme = sameRiver.scheme.scheme(scheme_fname='./scheme.xlsx')\n",
    "a = time.time()\n",
    "nb = sameRiver.statsForCountsNB.statsForCountsNB(negatives=negatives, positives=positives)\n",
    "nb.calculate_pvalues(cutoff_for_consideration=1, test_mode=True)\n",
    "b = time.time()\n",
    "print('time {}m'.format((b - a)/60.))\n",
    "nb.write_pvals_single_file(which='per_read', outfname='./tables/pvals_per_read.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fname = '/Users/dfporter/pma/dataAndScripts/clip/miseq/v2all/data/seqs/randoms_for_Rbfox1.fa'\n",
    "lines = open(fname).readlines()\n",
    "fa = ''.join([\">{}\\n{}\".format(line.split('\\t')[0], line.split('\\t')[1]) for line in lines])\n",
    "with open('/Users/dfporter/pma/dataAndScripts/clip/miseq/v2all/data/seqs/randoms_for_Rbfox1.fa', 'w') as f:\n",
    "    f.write(fa)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\"\"\"\n",
    "Assemble outputs (made in write_fasta_for_homer.ipynb) into one file.\n",
    "\"\"\"\n",
    "\n",
    "import glob\n",
    "from pathlib import Path\n",
    "\n",
    "os.system('export PATH=$PATH:/Users/dp/homer/bin/')\n",
    "print('export PATH=$PATH:/Users/dp/homer/bin/')\n",
    "for fname in glob.glob(top_dir+'/data/seqs/*fa'):\n",
    "    if 'random' in fname:\n",
    "        continue\n",
    "\n",
    "    gene = os.path.basename(fname).split('.fa')[0]\n",
    "    random_fname = os.path.dirname(fname) + '/randoms_for_{0}.fa'.format(gene)\n",
    "    homer = '/Users/dfporter/homer/bin/findMotifs_edit.pl'\n",
    "    cmd = \"{} {} fastq ./{} -fasta {} -rna -homer1\".format(homer, fname, gene, random_fname)\n",
    "    #print(cmd)\n",
    "    #os.system(cmd)\n",
    "    \n",
    "class logoAssembly:\n",
    "    \n",
    "    def __init__(self, homer_dir='./'):\n",
    "        self.homer_dir = homer_dir\n",
    "    \n",
    "    def find_all_svg_motif1_logos(self):\n",
    "        print(f'{self.homer_dir}/*/motif1.logo.svg')\n",
    "        svgs_motif1, svgs_motif2 = {}, {}\n",
    "        for path in glob.glob(f'{self.homer_dir}/*/*/motif1.logo.svg'):\n",
    "            with open(path) as f:\n",
    "                svgs_motif1[os.path.dirname(path)] = f.readlines()\n",
    "                \n",
    "        for path in glob.glob(f'{self.homer_dir}/*/*/motif2.logo.svg'):\n",
    "            print(path)\n",
    "            with open(path) as f:\n",
    "                svgs_motif2[os.path.dirname(path)] = f.readlines()\n",
    "        #print(svgs)\n",
    "        \n",
    "        header = \"\"\"<HTML><HEAD><TITLE>Homer de novo Motif Results</TITLE></HEAD>\n",
    "<BODY>\n",
    "<H1>Homer <i>de novo</i> Motif Results </H1>\n",
    "\"\"\"\n",
    "        print(os.getcwd())\n",
    "        comb = open(self.homer_dir+'/all_logos.html', 'w')\n",
    "        comb.write(header)\n",
    "        for path, lines in svgs_motif1.items():\n",
    "            \n",
    "            comb.write('<H2>{}</H2>\\n'.format(os.path.dirname(path)))\n",
    "            comb.write(''.join(lines))\n",
    "            if path in svgs_motif2:\n",
    "                comb.write(''.join(svgs_motif2[path]))                \n",
    "        comb.close()\n",
    "        \n",
    "logos = logoAssembly(\n",
    "#    homer_dir='/Users/dp/homer/output/'\n",
    "    homer_dir='/Users/dp/homer/top10Ktargets_5_min_height_no_Pvalue_cutoffs_20k_controlSeqs_mRNA_only_old_mapping_eg_repeats_first'\n",
    ")\n",
    "logos.find_all_svg_motif1_logos()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
