{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas, os, sys, re, collections, importlib, glob\n",
    "\n",
    "#sys.path.append('/Users/dfporter/pma/')\n",
    "#import sameRiver\n",
    "#import sameRiver.biotypeLookupFileMaker\n",
    "#from Bio.SeqUtils import MeltingTemp as mt\n",
    "from Bio.Seq import Seq\n",
    "import Bio\n",
    "#import pprint\n",
    "#from pprint import pprint as pp\n",
    "import primerMaker\n",
    "importlib.reload(primerMaker)\n",
    "from primerMaker import primerMaker\n",
    "\n",
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import IUPAC\n",
    "\n",
    "import time\n",
    "import Bio.SeqUtils  # For Bio.SeqUtils.molecular_weight "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "proteins = ['FUBP1', 'HNRNPCL1', 'RPL5', 'TDRKH', 'EEF1B2', 'PCBP1', 'SF3B1', 'PABPC4L', 'YTHDC2', 'BCLAF1', \n",
    "            'KHDRBS2', 'RARS2', 'A1CF', 'DDX50', 'UPF2', 'DCP1B', 'NUFIP1', 'DICER1', 'NOVA1', 'CNOT1', 'RBFOX1', \n",
    "            'KPNB1', 'SRSF2', 'CRNKL1', 'RBM39', 'RBM11', 'U2AF1', 'DDX3X', 'EIF1AX', 'RPL10', 'RQCD1', 'CNOT9',\n",
    "           'MYC', 'CDKN2A', 'BRCA1' , 'OAS1',\n",
    "           'BARD1']\n",
    "print(proteins, len(proteins))\n",
    "wells = pandas.read_excel('Sequencing_schemes_and_results.xlsx', sheet_name='midiprep amounts')\n",
    "prot_to_well = dict(zip(wells.Protein, wells['96 well DNA plate position']))\n",
    "prot_to_well['PABPC4L'] = prot_to_well['PABPCL4']\n",
    "prot_to_well['NOVA1'] = prot_to_well['NOVA']\n",
    "prot_to_well['TDRKH'] = prot_to_well['TDRKH1']\n",
    "#print(wells.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pandas.read_excel('./RBP_high_freq_mutations.xlsx', sheet_name='short_list_cancer')\n",
    "#df = pandas.read_excel('Sequencing_schemes_and_results.xlsx', sheet_name='midiprep amounts')\n",
    "#df['Gene'] = df['Protein']\n",
    "df = df.loc[:, [x for x in df.columns if 'notes' not in x]]\n",
    "df['URL'] = df['URL'].astype('str')\n",
    "df['URL'] = [x.split('/')[-1] for x in df.URL]\n",
    "print(df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# BamHI cut uORF Nt FHH (only see His tag in overlap primer.)\n",
    "\n",
    "left_overlap = 'ACATCATCACCACCATCATGGATCC' # Nt FHH insertion overlap.\n",
    "#left_overlap = 'GGCATAACGACTAATATTTATG' # Ct FHH insertion overlap (BamHI cut)\n",
    "\n",
    "#right_overlap = 'CCACACTGGACTAGTGGATC' # <- as in the primer (antisense)\n",
    "right_overlap = 'GATCCACTAGTCCAGTGTGG' # <- Nt FHH insertion overlap. as in the vector (sense)\n",
    "#right_overlap = 'GATCCAGATTATAAAGATGA'  # Ct FHH insertion overlap (BamHI cut)\n",
    "\n",
    "remove_stop = True # Remove stop codon from CDS if present.\n",
    "\n",
    "print(f\"left o/l length: {len(left_overlap)}, right o/l length: {len(right_overlap)}\")\n",
    "from Bio.Data import CodonTable\n",
    "standard_table = CodonTable.unambiguous_dna_by_id[1]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#https://www.ncbi.nlm.nih.gov/nuccore/BC040523\n",
    "\n",
    "\n",
    "def get_dna(_loc, seq_record) -> Bio.Seq:\n",
    "    seq = seq_record.seq\n",
    "    dna =  seq[_loc._start:_loc._end]#, IUPAC.ambiguous_dna)\n",
    "    if _loc._strand != 1:\n",
    "        dna = dna.reverse_complement()\n",
    "    return dna  # Seq object\n",
    "\n",
    "def get_cds_from_genbank_record(seq_record: Bio.SeqRecord.SeqRecord) -> tuple:\n",
    "\n",
    "        # Did we annotate what to PCR?\n",
    "        highlighted_cds = 0\n",
    "        starts_with = 0\n",
    "        for feature in seq_record.features:\n",
    "            y = feature.__dict__\n",
    "            \n",
    "            if 'qualifiers' in y and 'label' in y['qualifiers']:\n",
    "                \n",
    "                if y['qualifiers']['label'][0] == 'CDS_to_amplify':\n",
    "                    highlighted_cds = feature\n",
    "                    # 0-based.\n",
    "                    if 'starts_with_aa_num' in y['qualifiers']:\n",
    "                        starts_with = int(y['qualifiers']['starts_with_aa_num'][0])\n",
    "                    else:\n",
    "                        starts_with = 0\n",
    "        if highlighted_cds:\n",
    "            cds = highlighted_cds\n",
    "            \n",
    "        # If not, get the first CDS annotated.\n",
    "        else:\n",
    "            try:\n",
    "                cds = [x for x in seq_record.features if x.type=='CDS'][0]\n",
    "            except:\n",
    "                print(\"*\" * 14 + \"No CDS...\")\n",
    "                return\n",
    "        \n",
    "        cds_start = cds.location.start\n",
    "        cds_dna = seq_record.seq[cds.location.start:cds.location.end]\n",
    "        if cds.location.strand != 1:\n",
    "            cds_dna = Seq.reverse_complement(cds_dna)\n",
    "        cds_aa = cds_dna.translate()\n",
    "        \n",
    "        # 'RHKDESTNQCUGPAVILMFYW'\n",
    "        #if not str(cds_aa)[-1].isalpha():\n",
    "        #    cds_aa = cds_aa[:-1]\n",
    "            \n",
    "        return cds_dna, cds_aa, starts_with\n",
    "    \n",
    "def get_primers(\n",
    "    fname: str, aa_wt: str, aa_n: int, aa_mut: str, gene_name: str ='gene_name',\n",
    "    remove_stop: bool = False) -> str: \n",
    "    \n",
    "    for seq_record in SeqIO.parse(fname, \"genbank\"):\n",
    "\n",
    "        print(seq_record.id)\n",
    "        #print(seq_record.features)\n",
    "\n",
    "        # cds_aa: Amino acid sequence.\n",
    "        cds_dna, cds_aa, starts_with = get_cds_from_genbank_record(seq_record)\n",
    "        \n",
    "        if remove_stop and (cds_dna[-3:] in ['TAA', 'TAG', 'TGA']):\n",
    "            #print(f\"Converted {cds_dna} to {cds_dna[:-3]}\")\n",
    "            cds_dna = cds_dna[:-3]\n",
    "            \n",
    "        if remove_stop and (cds_aa[-1] == '*'):\n",
    "            cds_aa = cds_aa[:-1]\n",
    "            \n",
    "        aa_n -= starts_with\n",
    "    \n",
    "        #cds_dna = get_dna(cds.location, seq_record)\n",
    "        #if 'translation' in cds.qualifiers:\n",
    "        #    cds_aa = cds.qualifiers['translation'][0]\n",
    "        #else:\n",
    "        #    cds_aa = cds.translate(cds_dna)\n",
    "        print('aa_n = ', aa_n, ' seq length for cds: ', len(cds_dna))\n",
    "        if aa_n * 3 > len(cds_dna):\n",
    "            print(\"*\" * 14 + \"aa number out of bounds\")\n",
    "            return\n",
    "        \n",
    "        snp_left = (aa_n - 1) * 3\n",
    "        snp_right = aa_n * 3\n",
    "        codon = cds_dna[snp_left:snp_right]\n",
    "        \n",
    "        print(\"Observed codon translation:\", codon.translate(), ' expected ', aa_wt)\n",
    "        \n",
    "        if aa_wt != codon.translate():\n",
    "            print(\"*\" * 14 + \"Observed aa did not match expected\")\n",
    "            return\n",
    "        \n",
    "        product_size_full = len(cds_dna)\n",
    "        product_size_left = len(cds_dna[:snp_right])\n",
    "        product_size_right = len(cds_dna[snp_left:])\n",
    "        \n",
    "        if os.path.exists('outputs/pcr_product_sizes.txt'):\n",
    "            f = open('outputs/pcr_product_sizes.txt', 'a')\n",
    "        else:\n",
    "            f = open('outputs/pcr_product_sizes.txt', 'w')\n",
    "            f.write(\"seqid\\tfull_cds_length\\tleft\\tright\\t-\\t#AA\\tmut\\tkDa\\tseqid\\n\")\n",
    "\n",
    "        f.write(f\"{gene_name}\\t{product_size_full}\\t{product_size_left}\")\n",
    "        f.write(f'\\t{product_size_right}')\n",
    "        f.write(\"\\t-\\t{}\\t{}\".format(product_size_full/3 - 1, f\"{aa_wt}{aa_n}{aa_mut}\"))\n",
    "        f.write(\"\\t{:0.3}\".format(\n",
    "            Bio.SeqUtils.molecular_weight(''.join([x for x in cds_aa if x.isalpha()]), seq_type='protein')/1000))\n",
    "        f.write(f'\\t{gene_name}\\n')\n",
    "        \n",
    "        # Amplification primers for whole cds into\n",
    "        # BamHI cut Nt FHH pLEX uORF.\n",
    "        pm = primerMaker(blocks=[str(cds_dna)], left=left_overlap, right=right_overlap)\n",
    "        pm.pick_primers()\n",
    "        \n",
    "        with open('fasta/{}.fa'.format(gene_name), 'w') as f:\n",
    "            f.write(pm.get_seq_as_fasta())\n",
    "            \n",
    "        lines = pm.one_per_line()\n",
    "        \n",
    "        lines = '\\n'.join([x for x in lines.split('\\n') if 'stit' in x])\n",
    "        bc_acc = re.sub('BC0', '', seq_record.id)\n",
    "        lines = re.sub('_0_', '', lines)\n",
    "        lines = re.sub('leftstitch', 'L_{}'.format(gene_name), lines)\n",
    "        lines = re.sub('rightstitch', 'R_{}'.format(gene_name), lines)\n",
    "\n",
    "        return lines + '\\n'\n",
    "    \n",
    "\n",
    "order = ''\n",
    "genbank_files = glob.glob('gb/*.gb')\n",
    "\n",
    "def parse_mutation(_str, return_no_mutation=False) -> tuple:\n",
    "    a = re.split('(\\d+)', _str)\n",
    "    if return_no_mutation:\n",
    "        return a[0], int(a[1]), a[0]\n",
    "    return a[0], int(a[1]), a[2]\n",
    "\n",
    "\n",
    "def get_fname(gene_name: str) -> str:\n",
    "    \n",
    "    if type(gene_name) != type(''):\n",
    "        return False\n",
    "    \n",
    "    matching_gbs = [x for x in genbank_files if gene_name in str(x)]\n",
    "    if len(matching_gbs):\n",
    "        fname = matching_gbs[0]\n",
    "    else:\n",
    "        fname = 'gb/{}.gb'.format(df.iloc[n]['URL'])\n",
    "        \n",
    "    if not os.path.exists(fname):\n",
    "        print(f\"{gene_name}: No file {fname}\")\n",
    "        return False\n",
    "    \n",
    "    return fname\n",
    "\n",
    "# Read all CDSes from genbank files and write to a single fasta\n",
    "# in case this is handy later for BLASTing against expected \n",
    "# constructs.\n",
    "all_cds = ''\n",
    "cds = {}\n",
    "for n in range(df.shape[0]):\n",
    "    \n",
    "    gene_name = df.iloc[n]['Gene']\n",
    "    fname = get_fname(gene_name)\n",
    "    \n",
    "    print(f\"{gene_name}->{fname}\")\n",
    "    if fname:\n",
    "        print(\"...->\")\n",
    "        for seq_record in SeqIO.parse(fname, \"genbank\"):\n",
    "            cds_dna, cds_aa, starts_with = get_cds_from_genbank_record(seq_record)\n",
    "            all_cds += f\">{gene_name}_{seq_record.id}_{fname}\\n\"\n",
    "            all_cds += f\"{cds_aa}\\n\"\n",
    "            cds[gene_name] = cds_aa\n",
    "            \n",
    "with open('outputs/all_expected_cds_from_genbank_input_files.fa', 'w') as f:\n",
    "    f.write(all_cds)\n",
    "    \n",
    "print('Exported all CDSes from genbank files.')\n",
    "\n",
    "# Clear the file outputs/pcr_product_sizes.txt that is used\n",
    "# for checking PCR lengths and mutations by hand.\n",
    "if os.path.exists('outputs/pcr_product_sizes.txt'):\n",
    "    os.system('rm outputs/pcr_product_sizes.txt')\n",
    "\n",
    "# Get primers for each mutation.\n",
    "done = set()\n",
    "for n in range(df.shape[0]):\n",
    "    print('---=---' * 7)\n",
    "    \n",
    "    gene_name = df.iloc[n]['Gene']\n",
    "    \n",
    "    if gene_name not in proteins:\n",
    "        continue\n",
    "    \n",
    "    fname = get_fname(gene_name)\n",
    "    \n",
    "    if fname in done:\n",
    "        continue\n",
    "    done.add(fname)\n",
    "\n",
    "    print(gene_name, fname)\n",
    "    \n",
    "    if fname:\n",
    "        mut = parse_mutation(\n",
    "            df.iloc[n]['Mutation'])#, return_no_mutation=1)\n",
    "        #if gene_name in cds:\n",
    "        #    midpoint = int(len(cds[gene_name])/2)\n",
    "        #    mut = cds[gene_name][midpoint], midpoint + 1, cds[gene_name][midpoint]\n",
    "        #else:\n",
    "        #    print(f\"No CDS {gene_name}\")\n",
    "        \n",
    "        lines = get_primers(fname, *mut, gene_name=df.iloc[n]['Gene'], remove_stop=remove_stop)\n",
    "        print(lines)\n",
    "        if lines:\n",
    "            order += lines\n",
    "\n",
    "#Write a text order sheet.\n",
    "#Well Position\tName\tSequence\n",
    "#A1\ttest1\tagagagaagagagagag\n",
    "with open('order.txt', 'w') as f:\n",
    "    f.write(order)\n",
    "\n",
    "def well_value(well: str) -> int:\n",
    "    row = 'ABCDEFGHIJKLMNOPQRST'.index(well[0])\n",
    "    value = row * 12 + int(well[1:])\n",
    "    return value\n",
    "\n",
    "# Write excel order sheet.\n",
    "rows = []\n",
    "alphabet = 'ABCDEFGHIJKLMNOPQRST'\n",
    "letter_index = 0\n",
    "number_index = 1\n",
    "for li in order.split('\\n'):\n",
    "    \n",
    "    # For now, skip the odd man out.\n",
    "    #if 'RARS2' in li:\n",
    "    #    continue\n",
    "    #if li.count('_') < 2:\n",
    "    #    continue\n",
    "    if 'RARS2' in li or ('_0L_' not in li and '_1R_' not in li):\n",
    "        if len(li.split('\\t')) < 2:\n",
    "            continue\n",
    "            \n",
    "        gene_name = li.split('\\t')[0].split('_')[-1]\n",
    "        \n",
    "        if gene_name not in prot_to_well:\n",
    "            print(f'Skipping {gene_name}')\n",
    "            continue\n",
    "        \n",
    "        well = prot_to_well[gene_name]\n",
    "        _well_value = well_value(well)\n",
    "        if li.split('\\t')[0][:2] == 'R_':\n",
    "            well = well[0] + str(int(well[1:]) + 1)\n",
    "        #    _well_value += 1\n",
    "        _well_value = well_value(well)\n",
    "        rows.append({\n",
    "            'Well Position': well,#alphabet[letter_index]+str(number_index),\n",
    "            'Name': li.split('\\t')[0],\n",
    "            'Sequence': li.split('\\t')[1].rstrip('\\n'),\n",
    "            'Gene name': li.split('\\t')[0].split('_')[-1],\n",
    "            'sorting': _well_value,\n",
    "        })\n",
    "        \n",
    "        if number_index >= 8:\n",
    "            letter_index += 1\n",
    "            number_index = 1\n",
    "        else:\n",
    "            number_index += 1\n",
    "        if letter_index > 7:\n",
    "            letter_index = 0\n",
    "            number_index = 1\n",
    "            \n",
    "order_df = pandas.DataFrame(rows)\n",
    "order_df.sort_values(by='sorting', inplace=True)\n",
    "order_df.to_excel('order.xlsx', )\n",
    "\n",
    "print(order_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "\"\"\"\n",
    "Charlie8\n",
    "./hg38re/Charlie8\n",
    "stdout:  output//pair1_/Charlie8.txt\n",
    "bowtie2 -k 1 -p 1 --quiet -x ./hg38re/Charlie8 short_multimap_R1.fastq | grep \"repname\" -\n",
    "stdout:  output//pair2_/Charlie8.txt\n",
    "bowtie2 -k 1 -p 1 --quiet -x ./hg38re/Charlie8 short_multimap_R2.fastq | grep \"repname\" -\n",
    "\"\"\"\n",
    "b_opt = \"-k 1 -p 1 --quiet\"\n",
    "file1 = '/Users/dp/pma/dataAndScripts/clip/RepEnrich2/output//pair1_/U8.txt'\n",
    "metagenomepath = './hg38re/U8'\n",
    "fastqfile_1 = 'short_multimap_R1.fastq'\n",
    "\n",
    "with open(file1, 'w') as stdout:\n",
    "    cmd = 'bowtie2 ' + b_opt + ' ' + '-x ' + metagenomepath + ' ' + fastqfile_1 + ' | grep \\\"repname\\\" -'\n",
    "    cmd = ['bowtie2', '-k', '1', '-p', '1', '--quiet', '-x', metagenomepath, fastqfile_1]\n",
    "    #cmd = 'echo hi'\n",
    "    print(cmd)\n",
    "    p = subprocess.check_output(cmd, shell=True)#, stdout=stdout)\n",
    "#    p.communicate()\n",
    "    print(p)\n",
    "stdout.close()\n",
    "with open(file1) as f:\n",
    "    print(f.readlines())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(file1) as f:\n",
    "    print(f.readlines())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
