{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Analysis of all TCGA data for all RBPs.\n",
    "This is data downloaded iteratively from TCGA's API using Amigo RBP term to\n",
    " define RBPs.\n",
    "Analysis ends with\n",
    "mutations_in_rbps_all_tcga_downloaded\n",
    "\n",
    "\"\"\"\n",
    "import pandas, os, sys, re, time, collections, Bio, pprint, random, glob, importlib, json, pickle, matplotlib\n",
    "\n",
    "from Bio.SeqUtils import MeltingTemp as mt\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import proteinLengths\n",
    "import tcgaParser\n",
    "import tcgaAnnotater\n",
    "importlib.reload(tcgaAnnotater)\n",
    "importlib.reload(proteinLengths)\n",
    "importlib.reload(tcgaParser)\n",
    "\n",
    "from typing import List, Mapping, Union\n",
    "\n",
    "pma_dir = '/Users/dp/pma/'\n",
    "\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams['ps.fonttype'] = 42\n",
    "matplotlib.rc('font', **{'family':'sans-serif','sans-serif':['Arial']})\n",
    "\n",
    "lolipop_proteins = [\n",
    "    'SF3B1', 'SMAD4', 'U2AF1', 'SRSF2', 'PCBP1', 'CNOT9',\n",
    "    'RARS2', 'A1CF', 'EEF1B2', 'EIF1AX', 'KHDRBS2', 'FUBP1',\n",
    "    'RBM11', 'HNRNPCL1', 'DDX3X', 'DICER1', 'PABPC4L', 'RPL5',\n",
    "    'CRNKL1', 'RBM39', 'DCP1B', 'RBFOX1', 'TDRKH', 'NOVA1',\n",
    "    'DDX50', 'YTHDC2', 'KPNB1', 'NUFIP1', 'CNOT1']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def just_lolipop(directory):\n",
    "    \n",
    "    out_dir = f\"{directory}/subset_to_lolipop/\"\n",
    "    os.makedirs(out_dir, exist_ok=True)\n",
    "    \n",
    "    for study_fname in glob.glob(f\"{directory}/*txt\"):\n",
    "        outli = []\n",
    "        df = pandas.read_csv(study_fname, sep='\\t', comment='#', dtype='str')\n",
    "        if 'COMMON' not in df.columns:\n",
    "            print(f\"Could not find column COMMON in {study_fname}: {df.head()}\")\n",
    "            continue\n",
    "        df = df.loc[[(x in lolipop_proteins) for x in df.COMMON], :]\n",
    "        df.to_csv(f\"{out_dir}/{os.path.basename(study_fname)}\", sep='\\t', index=False)\n",
    "\n",
    "just_lolipop(\"./all_TCGA_data/Jan2020_census_api_outputs/\")\n",
    "#just_lolipop(\"./all_TCGA_data/Dec2019_api_outputs/\")\n",
    "#just_lolipop(\"./all_TCGA_data/Dec2019_api_outputs/our_rbps/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "importlib.reload(tcgaAnnotater)\n",
    "importlib.reload(proteinLengths)\n",
    "importlib.reload(tcgaParser)\n",
    "\n",
    "#amigo = pandas.read_csv('listsOfRbps/rna_binding_genes_from_amigo.txt', sep='\\t', comment='#')['Gene name'].tolist()\n",
    "\n",
    "protein_lengths = proteinLengths.proteinLengths()\n",
    "protein_lengths = protein_lengths.load()\n",
    "\n",
    "#print(protein_lengths['SMAD4'])\n",
    "\n",
    "# Study descriptions from https://www.cbioportal.org/webservice.do?cmd=getCancerTypes\n",
    "\n",
    "study_info = tcgaParser.currated_set_of_nonredundant_studies(\n",
    "        fname='./all_TCGA_data/currated_set_of_nonredundant_studies_list.txt',\n",
    "        study_desc_fname='cancerLists/tcga_study_ids_and_descriptions.do')\n",
    "\n",
    "a = study_info.study_categories['MSK-IMPACT Clinical Sequencing Cohort (MSKCC, Nat Med 2017)']\n",
    "print(a)\n",
    "print(study_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "census_dl = all_dl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "patient_data_fname = \"all_TCGA_data/msk_impact_2017_clinical_data.tsv\"\n",
    "all_dl = pickle.load(open('outputs/tcgaLoaderObj.p', 'rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def do_all(\n",
    "    nonredundant_study_ids: List, study_info: tcgaParser.currated_set_of_nonredundant_studies) -> tcgaParser.dataLoader:\n",
    "    \n",
    "    set_of_studies = set([\n",
    "        fname for study in nonredundant_study_ids if os.path.exists(\n",
    "            fname := 'all_TCGA_data/Jan2020_census_api_outputs/subset_to_lolipop//{}.txt'.format(study))])\n",
    "    \n",
    "    set_of_studies |= set([\n",
    "        fname for study in nonredundant_study_ids if os.path.exists(\n",
    "            fname := 'all_TCGA_data/Dec2019_api_outputs/our_rbps/subset_to_lolipop/{}.txt'.format(study))])\n",
    "            \n",
    "    dl = tcgaParser.dataLoader(study_info=study_info)\n",
    "\n",
    "    dl.do_a_set_of_studies(set_of_studies)\n",
    "    print(\"Total patients: {}\".format(dl.n_total_patients))\n",
    "    \n",
    "    return dl\n",
    "#print(study_info.nonredundant_study_ids)\n",
    "all_dl = do_all(study_info.nonredundant_study_ids, study_info)\n",
    "\n",
    "all_dl.map_mutations_to_the_number_of_patients_by_study()\n",
    "all_an = tcgaAnnotater.tcgaAnnotater(all_dl)\n",
    "print(study_info.study_id_to_study_name['cllsll_icgc_2011'])\n",
    "#print(study_info.get_study_name_from_study_id('all_TCGA_data/Dec2019_api_outputs/all_rbps_cllsll_icgc_2011.txt'))\n",
    "#print(all_dl.mutation_to_studies)\n",
    "#all_dl.by_study_cat1_n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#all_dl.get_study_cat_from_study_name('')\n",
    "#all_dl.map_mutations_to_the_number_of_patients_by_study()\n",
    "a = study_info.get_study_name_from_study_id('all_TCGA_data/Jan2020_census_api_outputs/census_skcm_broad.txt')\n",
    "print(list(all_dl.by_study_cat1_n.values())[:10])\n",
    "print(a)\n",
    "print('---')\n",
    "print(all_dl.__dict__.keys())\n",
    "\n",
    "print(f\"A1CF=\")\n",
    "print(all_dl.mutation_to_studies[('A1CF', 'E34K')])\n",
    "print(all_dl.mutation_to_studies[('SMAD4', 'R380S')])\n",
    "#study_info.study_categories "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aas = 'ARNDBCEQZGHILKMFPSTWYV'\n",
    "\n",
    "def interpret_mut(protein, mut):\n",
    "    m = re.match('([A-Za-z]+)(\\d+)$', mut)\n",
    "    if m is not None:\n",
    "        out_muts = []\n",
    "        for aa in aas:\n",
    "            out_muts.append((protein, mut + aa))\n",
    "        return out_muts\n",
    "\n",
    "    return [(protein, mut)]\n",
    "\n",
    "studied_rbps = pandas.read_excel('/Users/dp/pma/RBP missense mutations/Sequencing_schemes_and_results.xlsx', sheet_name='Proteins')\n",
    "#print(study_id_to_study_name)\n",
    "\n",
    "def get_mut_and_study_cats_dict() -> List[Mapping[str, Union[str, int]]]:\n",
    "    mut_and_study_cats = []\n",
    "    n = 0\n",
    "    for protein_mut in studied_rbps.Mutations.tolist()[::-1]:\n",
    "        \n",
    "        if pandas.isna(protein_mut):\n",
    "            continue\n",
    "        \n",
    "        n+= 1\n",
    "        \n",
    "        keys = interpret_mut(*protein_mut.split(' '))\n",
    "        mut_and_study_cats.append(all_dl.by_study_cat1_n[keys[0]])\n",
    "        mut_and_study_cats[-1]['Mutation'] = protein_mut\n",
    "\n",
    "        for key in keys[1:]:\n",
    "            results = all_dl.by_study_cat1_n[key]\n",
    "            for cancer, n_patients in results.items():\n",
    "                if cancer == 'Mutation':\n",
    "                    continue\n",
    "                mut_and_study_cats[-1].setdefault(cancer, 0)\n",
    "                mut_and_study_cats[-1][cancer] += n_patients\n",
    "    return mut_and_study_cats\n",
    "\n",
    "\n",
    "def mut_and_study_cats_dict_to_dataframes(_mut_and_study_cats: dict) -> List[pandas.DataFrame]:\n",
    "    \n",
    "    # Convert to dict and re-index by mutation.\n",
    "    mut_and_study_cats = pandas.DataFrame(_mut_and_study_cats)\n",
    "    mut_and_study_cats.index = mut_and_study_cats['Mutation']\n",
    "    del mut_and_study_cats['Mutation']\n",
    "    \n",
    "    # Replace na with 0.\n",
    "    mut_and_study_cats.fillna(0, inplace=True)\n",
    "    \n",
    "    # Sum over columns -> total by mutation.\n",
    "    sums = mut_and_study_cats.sum(axis=1)\n",
    "    \n",
    "    # Remove mutations with 0 sum.\n",
    "    mut_and_study_cats = mut_and_study_cats.loc[[x>0 for x in sums], :]\n",
    "    \n",
    "    # Get the sums again so the index of the sum series matches.\n",
    "    sums = mut_and_study_cats.sum(axis=1)\n",
    "\n",
    "    # Copy the dataframe to make a version by fraction.\n",
    "    mut_and_study_cats_fractions = mut_and_study_cats.copy()\n",
    "\n",
    "    # Make a version by fraction.\n",
    "    for n, index in enumerate(mut_and_study_cats_fractions.index):\n",
    "        mut_and_study_cats_fractions.loc[index, :] = [x/max([1, sums[n]]) for x in mut_and_study_cats_fractions.loc[index, :]]\n",
    "    \n",
    "    # Return the two dataframes.\n",
    "    return [mut_and_study_cats, mut_and_study_cats_fractions]\n",
    "\n",
    "mut_and_study_cats = get_mut_and_study_cats_dict()\n",
    "\n",
    "#with open('./outputs/cancer_type_graph_data_GO_term.json', 'w') as fp:\n",
    "#    json.dump(mut_and_study_cats, fp, indent=4)\n",
    "\n",
    "#with open('./outputs/cancer_type_graph_data_census_only.json', 'r') as fp:\n",
    "#    census = json.load(fp)\n",
    "#    print('census json:')\n",
    "#    print(census)\n",
    "lolipop_only = lambda a_df: a_df.loc[[  any([x in mut for x in lolipop_proteins]) for mut in a_df.index  ], :]\n",
    "\n",
    "[mut_and_study_cats, mut_and_study_cats_fractions] = mut_and_study_cats_dict_to_dataframes(mut_and_study_cats)\n",
    "mut_and_study_cats = lolipop_only(mut_and_study_cats)\n",
    "mut_and_study_cats_fractions = lolipop_only(mut_and_study_cats_fractions)\n",
    "\n",
    "print(mut_and_study_cats.head())\n",
    "#mut_and_study_cats_fractions.loc[(mut_and_study_cats_fractions > 0.2).any(axis=1), :]\n",
    "mut_and_study_cats_fractions = mut_and_study_cats_fractions.loc[:, (mut_and_study_cats_fractions > 0.05).any(axis=0)]\n",
    "print(mut_and_study_cats_fractions.head())\n",
    "\n",
    "cm = sns.clustermap(mut_and_study_cats_fractions, xticklabels=True, yticklabels=True)\n",
    "cm.ax_row_dendrogram.set_visible(False)\n",
    "cm.ax_col_dendrogram.set_visible(False)\n",
    "\n",
    "plt.savefig(f'{pma_dir}/dataAndScripts/clip/figs/heatmap_missense_freq_by_fraction.pdf')\n",
    "plt.show()\n",
    "plt.clf(); plt.close()\n",
    "\n",
    "cm = sns.clustermap(mut_and_study_cats, vmax=3, xticklabels=True, yticklabels=True)\n",
    "cm.ax_row_dendrogram.set_visible(False)\n",
    "cm.ax_col_dendrogram.set_visible(False)\n",
    "\n",
    "plt.savefig(f'{pma_dir}/dataAndScripts/clip/figs/heatmap_missense_freq_by_absolute_count.pdf')\n",
    "plt.show()\n",
    "plt.clf(); plt.close()\n",
    "print(\"Final dataframe plotted:\", mut_and_study_cats)\n",
    "#print(all_dl.by_study_cat1_n[('SF3B1', 'R625C')])\n",
    "#print(all_dl.mutation_to_studies_n[('SF3B1', 'K700E')])\n",
    "#print(all_dl.by_mutation[('SF3B1', 'K700E')])\n",
    "#print(all_dl.n_by_mutation[('SF3B1', 'K700E')])\n",
    "#print(all_an.top_snp_df(all_dl.n_by_mutation).head(60))\n",
    "all_an.top_snp_df(all_dl.n_by_mutation).to_excel('./outputs/top_snps.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(df_by_mutation_scc.head(10))\n",
    "#snp_scc = top_snp_df(n_by_mutation_scc)\n",
    "snp_all = top_snp_df(n_by_mutation_all)\n",
    "print(snp_all.head(40))\n",
    "#snp_scc.head(30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "snp_all = top_snp_df(n_by_mutation_all)\n",
    "print(snp_all.head(20))\n",
    "print('\\n'.join(snp_all['Gene'].tolist()[:20]))\n",
    "#print('\\n'.join(df_all['Gene'].tolist()[:20]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(df_by_mutation_all.head(2))\n",
    "df_by_gene_all.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import scipy as sp\n",
    "import scipy.stats as stats\n",
    "import matplotlib.pyplot as plt\n",
    "import lifelines\n",
    "from lifelines import KaplanMeierFitter\n",
    "from lifelines.statistics import logrank_test\n",
    "\n",
    "kmf = KaplanMeierFitter()\n",
    "\n",
    "def cf_survival(df, col='Overall Survival (Months)'):\n",
    "    dead = df.loc[\n",
    "        [not np.isnan(x) for x in df[col]], :]\n",
    "    res = logrank_test(dead.loc[dead['Mut'], col], dead.loc[~dead['Mut'], col])\n",
    "    #res.print_summary()\n",
    "    return res\n",
    "\n",
    "clinical = pandas.read_csv('./all_TCGA_data/scc_3816_samples_clinical_data.txt', sep='\\t')\n",
    "print(clinical.tail())\n",
    "clinical.index = [sample_id_to_patient_id(str(x)) for x in clinical['Patient ID']]\n",
    "del clinical['Patient ID']\n",
    "\n",
    "clinical = clinical.loc[:, ['Overall Survival (Months)']]\n",
    "#print(clinical.head(2))\n",
    "##print(clinical.shape)\n",
    "# Has patient IDS: by_mutation_scc\n",
    "pvals = []\n",
    "for n in range(100):\n",
    "    row = df_by_gene_all.iloc[n]\n",
    "    #tup = (row['Gene'], row['Variant'])\n",
    "    #print(row)\n",
    "    gene = row['Gene']\n",
    "    if gene in by_gene_scc:\n",
    "        #print('--->', gene)\n",
    "        patients = by_gene_scc[gene]\n",
    "        #print(clinical.loc[patients, :])\n",
    "        \n",
    "        clinical['Mut'] = [bool(x in patients) for x in clinical.index]\n",
    "        \n",
    "        if len(patients) < 3:\n",
    "            continue\n",
    "        res = cf_survival(clinical)\n",
    "        \n",
    "        pvals.append({\n",
    "            'Gene': gene,\n",
    "            #'tup': tup,\n",
    "            'pvalue': res.p_value,\n",
    "            't': res.test_statistic,\n",
    "            'Patients': ','.join(patients)\n",
    "        })\n",
    "        \n",
    "        #print(by_mutation_scc)\n",
    "df = pandas.DataFrame(pvals).sort_values(by='pvalue')\n",
    "print(df)\n",
    "print('\\n'.join(df.loc[df['pvalue']<0.05, 'Gene']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cds = \"\"\"ATG ATG GAA GCC GAT CGA CCT GAA AAA CTT TTT ATT GGG GGT CTC AAT CTC AAG ACG GAT GAG AAG GCC CTC AAA GCA GAG TTC GGA AAG TAT GGC CAC ATT ATC AAG GTC TTT CTT ATG AAA GAC CGC AAG ACG AAC AAA TCT AGG GGT TTC GCA TTC GTA ACA TTT GAG TCC CCA GCC GAC GCG AAG GCA GCG GCG CGC GAT ATG AAC GGG AAA TAC CTC GAT GGC AAG GCG ATT ATG GTT GCG CAG ACT ATC AAG CCG GCC TTC AAA TCC AGT CGA TGG GTG CCG CCG ACC CCT GGC TCA GGT TCA AGA TCC AGG TTC AGC CAC AGG ACC AGA GGT GGA GGA TCC AGT CCG CAA CGC CCC CCA TCA CAG GGG CGG CCG GAC GAC GGT AGA GGA TAT GCA GGG TAT TTC GAT CTC TGG CCG TAT CGC GCC CCT ATG CCC CGC AAA CGC GGT CCC CCT CCA CGA CAC TGG GCA TCT CCA CCT CAT AAA CGG GCC ACG CCT TCC TCA TTG GCC CAT TCT GTA GGG TGC GGT ATG CGA GGA AAG GCC CCG ACA GTA AGT GGT CAA GAT GGA TAT TCA GGA TTG CAG CCC CGA AGG TGG GCA GGT CCG CCC CAT AAA CGA GCA GTC CCG CGC TCC TCT TTG GCC CGG ATC GGT GGA TCA GGC ATG CCA GGG AAG GCC CCA GCA GTG TGG GGG CAG GAT GGT TAC TCA GGA CCC CGA GTA CGA GAG CCA CTT CCG CCC TGC CGA GAC CCC GGC GAC TTT GTT CCA GCA CTG AGG GAT TAT AGC AGG CGC TAT TAC GGC CAT TCA TCT GTC CCC GAC TAT CGG CCG CTT AGG GGG GAT GGC AAC CAA AAC GGT TAT CGG GGT AGA GAC CAT GAG TAC ACA GAC CAT CCA TCA AAA GGG TCA TAT AGG GAG CCG CTG AAA TCC TAC GGT GGG CCT TGT GGG GCG GCG CCT GTA TGG GGC ACT CCT CCA TCC TAT GGC GGT GGG TGT CGC TAC GAG GAA TAT CAA GGG AAT TCC CCA GAT GCA TGC AGT GAG GGG AGG AGT TCC GAA GCT CTC CCA GTA GTT CTG CCT GAT GCC TAT TCA AGA GAC CAC TCC CCT AAG GCA TAT TCA GGC GGC AGA TCC AGT TCT TCC AAT GGA TAT TCA AGG TCA GAC CGC TAC GGG GAG GAG GGT TGT TAT GAA GAA TAT CGC GGA CGA TCT CCG GAC GCA CAC TCT GGA GGC AGG AAC AGC AGC TCA AAT TCA TAC GGT CAG TCT CAC CAT TAC GGT GGC GAA GGT AGG TAT GAG GAG TAC AGA GGC CGG AGC CAC GAA GCC AGA AGC GGG GGA CGA TCT ACC GAC GCC CAT AGT CGG GGT CGG TCC GAC GAC GCC TAC AGC GGT GGC CAC GAT TCA TCA AGT TGG AGT GAC TGT TGC GGG GGT GGA GGA AGG TAC GAG GAG TAT CAA GGG CGG TCA CTG GAT GCG AAT TCA GGC GGT TGC TCC CCC GAA GCA TAT AGT GGC GGG CAC GAC AAC AGC TCA TGG AGC GAC CGA TAT GGT GTA GGG GGA CAC TAT GAA GAG AAT AGA GGT CAC TCC CTG GAT GCG AAC TCA GGA GGT CGG TCT CCC GAT ACT CAT TCA GGA GGT CAC TCA TCA AGT TCA AAC AGT TAT GGC CAA TCT CAT CGC TAC GGT GGG GAG GGT AGA TAC GAA TAC AGG GGT AGA TCT CAT GAT GCG CAT AGT GGT GGA TGC TCC GCA GAC GCC TAC TCA GGT GGC CAC GAC AGT AGT TCA CAG AGC AAC CGG TAT GGG GGC GGT GGT TGC TAT GAA GAG TAT AGA GGT CGG AGT TTG GAC GCA AAT TCT GGG GGT CGG TCC CCA AAT GCC TAC AGT GGC GGT CAC GAT AGT TCC TCC TGG TCT CAC AGG TAC GGC GGA GGC GGC CGC TAC GAA GAG TAC CGG GGC AGG TCA TTG GAC GCC AAC TCT GGC GGT AGG AGC CCA GAC GCT TAC AGT GGC GGG CAC GAT TCC TCT GGA CAG TCA AAC TGC TAT GGA GGA GGT GGG CGC TAT GAG GAG TAC CGA GGC CGC CTT CTG GAT GCG AAC TCA GGT GGT CGG TCT CCA GAT GCG TAT AGT GGT GGA CAC GAT AGT TCC TCA CAA TCC AAC CGG TAT GGG GGA GGA GGG AGG TAT GAA GAG TAC CGA GGA CAC AGT CTC GAC GCT AAC AGT GGA GGG AGA AGT CCC GAT ACC TAC AGC AGG GGC CAT GAC TCA TCT AGT CAG TCT GAC CAT TAC GGC GGC GGA GGG CGA AGC CTT GAT GCA AAC TCT TCA GGA CGA CTT CCA GAT GCC TAT TCT GGA GGA CAC GAT TCT TCA TCA CGC AGC CAT CGC TAT GGC GGG GGC GGA CGC TAC GAG GAG TAC AGG GGA AGG TCA CTT GAT GCG AAT AGC GGG GGA CGG TCC CCG AAT GCG TAT AGC GGA GGA CAC AAT TCA TCT TCT CGG AAT GAT CCT TGC AGG GGT GGC GGA AGA TAT GAG GAG AAT CGC GGT CAC AGT CTG GAC GCT AAT TCA GGG GGG CAC AGC CCG AAC GCG TAC TCA GGG GGA AGG GAC AGC AGC AGC AAC TCA TAC GAT CGA TCA CAT CGG TAC GGG GGG GGT GGT CAC TAC GAG GAA TAC AGA GGT CGA TCC CAT GAC ACT CAT AGC AGG GGG AGG TCA CCA GAC GCG CAT TCT GGT GAT CAT TAT ACT GAA GCA TAC TCT CGA GGG AGG GAC TCC TTT AGC AAC AGC TAC GGC CGC TCC GAC CAC TAT GGA CGA GGC GGT TGT TAT GAA GAA TAT CAG GGT AGA AGC CCT AAC GCA TAT GGG GGG GGC CGC GGC CTC AAT TCT TCA AAC AAT TCA CAT GGG CGG TCT CAC AGG TAT GGG GGA GGG GGC CGG TAT GAA GAG TAT CGA GGT CCG TCA CCA GAC GCA CAT AGC GGG GGA AGG GAC TCC AGC ATA AAG AGT TAC GGA CTG TCT GAT AGA TAT GGT GGC GGC GGA CAT TAT GAG GAG TAT CAA GGT TCC CTG CCC GAT GCC TAT TCT GGG GAT CAC GAC AGA TCC TCT AAC AGT TAT GGT CGG TCC GAC CGC TAT TCA AGA GGA CGG GAT CGC GTA GGC AGG CCA GAC AGG GGC TTG CCT CTG CCT ATG GAG ACA GGA TCT CCG CCG TTG CAC GAC AGT TAC TCT AGG AGT GGG TGC CGG GTC CCG AGG GGT GGT GGT AGG CAG GGA GGC AGA TTC GAG AGA GGG GAG GGC CGA TCC CGG TAC \"\"\"\n",
    "cds = cds.split(' ')\n",
    "cds = ''.join(cds)\n",
    "\n",
    "table = {\n",
    "    \"TTT\": \"F\", \"TTC\": \"F\", \"TTA\": \"L\", \"TTG\": \"L\", \n",
    "    \"TCT\": \"S\", \"TCC\": \"S\", \"TCA\": \"S\", \"TCG\": \"S\", \n",
    "    \"TAT\": \"Y\", \"TAC\": \"Y\", \"TGT\": \"C\", \"TGC\": \"C\", \n",
    "    \"TGG\": \"W\", \"CTT\": \"L\", \"CTC\": \"L\", \"CTA\": \"L\", \n",
    "    \"CTG\": \"L\", \"CCT\": \"P\", \"CCC\": \"P\", \"CCA\": \"P\", \n",
    "    \"CCG\": \"P\", \"CAT\": \"H\", \"CAC\": \"H\", \"CAA\": \"Q\", \n",
    "    \"CAG\": \"Q\", \"CGT\": \"R\", \"CGC\": \"R\", \"CGA\": \"R\", \n",
    "    \"CGG\": \"R\", \"ATT\": \"I\", \"ATC\": \"I\", \"ATA\": \"I\", \n",
    "    \"ATG\": \"M\", \"ACT\": \"T\", \"ACC\": \"T\", \"ACA\": \"T\", \n",
    "    \"ACG\": \"T\", \"AAT\": \"N\", \"AAC\": \"N\", \"AAA\": \"K\", \n",
    "    \"AAG\": \"K\", \"AGT\": \"S\", \"AGC\": \"S\", \"AGA\": \"R\", \n",
    "    \"AGG\": \"R\", \"GTT\": \"V\", \"GTC\": \"V\", \"GTA\": \"V\", \n",
    "    \"GTG\": \"V\", \"GCT\": \"A\", \"GCC\": \"A\", \"GCA\": \"A\", \n",
    "    \"GCG\": \"A\", \"GAT\": \"D\", \"GAC\": \"D\", \"GAA\": \"E\", \n",
    "    \"GAG\": \"E\", \"GGT\": \"G\", \"GGC\": \"G\", \"GGA\": \"G\", \n",
    "    \"GGG\": \"G\"}\n",
    "rev_translate = collections.defaultdict(list)\n",
    "for codon, aa in table.items():\n",
    "    rev_translate[aa].append(codon)\n",
    "import random\n",
    "def get_random_codon(aa):\n",
    "    return random.choice(rev_translate[aa])\n",
    "    \n",
    "seq = Bio.Seq.Seq(cds)\n",
    "print(seq)\n",
    "print(seq.translate())\n",
    "ct = Bio.Data.CodonTable.CodonTable\n",
    "\n",
    "from Bio.Data import CodonTable\n",
    "standard_table = CodonTable.unambiguous_dna_by_id[1]\n",
    "print(standard_table)\n",
    "standard_table.back_table['H']\n",
    "\n",
    "randomized_codons = []\n",
    "for n_codon in range(0, len(seq) - 3, 3):\n",
    "    #print(n_codon, seq[n_codon:n_codon+3])\n",
    "    codon = seq[n_codon:n_codon+3]\n",
    "    aa = codon.translate()\n",
    "    #print(aa)\n",
    "    #print(standard_table.back_table[aa])\n",
    "    #print(get_random_codon(aa))\n",
    "    #if random.randint(0, 100) == 1:\n",
    "    #    break\n",
    "    randomized_codons.append(get_random_codon(aa))\n",
    "    \n",
    "    if len(randomized_codons) >= 550:\n",
    "        break\n",
    "randomized_codons = ''.join(randomized_codons)\n",
    "print(randomized_codons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
