{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas, os, sys, re, time, collections, Bio, pprint, random\n",
    "\n",
    "sys.path.append('/Users/dp/pma/')\n",
    "import sameRiver\n",
    "import sameRiver.biotypeLookupFileMaker\n",
    "\n",
    "from Bio.SeqUtils import MeltingTemp as mt\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "from pprint import pprint as pp\n",
    "from functools import reduce\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pandas.read_csv('uniprot_id_to_gene_name.tab', sep='\\t', index_col=False)\n",
    "df.columns = ['Uniprot', 'Gene']\n",
    "uniprot_to_gene_name = dict(zip([x.upper() for x in df.Uniprot], [x.upper() for x in df.Gene]))\n",
    "amigo = pandas.read_csv('listsOfRbps/rna_binding_genes_from_amigo.txt', sep='\\t', comment='#')['Gene name'].tolist()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_nonrbps():\n",
    "    nonrbps = ['CDKN2A', 'HIST1H1C', 'HLA-A', 'HLA-B']\n",
    "    return []#nonrbps\n",
    "\n",
    "def remove_nonrbps(x):\n",
    "    nonrbps = get_nonrbps()\n",
    "    return [_ for _ in x if _ not in nonrbps]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get protein lengths.\n",
    "\n",
    "protein_length = collections.defaultdict(int)\n",
    "li = ''\n",
    "\n",
    "for seq_record in SeqIO.parse('./uniprot-proteome%3AUP000005640.fasta', \"fasta\"):\n",
    "    m = re.search(' GN=(\\w+) ', seq_record.description)\n",
    "    gene = seq_record.id.split('|')[-1].split('_')[0].upper()\n",
    "\n",
    "    if m is not None:\n",
    "        gene = m.group(1)\n",
    "\n",
    "    if 'Fragment' in seq_record.id:\n",
    "        continue\n",
    "    if gene in uniprot_to_gene_name:\n",
    "        protein_length[uniprot_to_gene_name[gene]] = max([\n",
    "            protein_length[uniprot_to_gene_name[gene]], len(seq_record.seq)])\n",
    "        protein_length[gene] = max([protein_length[uniprot_to_gene_name[gene]], len(seq_record.seq)])\n",
    "    else:\n",
    "        protein_length[gene] = max([protein_length[gene], len(seq_record.seq)])\n",
    "\n",
    "    li += gene + '\\n'\n",
    "    \n",
    "with open('uniprot_gene_ids.txt', 'w') as f:\n",
    "    f.write(li)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "bailey = pandas.read_excel('cancerLists/bailey_mh_cell_cancerDrivers.xlsx', sheet_name='Table S1')\n",
    "bailey_amigo = bailey.loc[[bool(x in amigo) for x in bailey.Gene], :]\n",
    "bailey_scc = bailey_amigo.loc[bailey['Cancer']=='PANCAN', :]\n",
    "bailey_amigo_uniq = bailey_amigo.drop_duplicates('Gene')\n",
    "print(bailey_amigo_uniq.shape)\n",
    "print('\\n'.join(bailey_scc['Gene'].tolist()))\n",
    "#print(bailey_amigo_uniq)\n",
    "#print(bailey_amigo)\n",
    "bailey_amigo_uniq.to_excel('outputs/bailey_uniq_amigo.xlsx')\n",
    "bailey_amigo.to_excel('outputs/bailey_amigo.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "oncokb = pandas.read_csv('cancerLists/cancerGeneList.txt', sep='\\t')\n",
    "print(oncokb.head(1))\n",
    "print(oncokb.shape)\n",
    "oncokb_amigo = oncokb.loc[[bool(x in amigo) for x in oncokb['Hugo Symbol']], :]\n",
    "oncokb_amigo.sort_index(by='# of occurrence within resources (Column D-J)', ascending=False, inplace=True)\n",
    "print(oncokb_amigo.shape)\n",
    "print(oncokb_amigo.head(2))\n",
    "oncokb_amigo.to_excel('outputs/oncokb_amigo.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Single cell SCC sequencing data (Andrew Ji)\n",
    "sc_tss_vs_norm_basal = pandas.read_excel('cancerLists/pt7new_de_tss_0.01.xlsx')\n",
    "print(sc_tss_vs_norm_basal.head(1))\n",
    "sc_tss_vs_norm_basal_amigo = sc_tss_vs_norm_basal.loc[[bool(x in amigo) for x in sc_tss_vs_norm_basal.Gene], :]\n",
    "df = sc_tss_vs_norm_basal_amigo\n",
    "df = df.loc[df['p_val']< 0.01, :]\n",
    "df = df.loc[[abs(x)>=0.7 for x in df['avg_logFC']], :]\n",
    "df.sort_index(by='avg_logFC', inplace=True)\n",
    "sc_tss_vs_norm_basal_amigo = df\n",
    "print(sc_tss_vs_norm_basal_amigo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hotspots = pandas.read_csv('cancerLists/hotspots.txt', sep='\\t')\n",
    "print(hotspots.head(1))\n",
    "print(hotspots.shape)\n",
    "hotspots_amigo = hotspots.loc[[bool(x in amigo) for x in hotspots['Gene']], :]\n",
    "print(hotspots_amigo.shape)\n",
    "hotspots_amigo.drop_duplicates('Gene', inplace=True,)\n",
    "print(len(set(hotspots_amigo['Gene'])))\n",
    "print(hotspots_amigo)\n",
    "print('\\n'.join(hotspots_amigo['Gene'].tolist()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "cosmic = pandas.read_csv('cancerLists/Census_allThu May 30 21_44_43 2019.tsv', sep='\\t')\n",
    "cosmic['SCC?'] = [\n",
    "    {True: 'Y', False: 'N'}[bool(re.search('SCC', str(x)) is not None)] for x in cosmic['Tumour Types(Somatic)']]\n",
    "#print([re.search('SCC', str(x) is not None) for x in cosmic['Tumour Types(Somatic)']])\n",
    "#print(cosmic.head())\n",
    "print(cosmic.shape)\n",
    "cosmic_amigo = cosmic.loc[[bool(x in amigo) for x in cosmic['Gene Symbol']], :]\n",
    "print(cosmic_amigo.shape)\n",
    "scc_cos_amigo = cosmic_amigo.loc[[bool(x=='Y') for x in cosmic_amigo['SCC?']], :]\n",
    "print(scc_cos_amigo)\n",
    "cosmic_amigo.to_excel('outputs/cosmic_amigo.xlsx')\n",
    "print(cosmic_amigo.head(2))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "skin_tcga = pandas.read_csv('Mutated_Genes.tsv', sep='\\t')\n",
    "skin_tcga['length'] = [protein_length.get(x, 1E6) for x in skin_tcga['Gene']]\n",
    "skin_tcga = skin_tcga.loc[skin_tcga['length']>0, :]\n",
    "skin_tcga['Patients/length'] = [1000*x/y for x, y in zip(skin_tcga['#'], skin_tcga.length)]\n",
    "skin_tcga.sort_values(by='Patients/length', ascending=False, inplace=True)\n",
    "print(skin_tcga.head(2))\n",
    "print(skin_tcga.shape)\n",
    "skin_tcga_amigo = skin_tcga.loc[[bool(x in amigo) for x in skin_tcga.Gene], :]\n",
    "\n",
    "\n",
    "print(skin_tcga_amigo.shape)\n",
    "print(skin_tcga_amigo.head(10))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "exomes = pandas.read_excel('cancerLists/SCC exomes and targeted.100 integrated exomes.xlsx')\n",
    "exomes['Gene'] = [str(x).upper() for x in exomes.Gene]\n",
    "exomes['length'] = [protein_length.get(x, 1E6) for x in exomes['Gene']]\n",
    "exomes = exomes.loc[exomes['length']>0, :]\n",
    "exomes['Patients/length'] = [1000*x/y for x, y in zip(exomes['# of Patients Mutated'], exomes.length)]\n",
    "exomes.sort_values(by='Patients/length', ascending=False, inplace=True)\n",
    "exomes.index = range(0, exomes.shape[0])\n",
    "print(exomes.head(1))\n",
    "print(exomes.shape)\n",
    "\n",
    "exome_rbps = exomes.loc[[bool(x in amigo) for x in exomes['Gene']], :]\n",
    "print(exome_rbps.shape)\n",
    "#print(exome_miss)4\n",
    "exome_rbps['Patient codes'] = exome_rbps['Patient codes'].astype(str)\n",
    "patients = reduce(lambda a, b: a + '_' + b, exome_rbps['Patient codes'].tolist())\n",
    "patients = set(patients.split('_')) - set([''])\n",
    "n_patients = len(patients)\n",
    "print(n_patients)\n",
    "print(exome_rbps.loc[:, ['Gene', #'# of Genes Mutated', \n",
    "               '# of Patients Mutated', 'length', 'Patients/length']])\n",
    "# First with any real chance of being an RBP: HELZ(18 patients)\n",
    "# ADAD1: adenosine deamindase. Normally only expressed in testes. Occasionally expressed in some other\n",
    "# cancers, but typically not in a mutated form.\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "arr_len = np.array(exomes['Patients/length'])\n",
    "arr_len = arr_len[arr_len<1E6]\n",
    "plt.hist(arr_len, bins=50)\n",
    "plt.title('All genes')\n",
    "plt.show()\n",
    "plt.clf()\n",
    "plt.hist(exome_rbps['Patients/length'].tolist(), bins=50)\n",
    "plt.title('RBPs')\n",
    "plt.show()\n",
    "plt.clf()\n",
    "print(np.median(arr_len), np.median(exome_rbps['Patients/length']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import tcgaParser\n",
    "import tcgaAnnotater\n",
    "\n",
    "def do_all():\n",
    "\n",
    "    # Read studies\n",
    "    studies = []\n",
    "    by_mutation = collections.defaultdict(set)\n",
    "    \n",
    "    set_of_studies = set()\n",
    "    for fname in glob.glob('./all_TCGA_data/Jan2020_census_api_outputs/'):\n",
    "        set_of_studies.add(fname)\n",
    "        \n",
    "    dl = tcgaParser.dataLoader()\n",
    "    dl.do_a_set_of_studies(set_of_studies)\n",
    "    return dl\n",
    "\n",
    "#all_dl = do_all()\n",
    "#all_an = tcgaAnnotater.tcgaAnnotater(all_dl)\n",
    "import numpy as np\n",
    "class tcgaConvolve():\n",
    "\n",
    "    @staticmethod\n",
    "    def smooth(y, box_pts):\n",
    "        #https://stackoverflow.com/questions/20618804/how-to-smooth-a-curve-in-the-right-way\n",
    "        box = np.ones(box_pts)/box_pts\n",
    "        y_smooth = np.convolve(y, box, mode='same')\n",
    "        return y_smooth\n",
    "\n",
    "    def for_each_mut(self, by_mutation, box_width=1):\n",
    "        gene_array = {}\n",
    "        for (gene, mut), num_patients in by_mutation.items():\n",
    "            #if gene != 'TENT4B':\n",
    "            #    continue\n",
    "            if '*' in mut:\n",
    "                continue\n",
    "            if '?' in mut:\n",
    "                continue\n",
    "            pos = re.search('\\w(\\d+)\\w', mut)\n",
    "\n",
    "            if pos is None:\n",
    "                continue\n",
    "                \n",
    "            pos = int(pos.group(1)) - 1\n",
    "            if pos < 0:\n",
    "                print(gene, mut, pos)\n",
    "            if gene not in gene_array:\n",
    "                gene_array[gene] = np.zeros(pos)\n",
    "            if pos >= len(gene_array[gene]):\n",
    "                gene_array[gene] = np.append(\n",
    "                    gene_array[gene], np.zeros(1+pos-len(gene_array[gene])))\n",
    "            gene_array[gene][pos] += len(num_patients)\n",
    "            \n",
    "        smoothed = {}\n",
    "        self.max_positions = {}\n",
    "        self.val_at_max = {}\n",
    "        \n",
    "        for gene in gene_array:\n",
    "\n",
    "            smoothed[gene] = self.smooth(gene_array[gene], box_width)\n",
    "\n",
    "            self.max_positions[gene] = self.highest_point(smoothed[gene])\n",
    "\n",
    "            if self.max_positions[gene] is None:\n",
    "                self.val_at_max[gene] = 0\n",
    "            else:\n",
    "                self.val_at_max[gene] = box_width * smoothed[gene][self.max_positions[gene]]\n",
    "\n",
    "        #print(gene_array)\n",
    "        #print(smoothed)\n",
    "        \n",
    "    def highest_point(self, arr):\n",
    "\n",
    "        if len(arr) != 0:\n",
    "            _max = np.max(arr)\n",
    "        else:\n",
    "            _max = 0\n",
    "            return None\n",
    "\n",
    "        indexes_of_maxes = []\n",
    "        for n, v in enumerate(arr):\n",
    "            if v == _max:\n",
    "                indexes_of_maxes.append(n)\n",
    "\n",
    "        if len(indexes_of_maxes) > 1:\n",
    "            consecutive = []\n",
    "            for n, index in enumerate(indexes_of_maxes):\n",
    "                if n == 0:\n",
    "                    consecutive.append(index)\n",
    "                    continue\n",
    "                if index <= (consecutive[n-1]+3):  # Allow 2 nt gaps.\n",
    "                    consecutive.append(index)\n",
    "                else:  # Reject if there are nonadjacent maxima\n",
    "                    return None\n",
    "            # Take the peak as the middle point.\n",
    "            max_index = int((consecutive[0] + consecutive[-1])/2)\n",
    "\n",
    "#            max_index = random.choice(indexes_of_maxes)\n",
    "        else:\n",
    "            max_index = indexes_of_maxes[0]\n",
    "\n",
    "        \n",
    "        return max_index\n",
    "\n",
    "# total patients = 35313\n",
    "n_total_patients = 35313\n",
    "\n",
    "width_1_convolver = tcgaConvolve()\n",
    "width_1_convolver.for_each_mut(all_dl.by_mutation, box_width=1)\n",
    "\n",
    "width_n_convolver = tcgaConvolve()\n",
    "width_n_convolver.for_each_mut(all_dl.by_mutation, box_width=5)\n",
    "\n",
    "convolver = width_n_convolver\n",
    "print(convolver.max_positions['EIF1AX'])\n",
    "print(convolver.val_at_max['YTHDC2'])\n",
    "\n",
    "cutoff = 1.8\n",
    "\n",
    "above = {}\n",
    "for gene in convolver.val_at_max.keys():\n",
    "    if bool(convolver.val_at_max[gene]>=cutoff):\n",
    "        above[gene] = convolver.val_at_max[gene]\n",
    "\n",
    "for k in ['RBFOX1', 'DDX3X', 'MSI2', 'CELF1', 'SRSF1', 'SRSF2', 'EIF4H', 'TIAL1']:\n",
    "    print(k, convolver.val_at_max[k])\n",
    "sorted_above = sorted(above.keys(), key=lambda x: above[x])\n",
    "print('{} above {}'.format(len(above), cutoff))\n",
    "for k in sorted_above[-100:]:\n",
    "    print(k, above[k], convolver.max_positions[k])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_exomes = pandas.read_excel('cancerLists/updated_100WES_for_cari.xlsx')\n",
    "all_exomes.index = all_exomes['Gene'].tolist()\n",
    "exomes = all_exomes.loc[[bool(x in amigo) for x in df.Gene], :]\n",
    "\n",
    "print(exomes.head(2))\n",
    "exomes.sort_values(by='mutations_normalized', ascending=False, inplace=True)\n",
    "ex_cols = ['Gene', 'Non/Syn', 'NumberOfSCCs_filtered', 'mutations_normalized'] \n",
    "print(exomes.head(1))\n",
    "print(exomes[ex_cols].head(20))\n",
    "non_over_syn = exomes.loc[[x>=3. for x in exomes['Non/Syn']], :]\n",
    "print(non_over_syn[ex_cols].head(20))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pandas.read_excel('RBP_high_freq_mutations.xlsx')\n",
    "df.index = zip(df['Gene'].tolist(), df.Mutation)\n",
    "scc_genes = set(all_exomes.index)\n",
    "\n",
    "def add_if_there(x, col):\n",
    "    if x in scc_genes:\n",
    "        return all_exomes.loc[x, col]\n",
    "    else:\n",
    "        return ''\n",
    "import numpy as np\n",
    "\n",
    "def yes_if_in(name, a_set):\n",
    "    if name in a_set:\n",
    "        return 'Yes'\n",
    "    return 'No'\n",
    "def expressed(x):\n",
    "    if x not in scc_genes:\n",
    "        return 'Not found'\n",
    "    vals = all_exomes.loc[x, ['day_0', 'day_3', 'day_6']]\n",
    "    vals = np.nan_to_num(vals)\n",
    "    return np.average(vals)\n",
    "\n",
    "df['cSCC Nonsyn./Synom.'] = [add_if_there(x, 'Non/Syn') for x in df.Gene]\n",
    "df['cSCC # Patients/100 with ANY non_synon. mutation in gene'] = [\n",
    "    add_if_there(x, 'NumberOfSCCs_filtered') for x in df.Gene]\n",
    "df['cSCC expression'] = [expressed(x) for x in df.Gene]\n",
    "\n",
    "df['# TCGA patients (out of 35313) with ANY non-truncating mutation at a.a. X'] = [\n",
    "    width_1_convolver.val_at_max.get(x, '') for x in df.Gene\n",
    "]\n",
    "df['a.a. X'] = [\n",
    "    width_1_convolver.max_positions.get(x, '') for x in df.Gene\n",
    "]\n",
    "def add_one(val):\n",
    "    try:\n",
    "        return val + 1\n",
    "    except:\n",
    "        return ''\n",
    "df['a.a. X'] = [add_one(x) for x in df['a.a. X']]\n",
    "df['Convolved # patients (out of 35313) with ANY non-truncating mutation near a.a. Y'] = [\n",
    "    width_n_convolver.val_at_max.get(x, '') for x in df.Gene\n",
    "]\n",
    "df['a.a. Y'] = [\n",
    "    add_one(width_n_convolver.max_positions.get(x, '')) for x in df.Gene\n",
    "]\n",
    "df['COSMIC cancer gene?'] = [yes_if_in(x, cosmic['Gene Symbol'].tolist()) for x in df.Gene]\n",
    "df['MH Bailey cancer driver?'] = [yes_if_in(x, bailey['Gene'].tolist()) for x in df.Gene]\n",
    "df['OncoKB cancer gene?'] = [yes_if_in(x, oncokb['Hugo Symbol'].tolist()) for x in df.Gene]\n",
    "df['Cancer mutation hotspot?'] = [yes_if_in(x, hotspots['Gene'].tolist()) for x in df.Gene]\n",
    "df['Amigo RBP?'] = [yes_if_in(x, amigo) for x in df.Gene]\n",
    "\n",
    "cols_to_remove = [\n",
    "    '# cases in all TCGA', \n",
    "    '# cases with ANY nonsyn mutation in this gene in 100 exomes cSCC Lee',\n",
    "    #'Amigo RBP?', \n",
    "    'URL', 'notes', 'notes2', 'notes3', 'MGC with CDS',\n",
    "    #'Ordered or have already',\n",
    "] + [x for x in df.columns if 'Unnamed' in x]\n",
    "for col in set(cols_to_remove):\n",
    "    if col in df.columns:\n",
    "        del df[col]\n",
    "\n",
    "col_explanation = {\n",
    "    '# cases/10,161': 'From original TCGA analysis by Amin',\n",
    "    'Mutation rank/1000': 'From original TCGA analysis by Amin',\n",
    "    'Old dataset?': 'Rationale for inclusion includes the TCGA analysis by Amin',\n",
    "    'Rationale': 'Reason for inclusion in this list',\n",
    "    'Amigo RBP?': 'Amigo GO term for RNA-binding',\n",
    "    'Polyphen summary': 'Predicted mutation impact',\n",
    "    'Polyphen score': 'Higher number means a negative predicted impact on protein',\n",
    "    'Ordered or have already': 'Whether this protein will be included in our CLIP study',\n",
    "    'cSCC Nonsyn./Synom': 'From Cari Lee, the ratio of nonsynon/synonymous mutations in cSCC',\n",
    "    'cSCC # Patients/100 with ANY non_synon. mutation in gene': 'From Cari Lee data',\n",
    "    'cSCC expression': \"From Lee lab data\",\n",
    "    '# TCGA patients (out of 35313) with ANY non-truncating mutation at a.a. X': \"I downloaded data for every RBP (by Amigo term) for every TCGA dataset for this count.\",\n",
    "    \"a.a. X\": \"Amino acid position X\",\n",
    "    'Convolved # patients (out of 35313) with ANY non-truncating mutation near a.a. Y': 'Box convolution with width 5 nt for missense mutations. This leads to a peak at a.a. Y, and the total number of mutated patients in the 5 nt around Y in this column.',\n",
    "    'a.a. Y': \"Amino acid position Y\",\n",
    "    'COSMIC cancer gene?': 'In COSMIC for any cancer type.',\n",
    "    'MH Bailey cancer driver?': \"In M.H. Bailey, 'Comprehensive Characterization of Cancer Driver Genes and Mutations.' (2018) as a cancer driver\",\n",
    "    'OncoKB cancer gene?': 'In OncoKB for any cancer',\n",
    "    'Cancer mutation hotspot?': 'In http://www.cancerhotspots.org (Memorial Sloan Kettering)',\n",
    "}\n",
    "col_explanation_df = pandas.DataFrame.from_dict(col_explanation, orient='index')\n",
    "col_explanation_df.columns = ['Description']\n",
    "col_explanation_df.index.name = 'Column'\n",
    "print(col_explanation_df.head(1))\n",
    "import xlsxwriter\n",
    "writer = pandas.ExcelWriter('outputs/Recurrently_mutated_RBPs.xlsx', engine='xlsxwriter')\n",
    "\n",
    "# Convert the dataframe to an XlsxWriter Excel object.\n",
    "df.to_excel(writer, sheet_name='Mutations', index=False)\n",
    "col_explanation_df.to_excel(writer, sheet_name='Column explanations', index=True)\n",
    "writer.save()\n",
    "\n",
    "#df.to_excel('outputs/Recurrently_mutated_RBPs.xlsx', index=False)\n",
    "\n",
    "print(df.head(2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
