{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re, os, glob, pandas, importlib, collections, pickle, matplotlib\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "import proteinLengths\n",
    "import tcgaParser\n",
    "import tcgaAnnotater\n",
    "importlib.reload(tcgaAnnotater)\n",
    "importlib.reload(proteinLengths)\n",
    "importlib.reload(tcgaParser)\n",
    "\n",
    "\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams['ps.fonttype'] = 42\n",
    "matplotlib.rc('font',**{'family':'sans-serif','sans-serif':['Arial']})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def insert_newlines_into_svg_files() -> None:\n",
    "    # Make reformatted svg files by inserting some newlines.\n",
    "    newline_before = ['<g>', '</g>', '<text', '<circle', '<rect', '<line']\n",
    "    for fname in glob.glob('figs/unedited_lolipop/*.svg'):\n",
    "        #if 'BRCA1' not in fname:\n",
    "        #    continue\n",
    "        outlines = []\n",
    "        with open(fname) as fh:\n",
    "            print(fname)\n",
    "            for line in fh:\n",
    "                #print(line)\n",
    "                for pattern in newline_before:\n",
    "                    line = re.sub(pattern, '\\n' + pattern, line)#'\\n'.join(line.split(pattern))\n",
    "                #print(f\"--{line}--\")\n",
    "                outlines.append(line)\n",
    "        basename = re.sub('.svg.svg', '.svg', os.path.basename(fname))\n",
    "        print(basename)\n",
    "        with open(f'figs/edited_lolipop/{basename}', 'w') as fh:\n",
    "            fh.write(''.join(outlines))\n",
    "            \n",
    "insert_newlines_into_svg_files()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "import get_protein_lengths_from_tcga_lolipop_svg_files as get_prot_lens\n",
    "protein_lengths = get_prot_lens.get_protein_lengths_from_tcga_lolipop_svg_files()\n",
    "[protein_lengths.pop(name, '') for name in ['JUN', 'CDKN2A']]\n",
    "for (_name, _kda) in [(name, 110 * aa) for name, aa in protein_lengths.items()]:\n",
    "    print(f\"{_name}\\t~{_kda/1000} kDa\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Make tcga csv files for mutation data by subsetting to the RBPs of interest.\n",
    "This is as long as it is in order to use multiple input tcga csv files for a study that add genes.\n",
    "\"\"\"\n",
    "def read_into_dict(study_fname: str, our_rbps=[]):\n",
    "    \n",
    "    df = pandas.read_csv(study_fname, sep='\\t', comment='#', dtype='str')\n",
    "    try:\n",
    "        df = df.loc[[(x in our_rbps) for x in df.loc[:,'COMMON']], :]\n",
    "    except:\n",
    "        print(study_fname, \" could not be parsed.\")\n",
    "        print(df.head())\n",
    "        return {}\n",
    "    \n",
    "    df.drop_duplicates(subset='COMMON', inplace=True)\n",
    "    df.index = list(df.COMMON)\n",
    "\n",
    "    _d = df.to_dict(orient='index')\n",
    "    return _d\n",
    "\n",
    "def make_tcga_api_output_files_with_only_our_rbps(\n",
    "    input_folder: str, output_folder: str, our_rbps=[]):\n",
    "    \"\"\"Read studies into dict objects saved in output_folder/.\"\"\"\n",
    "\n",
    "    os.makedirs(output_folder, exist_ok=True)    \n",
    "    \n",
    "    study_fnames = [_ for _ in glob.glob(f\"{input_folder}/*.txt\")]\n",
    "    \n",
    "    for n, study_fname in enumerate(study_fnames):\n",
    "        print(f\"Study {n+1}/{len(study_fnames)}\", end=', ')\n",
    "        \n",
    "        _d = read_into_dict(study_fname, our_rbps=our_rbps)\n",
    "        if _d is None or len(_d) == 0:\n",
    "            continue\n",
    "            \n",
    "        basename = re.sub('all_rbps_', '', os.path.basename(study_fname))\n",
    "        out_fname = f\"{output_folder}/{basename}\"\n",
    "        with open(out_fname, 'wb') as fh:\n",
    "            pickle.dump(_d, fh)\n",
    "        #df.to_csv(out_fname, sep='\\t', index=False)\n",
    "\n",
    "def add_folder(\n",
    "    input_folder: str, output_folder: str, our_rbps=[]):\n",
    "    \n",
    "    existing_fnames = [_ for _ in glob.glob(f'{output_folder}/*txt')]\n",
    "    data = {os.path.basename(fname):pickle.load(open(fname, 'rb')) for fname in existing_fnames}    \n",
    "    \n",
    "    study_fnames = [_ for _ in glob.glob(f\"{input_folder}/*.txt\")]\n",
    "    \n",
    "    for n, study_fname in enumerate(study_fnames):\n",
    "        print(f\"Study {n+1}/{len(study_fnames)}\", end=', ')\n",
    "        \n",
    "        _d = read_into_dict(study_fname,  our_rbps=our_rbps)\n",
    "        \n",
    "        basename = os.path.basename(study_fname)\n",
    "        basename = re.sub('census_', '', basename)\n",
    "        \n",
    "        if basename in data:\n",
    "            new_genes = set(_d.keys()) - set(data[basename].keys())\n",
    "            for gene in new_genes:\n",
    "                data[basename][gene] = _d[gene]\n",
    "        \n",
    "    for fname in existing_fnames:\n",
    "        with open(fname, 'wb') as fh:\n",
    "            pickle.dump(data[os.path.basename(fname)], fh)\n",
    "                \n",
    "make_tcga_api_output_files_with_only_our_rbps(\n",
    "    'all_TCGA_data/Dec2019_api_outputs/',\n",
    "    output_folder='all_TCGA_data/Dec2019_api_outputs/our_rbps_dicts/',\n",
    "    our_rbps=list(protein_lengths.keys()))\n",
    "\n",
    "add_folder(\n",
    "    'all_TCGA_data/Jan2020_census_api_outputs/',\n",
    "    'all_TCGA_data/Dec2019_api_outputs/our_rbps_dicts/',\n",
    "    our_rbps=list(protein_lengths.keys()))\n",
    "\n",
    "fnames = [_ for _ in glob.glob('all_TCGA_data/Dec2019_api_outputs/our_rbps_dicts/*txt')]\n",
    "d = {fname:pickle.load(open(fname, 'rb')) for fname in fnames}\n",
    "os.system('cp -r all_TCGA_data/Dec2019_api_outputs/our_rbps_dicts/ all_TCGA_data/our_rbps_dict_object_backups/')\n",
    "\n",
    "output_folder = 'all_TCGA_data/Dec2019_api_outputs/our_rbps/'\n",
    "for fname, dict_obj in d.items():\n",
    "    with open(f\"{output_folder}/{os.path.basename(fname)}\", 'w') as fh:\n",
    "        a = pandas.DataFrame.from_dict(dict_obj, orient='index')\n",
    "        a.to_csv(fh, sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import get_domain_locations_as_fractions_from_tcga_lolipop_files as get_domains\n",
    "\n",
    "input_fnames = [fname for fname in glob.glob('figs/edited_lolipop/*svg')]\n",
    "input_fnames = [_ for _ in input_fnames if 'DIS3' not in _ and 'IGF2BP1' not in _]\n",
    "\n",
    "all_outlines = []\n",
    "for n, fname in enumerate(input_fnames):\n",
    "    \n",
    "    gene = re.search('lolipop/(.+)_lollipop.svg', fname).group(1)\n",
    "    #print(gene, fname)\n",
    "    gene_dict, outlines = get_domains.get_rect(fname, verbose=False)\n",
    "    #print(outlines)\n",
    "    all_outlines.extend(outlines)\n",
    "    with open('outputs/domain_positions_as_fractions.txt', 'w') as f:\n",
    "        f.write(\"Gene\\tDomain\\tStart\\tEnd\\n\")\n",
    "        f.write(''.join(all_outlines))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "study_info = pandas.read_csv('./all_TCGA_data/getCancerStudies_result.txt', sep='\\t')\n",
    "study_info = dict(zip(study_info['cancer_study_id'], study_info['name']))\n",
    "#study_info.index = study_info['cancer_study_id']\n",
    "\n",
    "def get_study_info(study_fname, study_info):\n",
    "    if '/' in study_fname:\n",
    "        study_fname = os.path.basename(study_fname)\n",
    "    study_fname = study_fname.split('.')[0]\n",
    "    study_name = re.sub('our_rbps_', '', study_fname)\n",
    "\n",
    "    if study_name in study_info:\n",
    "        #print(study_name, study_info[study_name])\n",
    "        return study_info[study_name]\n",
    "    else:\n",
    "        print(\"Couldn't find study id for\", study_fname)\n",
    "        return 'Unknown'\n",
    "        \n",
    "    \n",
    "\n",
    "for fname in glob.glob('all_TCGA_data/Dec2019_api_outputs//our_rbps/*txt'):\n",
    "    _ = get_study_info(fname, study_info)\n",
    "    #print('>', _, ', ', fname)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tcga_dir = './all_TCGA_data/Dec2019_api_outputs/our_rbps/'\n",
    "\n",
    "importlib.reload(tcgaParser)\n",
    "\n",
    "study_info = tcgaParser.currated_set_of_nonredundant_studies(\n",
    "        fname='./all_TCGA_data/currated_set_of_nonredundant_studies_list.txt',\n",
    "        study_desc_fname='cancerLists/tcga_study_ids_and_descriptions.do')\n",
    "\n",
    "dl = tcgaParser.dataLoader(study_info=study_info, missense_only=True)\n",
    "\n",
    "def muts_as_array(by_mutation):\n",
    "    gene_array = {}\n",
    "\n",
    "    \n",
    "    for (gene, mut), num_patients in by_mutation.items():\n",
    "        \n",
    "        if gene not in protein_lengths:\n",
    "            continue\n",
    "            \n",
    "        pos = re.search('\\w(\\d+)\\w', mut)\n",
    "\n",
    "        if pos is None:\n",
    "            continue\n",
    "\n",
    "        pos = int(pos.group(1)) - 1\n",
    "        if pos < 0:\n",
    "            print(gene, mut, pos)\n",
    "            \n",
    "        if gene not in gene_array:\n",
    "            gene_array[gene] = np.zeros(protein_lengths[gene])\n",
    "        if pos >= len(gene_array[gene]):\n",
    "            gene_array[gene] = np.append(\n",
    "                gene_array[gene], np.zeros(1+pos-len(gene_array[gene])))\n",
    "        gene_array[gene][pos] += len(num_patients)\n",
    "    if 'RQCD1' in gene_array:\n",
    "        gene_array['CNOT9'] = gene_array['RQCD1']\n",
    "    return gene_array\n",
    "\n",
    "def get_tcga_data_from_file(fname):\n",
    "    df = pandas.read_csv(fname, sep='\\t', comment='#')\n",
    "    print(df.shape)\n",
    "    print(df.head())\n",
    "    \n",
    "def get_tcga_data(tcga_dir, dl):\n",
    "    set_of_studies = set()\n",
    "    for fname in glob.glob(tcga_dir + '/*txt'):\n",
    "\n",
    "        set_of_studies.add(fname)\n",
    "        #break\n",
    "    dl.do_a_set_of_studies(set_of_studies)\n",
    "    print([x for x in dl.by_mutation.keys() if x[0]=='PABPC4L'])\n",
    "    #print(set([x[0] for x in  dl.by_mutation.keys()]))\n",
    "    \n",
    "get_tcga_data(tcga_dir, dl)\n",
    "print(dl.__dict__.keys())\n",
    "mutation_counts_as_arrays = muts_as_array(dl.by_mutation)\n",
    "#print(a)\n",
    "#print(mutation_counts_as_arrays)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pprint as pp\n",
    "pp.pprint(dl.mutation_to_studies[('RARS2', 'R6C')])\n",
    "\n",
    "def classify_study(study_name):\n",
    "    print(study_name)\n",
    "    study_name = re.sub(' +\\(.+\\)', '', study_name)\n",
    "    print(f\"-{study_name}-\")\n",
    "    for keyword in ['Melanoma', 'Breast Cancer']:\n",
    "        if keyword in study_name:\n",
    "            return keyword#'Bladder Urothelial Carcinoma']\n",
    "for study in dl.mutation_to_studies[('RARS2', 'R6C')].keys():\n",
    "    classify_study(study)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(dl.n_by_gene['SRSF2'])\n",
    "our_snps = pandas.read_excel('RBP_high_freq_mutations.xlsx', sheet_name='All cancer')\n",
    "def get_pos(_str):\n",
    "    m = re.search('[A-Z]+(\\d+)[A-Z]', _str)\n",
    "    if not m:\n",
    "        print(\"No SNP?\", _str)\n",
    "        return ''\n",
    "    return int(m.group(1))\n",
    "\n",
    "our_snps['Pos'] = [get_pos(str(_)) for _ in our_snps['Mutation']]\n",
    "our_snps = our_snps.loc[our_snps['Pos']!='', :]\n",
    "gene_to_our_muts = collections.defaultdict(set)\n",
    "gene_to_our_mut_string = collections.defaultdict(set)\n",
    "for row in our_snps.to_dict('records'):\n",
    "    gene_to_our_muts[row['Gene']].add(row['Pos'])\n",
    "    gene_to_our_mut_string[row['Gene']].add(row['Mutation'])\n",
    "if 'RQCD1' in gene_to_our_muts:\n",
    "    gene_to_our_muts['CNOT9'] = gene_to_our_muts['RQCD1']\n",
    "    gene_to_our_mut_string['CNOT9'] = gene_to_our_mut_string['RQCD1']\n",
    "gene_to_our_muts['FUBP1'] = set([430])\n",
    "gene_to_our_muts['RBFOX1'] = set([49])\n",
    "gene_to_our_muts['BCLAF1'] = set([163])\n",
    "print(gene_to_our_muts['YTHDC2'])\n",
    "#print(mutation_counts_as_arrays['RQCD1'])\n",
    "mut_strings = list(gene_to_our_mut_string['YTHDC2'])\n",
    "if len(mut_strings) > 1:\n",
    "    pos_to_string = dict(zip(\n",
    "        [int(re.search('[A-Z](\\d+)[A-Z]', x).group(1)) for x in mut_strings],\n",
    "        mut_strings\n",
    "        ))\n",
    "print('>>>', pos_to_string)\n",
    "#def get_muts(dl, a_prot):\n",
    "#    return [(prot, mut) for prot, mut in dl.by_mutation.keys() if prot==a_prot]\n",
    "\n",
    "print(len(gene_to_our_muts))\n",
    "\n",
    "to_max = {name:np.max(arr) for name, arr in mutation_counts_as_arrays.items()}\n",
    "by_max = sorted(\n",
    "    mutation_counts_as_arrays.keys(), key=lambda x: np.max(mutation_counts_as_arrays[x]), reverse=True)\n",
    "print(to_max)\n",
    "print(by_max)\n",
    "print(len(mutation_counts_as_arrays))\n",
    "print(mutation_counts_as_arrays.keys())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_domains = pandas.read_csv('outputs/domain_positions_as_fractions.txt', sep='\\t')\n",
    "name_to_domain_list = collections.defaultdict(list)\n",
    "print(protein_domains)\n",
    "for row in protein_domains.to_dict('records'):\n",
    "    name_to_domain_list[row['Gene']].append(row)\n",
    "#protein_domains.index = protein_domains['Gene']\n",
    "print(name_to_domain_list['PABPC4L'])\n",
    "\n",
    "\n",
    "def domain_name_to_color(name, get_legend=False):\n",
    "    current_palette = sns.color_palette(palette='Set2', n_colors=4)\n",
    "    if get_legend:\n",
    "        pat0 = plt.Rectangle((0,0), 0, 0, color=current_palette[0])\n",
    "        pat1 = plt.Rectangle((0,0), 0, 0, color=current_palette[1])\n",
    "        pat2 = plt.Rectangle((0,0), 0, 0, color=current_palette[2])\n",
    "        pat3 = plt.Rectangle((0,0), 0, 0, color=current_palette[3])\n",
    "        pat4 = plt.Rectangle((0,0), 0, 0, color='k')\n",
    "        return plt.legend((pat0, pat1, pat2, pat3, pat4), ('KH', 'RRM', 'Zinc finger', 'Helicase', 'Other'))\n",
    "    if 'KH_' in name:\n",
    "        return current_palette[0]\n",
    "    if 'RRM' in name:\n",
    "        return current_palette[1]\n",
    "    if re.search('Zf', name, re.IGNORECASE) is not None:\n",
    "        return current_palette[2]\n",
    "    if (('DEAD' in name) or ('Heli' in name)):\n",
    "        return current_palette[3]\n",
    "    else:\n",
    "        return 'k'\n",
    "    \n",
    "    \n",
    "def plot_gene(mut_arr, our_muts, ax, gene):\n",
    "    \n",
    "    y_rel = 0\n",
    "    line_len = len(mut_arr)\n",
    "    gene_body_y = max(mut_arr) * -0.07\n",
    "    gene_body_y_height = max(mut_arr) * 0.05\n",
    "    print(f\"{gene_body_y}, {gene_body_y_height}\")\n",
    "    #scale_factor = line_len/len(mut_arr)\n",
    "    line = [[0, line_len], \n",
    "           [0, 0]]\n",
    "            #[gene_body_y + gene_body_y_height * 0.5, gene_body_y + gene_body_y_height * 0.5]]\n",
    "    \n",
    "    ax.plot(line[0], line[1], 'k-', linewidth=0.5, zorder=1)\n",
    "    \n",
    "    #print('>>>dafsd>>>', gene_body_y + (gene_body_y_height * 0.5))\n",
    "    ax.plot([0, line_len],\n",
    "            [gene_body_y - (gene_body_y_height * 0.2), gene_body_y - (gene_body_y_height * 0.2)], 'k-',\n",
    "            linewidth=0.5)\n",
    "    for n, freq in enumerate(mut_arr):\n",
    "        \n",
    "        if n+1 in our_muts:\n",
    "            ax.plot([n, n], [y_rel, freq+y_rel], 'r-', linewidth=0.4)\n",
    "            \n",
    "            \n",
    "            mut_strings = list(gene_to_our_mut_string[gene])\n",
    "            if len(mut_strings) > 1:\n",
    "                pos_to_string = dict(zip(\n",
    "                    [int(re.search('[A-Z](\\d+)[A-Z]', x).group(1)) for x in mut_strings],\n",
    "                    mut_strings\n",
    "                    ))\n",
    "                mut_label = pos_to_string[n+1]\n",
    "            else:\n",
    "                mut_label = mut_strings[0]\n",
    "            ax.text(n, freq+y_rel, s=mut_label, fontdict={'fontsize': 7})\n",
    "        else:\n",
    "            ax.plot([n, n], [y_rel, freq+y_rel], 'k-', linewidth=0.6)\n",
    "\n",
    "    if gene in name_to_domain_list:\n",
    "        domains = name_to_domain_list[gene]\n",
    "        #print('adding domains...', domains)\n",
    "        \n",
    "        \n",
    "        for domain in domains:\n",
    "            box_x = domain['Start'] * line_len\n",
    "            box_width = (domain['End']-domain['Start']) * line_len\n",
    "            #print(max(mut_arr))\n",
    "            #print(box_x, box_width, \"< x, width for domain\")\n",
    "            ax.add_patch(\n",
    "                plt.Rectangle(\n",
    "                    (box_x, gene_body_y-gene_body_y_height), box_width, gene_body_y_height, \n",
    "                    color=domain_name_to_color(domain['Domain']), zorder=3) )\n",
    "    else:\n",
    "        print(f\"{gene} not in protein domain list {name_to_domain_list.keys()}\")\n",
    "    ax.set_yticks([max(mut_arr)])\n",
    "    ax.set_xticks([len(mut_arr)])\n",
    "    for tick in ax.xaxis.get_major_ticks():\n",
    "            tick.label.set_fontsize(7) \n",
    "    for tick in ax.yaxis.get_major_ticks():\n",
    "            tick.label.set_fontsize(7)\n",
    "    ax.set_ylim(gene_body_y-gene_body_y_height, ax.get_ylim()[1])\n",
    "\n",
    "    \n",
    "ax_rows = 6\n",
    "ax_cols = 6\n",
    "on_row = -1\n",
    "#sns.set_style('ticks')\n",
    "fig = plt.figure()\n",
    "domain_name_to_color('', get_legend=True)\n",
    "\n",
    "fig.savefig('./figs/legend_for_lolipop.pdf')\n",
    "plt.show()\n",
    "plt.clf()\n",
    "plt.close()\n",
    "fig, axes = plt.subplots(ax_rows, ax_cols)\n",
    "\n",
    "input_fnames = [fname for fname in glob.glob('figs/edited_lolipop/*svg')]\n",
    "# KPNB1 and DICER1 have stability data. \n",
    "for gene in ['DIS3', 'IGF2BP1', 'UPF2', 'CDKN2A', 'JUN', 'BRCA1', 'BCLAF1', 'RPL10',]:\n",
    "    input_fnames = [_ for _ in input_fnames if gene not in _]\n",
    "\n",
    "def to_max(fname):\n",
    "    gene = re.search('lolipop/(.+)_lollipop.svg', fname).group(1)\n",
    "    #to_max = {name:np.max(arr) for name, arr in mutation_counts_as_arrays.items()}\n",
    "    return np.max(mutation_counts_as_arrays[gene])\n",
    "\n",
    "input_fnames = sorted(input_fnames, key=lambda x: to_max(x), reverse=True)\n",
    "\n",
    "for n, fname in enumerate(input_fnames):\n",
    "    \n",
    "    gene = re.search('lolipop/(.+)_lollipop.svg', fname).group(1)\n",
    "    #if gene == 'CNOT9':\n",
    "    #    continue\n",
    "    if n % ax_cols == 0:\n",
    "        on_row += 1\n",
    "    on_col = n % ax_cols\n",
    "    print(fname)\n",
    "\n",
    "    ax = axes[on_row][on_col]\n",
    "    plot_gene(mutation_counts_as_arrays[gene], gene_to_our_muts[gene], ax, gene)\n",
    "\n",
    "    axes[on_row][on_col].set_title(gene, fontdict={'fontsize': 7}, y=-0.2)\n",
    "    axes[on_row][on_col].patch.set_visible(False)\n",
    "    #ax.xaxis.set_visible(False)\n",
    "    \n",
    "    ax.set_xlim(-1, len(mutation_counts_as_arrays[gene]))\n",
    "    if on_col == 0:\n",
    "        ax.set_ylabel('# Patients', fontdict={'fontsize': 7})\n",
    "    #if on_row == ax_rows - 1:\n",
    "    #ax.set_xlabel(gene, fontdict={'fontsize': 7})\n",
    "    #if on_row > 1:\n",
    "    #break\n",
    "sns.despine(bottom=True)\n",
    "fig.set_figheight(9)\n",
    "fig.set_figwidth(9)\n",
    "plt.subplots_adjust(left=0, bottom=0, right=1, top=1, wspace=0.25, hspace=0.25)\n",
    "fig.savefig('figs/combined_pops.pdf')\n",
    "plt.show()\n",
    "plt.clf()\n",
    "plt.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "max_val = max(mutation_counts_as_arrays['BCLAF1'])\n",
    "index_of_max_val = np.argmax(mutation_counts_as_arrays['BCLAF1'])\n",
    "print(f\"{max_val} at {index_of_max_val}, our mut at {gene_to_our_muts['BCLAF1']}\")\n",
    "muts = [x for x in dl.by_mutation.keys() if x[0]=='BCLAF1']\n",
    "counts = [len(dl.by_mutation[x]) for x in muts]\n",
    "_d = dict(zip(muts, counts))\n",
    "#print(muts)\n",
    "df = pandas.DataFrame.from_dict(data=_d, orient='index')\n",
    "#print(df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hess = pandas.read_excel('cancerLists/HessJM_2019.xlsx')\n",
    "prots = pandas.read_excel('Sequencing_schemes_and_results.xlsx', sheet_name='midiprep amounts')\n",
    "#print(prots)\n",
    "proteins = [x.split(' ')[0] for x in prots['Protein']] + ['PCBP1']\n",
    "df = hess.loc[[x in proteins for x in hess.gene], :]\n",
    "print(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
